// @microsoft/sp-http/lib/spHttpClient/DigestCache.js
import { _LogSource, _TraceLogger } from '@microsoft/sp-diagnostics';
import { ServiceKey, UrlUtilities, TimeProvider } from '@microsoft/sp-core-library';
import { fetchProviderServiceKey } from '../httpClient/FetchProvider';
var DigestCache =  (function () {
    function DigestCache(serviceScope) {
        var _this = this;
        this._digestsByUrl = new Map();
        serviceScope.whenFinished(function () {
            _this._fetchProvider = serviceScope.consume(fetchProviderServiceKey);
            _this._timeProvider = serviceScope.consume(TimeProvider.serviceKey);
        });
    }
    DigestCache.prototype.fetchDigest = function (webUrl) {
        var _this = this;
        var normalizedWebUrl = UrlUtilities.removeEndSlash(webUrl);
        var cachedDigest = this._digestsByUrl.get(normalizedWebUrl);
        if (cachedDigest) {
            var timestamp = this._timeProvider.getTimestamp();
            if (timestamp < cachedDigest.expirationTimestamp) {
                _TraceLogger.logVerbose(DigestCache._logSource, 'DigestCache: Reusing cached digest.  Expiration: ' + cachedDigest.expirationTimestamp);
                return Promise.resolve(cachedDigest.value);
            }
        }
        var digestUrl = normalizedWebUrl + '/_api/contextinfo';
        var rawRequest = new Request(digestUrl, {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-type': 'application/json;odata=verbose;charset=utf-8'
            },
            credentials: 'same-origin',
            cache: 'no-cache'
        });
        var preFetchDigestTime = this._timeProvider.getTimestamp();
        return this._fetchProvider.fetch(rawRequest).then(function (response) {
            return response.json();
        }).then(function (responseObject) {
            var digestValue = responseObject.FormDigestValue;
            var seconds = responseObject.FormDigestTimeoutSeconds;
            var expirationTimeMs = preFetchDigestTime +
                (1000 * seconds) -
                DigestCache.REST_EXPIRATION_SLOP_MS;
            _this.addDigestToCache(normalizedWebUrl, digestValue, expirationTimeMs);
            _TraceLogger.logVerbose(DigestCache._logSource, 'DigestCache: Fetched new digest');
            return digestValue;
        });
    };
    DigestCache.prototype.addDigestToCache = function (webUrl, digestValue, expirationTimestamp) {
        var normalizedWebUrl = UrlUtilities.removeEndSlash(webUrl);
        var newCachedDigest = {
            value: digestValue,
            expirationTimestamp: expirationTimestamp
        };
        this._digestsByUrl.set(normalizedWebUrl, newCachedDigest);
    };
    DigestCache.prototype.clearDigest = function (webUrl) {
        var normalizedWebUrl = UrlUtilities.removeEndSlash(webUrl);
        var found = this._digestsByUrl.delete(normalizedWebUrl);
        _TraceLogger.logVerbose(DigestCache._logSource, "DigestCache: Requested to clear cache entry: " + (found ? 'found' : 'not found'));
        return found;
    };
    DigestCache.prototype.clearAllDigests = function () {
        this._digestsByUrl.clear();
    };
    DigestCache.serviceKey = ServiceKey.create('sp-client-base:DigestCache', DigestCache);
    DigestCache.REST_EXPIRATION_SLOP_MS = 15000; 
    DigestCache._logSource = _LogSource.create('DigestCache');
    return DigestCache;
}());
export default DigestCache;
