// @uifabric/fluent-theme/lib-amd/fluent/styles/HoverCard.styles.js
define(["require", "exports", "tslib"], function (require, exports, tslib_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var commonCardStyles = function (props) {
        var theme = props.theme;
        var effects = theme.effects;
        return {
            border: 'none',
            boxShadow: effects.elevation16,
            borderRadius: effects.roundedCorner2,
            selectors: {
                '.ms-Callout-main': { borderRadius: effects.roundedCorner2 }
            }
        };
    };
    exports.ExpandingCardStyles = function (props) {
        return {
            root: tslib_1.__assign({}, commonCardStyles(props), { width: 320 }),
            expandedCard: {
                selectors: {
                    ':before': {
                        width: 272 // needs to change due to above change
                    }
                }
            }
        };
    };
    exports.PlainCardStyles = function (props) {
        return {
            root: tslib_1.__assign({}, commonCardStyles(props))
        };
    };
});
