// @uifabric/fluent-theme/lib-amd/fluent/styles/Pivot.styles.js
define(["require", "exports", "office-ui-fabric-react/lib/Styling"], function (require, exports, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.PivotStyles = function (props) {
        var theme = props.theme, rootIsTabs = props.rootIsTabs;
        var palette = theme.palette;
        return {
            link: [
                !rootIsTabs && {
                    height: 44,
                    lineHeight: 44,
                    selectors: {
                        ':hover': {
                            backgroundColor: palette.neutralLighter,
                            color: palette.neutralDark
                        },
                        ':active': {
                            backgroundColor: palette.neutralLight
                        }
                    }
                }
            ],
            linkIsSelected: [
                !rootIsTabs && {
                    height: 44,
                    lineHeight: 44,
                    selectors: {
                        ':hover': {
                            backgroundColor: palette.neutralLighter
                        },
                        ':active': {
                            backgroundColor: palette.neutralLight
                        },
                        ':before': {
                            transition: "left " + Styling_1.AnimationVariables.durationValue2 + " " + Styling_1.AnimationVariables.easeFunction2 + ",\n                        right " + Styling_1.AnimationVariables.durationValue2 + " " + Styling_1.AnimationVariables.easeFunction2
                        },
                        ':hover::before': {
                            left: 0,
                            right: 0
                        }
                    }
                }
            ]
        };
    };
});
