// @uifabric/fluent-theme/lib-amd/fluent/styles/TeachingBubble.styles.js
define(["require", "exports", "../FluentType", "office-ui-fabric-react/lib/Styling"], function (require, exports, FluentType_1, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TeachingBubbleStyles = function (props) {
        var hasCondensedHeadline = props.hasCondensedHeadline, hasSmallHeadline = props.hasSmallHeadline, theme = props.theme;
        var effects = theme.effects, palette = theme.palette;
        var headlineSize = FluentType_1.FontSizes.size14;
        if (!hasCondensedHeadline && !hasSmallHeadline) {
            headlineSize = FluentType_1.FontSizes.size20;
        }
        return {
            headline: {
                fontSize: headlineSize,
                fontWeight: Styling_1.FontWeights.semibold
            },
            footer: {
                selectors: {
                    '.ms-Button:not(:first-child)': {
                        marginLeft: '16px'
                    }
                }
            },
            closeButton: {
                backgroundColor: 'transparent',
                selectors: {
                    '&:hover': {
                        color: palette.black
                    },
                    '&:active': {
                        background: palette.themeDarkAlt,
                        color: palette.black
                    }
                }
            },
            subComponentStyles: {
                callout: {
                    root: {
                        boxShadow: effects.elevation16
                    }
                }
            }
        };
    };
});
