// @uifabric/fluent-theme/lib-amd/fluent/styles/Toggle.styles.js
define(["require", "exports", "office-ui-fabric-react/lib/Styling"], function (require, exports, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ToggleStyles = function (props) {
        var disabled = props.disabled, checked = props.checked, theme = props.theme;
        var palette = theme.palette;
        return {
            pill: [
                {
                    width: '40px',
                    height: '20px',
                    borderRadius: '10px',
                    padding: '0 3px'
                },
                !disabled && [
                    checked && {
                        selectors: {
                            ':hover': [
                                {
                                    backgroundColor: palette.themeDark
                                }
                            ]
                        }
                    },
                    !checked && {
                        selectors: {
                            ':hover .ms-Toggle-thumb': {
                                backgroundColor: palette.neutralDark
                            }
                        }
                    }
                ]
            ],
            thumb: [
                {
                    width: '12px',
                    height: '12px',
                    borderRadius: '12px',
                    borderColor: 'transparent'
                },
                !disabled &&
                    !checked && {
                    backgroundColor: palette.neutralSecondary
                }
            ],
            text: {
                selectors: {
                    '&.ms-Toggle-stateText': {
                        fontWeight: Styling_1.FontWeights.regular
                    }
                }
            }
        };
    };
});
