// @uifabric/utilities/lib-amd/Customizer.js
define(["require", "exports", "tslib", "react", "./BaseComponent", "./Customizations"], function (require, exports, tslib_1, React, BaseComponent_1, Customizations_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.CustomizerContext = React.createContext({
        customizations: {
            inCustomizerContext: false,
            settings: {},
            scopedSettings: {}
        }
    });
    /**
     * The Customizer component allows for default props to be mixed into components which
     * are decorated with the customizable() decorator, or use the styled HOC. This enables
     * injection scenarios like:
     *
     * 1. render svg icons instead of the icon font within all buttons
     * 2. inject a custom theme object into a component
     *
     * Props are provided via the settings prop which should be one of the following:
     * - A json map which contains 1 or more name/value pairs representing injectable props.
     * - A function that receives the current settings and returns the new ones that apply to the scope
     *
     * @public
     */
    var Customizer = /** @class */ (function (_super) {
        tslib_1.__extends(Customizer, _super);
        function Customizer() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this._changeCount = 0;
            _this._onCustomizationChange = function () { return _this.forceUpdate(); };
            return _this;
        }
        Customizer.prototype.componentDidMount = function () {
            Customizations_1.Customizations.observe(this._onCustomizationChange);
        };
        Customizer.prototype.componentWillUnmount = function () {
            Customizations_1.Customizations.unobserve(this._onCustomizationChange);
        };
        Customizer.prototype.render = function () {
            var _this = this;
            var contextTransform = this.props.contextTransform;
            return (React.createElement(exports.CustomizerContext.Consumer, null, function (parentContext) {
                var newContext = mergeCustomizations(_this.props, parentContext);
                if (contextTransform) {
                    newContext = contextTransform(newContext);
                }
                return React.createElement(exports.CustomizerContext.Provider, { value: newContext }, _this.props.children);
            }));
        };
        return Customizer;
    }(BaseComponent_1.BaseComponent));
    exports.Customizer = Customizer;
    /**
     * Merge props and customizations giving priority to props over context.
     * NOTE: This function will always perform multiple merge operations. Use with caution.
     * @param props - New settings to merge in.
     * @param parentContext - Context containing current settings.
     * @returns Merged customizations.
     */
    function mergeCustomizations(props, parentContext) {
        var _a = (parentContext || {}).customizations, customizations = _a === void 0 ? { settings: {}, scopedSettings: {} } : _a;
        return {
            customizations: {
                settings: mergeSettings(customizations.settings, props.settings),
                scopedSettings: mergeScopedSettings(customizations.scopedSettings, props.scopedSettings),
                inCustomizerContext: true
            }
        };
    }
    exports.mergeCustomizations = mergeCustomizations;
    /**
     * Merge new and old settings, giving priority to new settings.
     * New settings is optional in which case oldSettings is returned as-is.
     * @param oldSettings - Old settings to fall back to.
     * @param newSettings - New settings that will be merged over oldSettings.
     * @returns Merged settings.
     */
    function mergeSettings(oldSettings, newSettings) {
        if (oldSettings === void 0) { oldSettings = {}; }
        var mergeSettingsWith = isSettingsFunction(newSettings) ? newSettings : settingsMergeWith(newSettings);
        return mergeSettingsWith(oldSettings);
    }
    exports.mergeSettings = mergeSettings;
    function mergeScopedSettings(oldSettings, newSettings) {
        if (oldSettings === void 0) { oldSettings = {}; }
        var mergeSettingsWith = isSettingsFunction(newSettings) ? newSettings : scopedSettingsMergeWith(newSettings);
        return mergeSettingsWith(oldSettings);
    }
    function isSettingsFunction(settings) {
        return typeof settings === 'function';
    }
    function settingsMergeWith(newSettings) {
        return function (settings) { return (newSettings ? tslib_1.__assign({}, settings, newSettings) : settings); };
    }
    function scopedSettingsMergeWith(scopedSettingsFromProps) {
        if (scopedSettingsFromProps === void 0) { scopedSettingsFromProps = {}; }
        return function (oldScopedSettings) {
            var newScopedSettings = tslib_1.__assign({}, oldScopedSettings);
            for (var scopeName in scopedSettingsFromProps) {
                if (scopedSettingsFromProps.hasOwnProperty(scopeName)) {
                    newScopedSettings[scopeName] = tslib_1.__assign({}, oldScopedSettings[scopeName], scopedSettingsFromProps[scopeName]);
                }
            }
            return newScopedSettings;
        };
    }
});
