// office-ui-fabric-react/lib-es2015/components/Overlay/Overlay.base.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { BaseComponent, customizable, classNamesFunction, getNativeProps, divProperties, enableBodyScroll, disableBodyScroll } from '../../Utilities';
var getClassNames = classNamesFunction();
var OverlayBase = /** @class */ (function (_super) {
    tslib_1.__extends(OverlayBase, _super);
    function OverlayBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OverlayBase.prototype.componentDidMount = function () {
        disableBodyScroll();
    };
    OverlayBase.prototype.componentWillUnmount = function () {
        enableBodyScroll();
    };
    OverlayBase.prototype.render = function () {
        var _a = this.props, isDark = _a.isDarkThemed, className = _a.className, theme = _a.theme, getStyles = _a.getStyles;
        var divProps = getNativeProps(this.props, divProperties);
        var classNames = getClassNames(getStyles, {
            theme: theme,
            className: className,
            isDark: isDark,
        });
        return (React.createElement("div", tslib_1.__assign({}, divProps, { className: classNames.root })));
    };
    OverlayBase = tslib_1.__decorate([
        customizable('Overlay', ['theme'])
    ], OverlayBase);
    return OverlayBase;
}(BaseComponent));
export { OverlayBase };
