// office-ui-fabric-react/lib-es2015/components/Keytip/KeytipContent.base.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { BaseComponent, classNamesFunction, customizable } from '../../Utilities';
/**
 * A component corresponding the the content rendered inside the callout of the keytip component.
 *
 * @export
 * @class KeytipContent
 * @extends {BaseComponent<IKeytipProps>}
 */
var KeytipContentBase = /** @class */ (function (_super) {
    tslib_1.__extends(KeytipContentBase, _super);
    function KeytipContentBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    KeytipContentBase.prototype.render = function () {
        var _a = this.props, content = _a.content, getStyles = _a.getStyles, theme = _a.theme, disabled = _a.disabled, visible = _a.visible;
        var getClassNames = classNamesFunction();
        var classNames = getClassNames(getStyles, {
            theme: theme,
            disabled: disabled,
            visible: visible
        });
        return (React.createElement("div", { className: classNames.container },
            React.createElement("span", { className: classNames.root }, content)));
    };
    KeytipContentBase = tslib_1.__decorate([
        customizable('KeytipContent', ['theme'])
    ], KeytipContentBase);
    return KeytipContentBase;
}(BaseComponent));
export { KeytipContentBase };
