// @pnp/spfx-controls-react/lib/controls/taxonomyPicker/TaxonomyPicker.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./TaxonomyPicker.module.css');
var styles = {
    contextualMenu: 'contextualMenu_a19fc10a',
    listItem: 'listItem_a19fc10a',
    termField: 'termField_a19fc10a',
    termFieldInput: 'termFieldInput_a19fc10a',
    termFieldButton: 'termFieldButton_a19fc10a',
    termset: 'termset_a19fc10a',
    termSetSelectable: 'termSetSelectable_a19fc10a',
    termSetSelector: 'termSetSelector_a19fc10a',
    term: 'term_a19fc10a',
    termEnabled: 'termEnabled_a19fc10a',
    termDisabled: 'termDisabled_a19fc10a',
    termNoTagging: 'termNoTagging_a19fc10a',
    actions: 'actions_a19fc10a',
    termBasePicker: 'termBasePicker_a19fc10a',
    termSuggestion: 'termSuggestion_a19fc10a',
    termSuggestionSubTitle: 'termSuggestionSubTitle_a19fc10a',
    pickedTermRoot: 'pickedTermRoot_a19fc10a',
    pickedTermText: 'pickedTermText_a19fc10a',
    pickedTermCloseIcon: 'pickedTermCloseIcon_a19fc10a',
    errorMessage: 'errorMessage_a19fc10a',
    errorIcon: 'errorIcon_a19fc10a',
};
exports.default = styles;
/* tslint:enable */
