// office-ui-fabric-react/lib-amd/components/Button/SplitButton/SplitButton.styles.js
define(["require", "exports", "../../../Styling", "../../../Utilities"], function (require, exports, Styling_1, Utilities_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = Utilities_1.memoizeFunction(function (theme, customStyles) {
        var effects = theme.effects, palette = theme.palette;
        var buttonHighContrastFocus = {
            left: -2,
            top: -2,
            bottom: -2,
            right: -2,
            border: 'none'
        };
        var splitButtonDivider = {
            position: 'absolute',
            width: 1,
            right: 31,
            top: 8,
            bottom: 8
        };
        var splitButtonStyles = {
            splitButtonContainer: [
                Styling_1.getFocusStyle(theme, { highContrastStyle: buttonHighContrastFocus }),
                {
                    display: 'inline-flex',
                    selectors: {
                        '.ms-Button--default': {
                            borderTopRightRadius: '0',
                            borderBottomRightRadius: '0',
                            borderRight: 'none'
                        },
                        '.ms-Button--primary': {
                            borderTopRightRadius: '0',
                            borderBottomRightRadius: '0',
                            border: 'none'
                        },
                        '.ms-Button--primary + .ms-Button': {
                            border: 'none'
                        }
                    }
                }
            ],
            splitButtonContainerFocused: {
                outline: 'none!important'
            },
            splitButtonMenuButton: {
                padding: 6,
                height: 'auto',
                boxSizing: 'border-box',
                borderRadius: 0,
                borderTopRightRadius: effects.roundedCorner2,
                borderBottomRightRadius: effects.roundedCorner2,
                border: "1px solid " + palette.neutralSecondaryAlt,
                borderLeft: 'none',
                outline: 'transparent',
                userSelect: 'none',
                display: 'inline-block',
                textDecoration: 'none',
                textAlign: 'center',
                cursor: 'pointer',
                verticalAlign: 'top',
                width: 32,
                marginLeft: -1,
                marginTop: 0,
                marginRight: 0,
                marginBottom: 0
            },
            splitButtonDivider: splitButtonDivider,
            splitButtonDividerDisabled: splitButtonDivider,
            splitButtonMenuButtonDisabled: {
                pointerEvents: 'none',
                border: 'none',
                selectors: {
                    ':hover': {
                        cursor: 'default'
                    }
                }
            },
            splitButtonFlexContainer: {
                display: 'flex',
                height: '100%',
                flexWrap: 'nowrap',
                justifyContent: 'center',
                alignItems: 'center'
            },
            splitButtonContainerDisabled: {
                outline: 'none',
                border: 'none'
            }
        };
        return Styling_1.concatStyleSets(splitButtonStyles, customStyles);
    });
});
