// @microsoft/sp-application-base/lib/BaseApplication.js
import * as tslib_1 from "tslib";
import { override, virtual } from '@microsoft/decorators';
import { BaseComponent } from '@microsoft/sp-component-base';
import { Guid, UrlQueryParameterCollection, Validate, _SPEventManager, _SPKillSwitch, _SPFlight } from '@microsoft/sp-core-library';
import { _QosMonitor } from '@microsoft/sp-diagnostics';
import { _SPLoaderFlights } from '@microsoft/sp-loader';
import { SuiteNavManager, SuiteNavManagerConfiguration } from '@ms/sp-suite-nav';
import BaseApplicationContext from './BaseApplicationContext';
import AadPlaceholderManager from './frameworkPlaceholders/AadPlaceholderManager';
import * as loadThemedStyles from '@ms/sp-load-themed-styles';
import SPThemeProvider from './pageChrome/SPThemeProvider';
import { ApplicationLoadType } from './ApplicationLoadType';
import { Flights } from './common/Flights';
import { Killswitches } from './common/Killswitches';
import ApplicationManager from './ApplicationManager';
var SUITE_NAV_USE_SPO_BEHAVIOR_KILL_SWITCH = Guid.parse('22F8084E-9DEB-4642-B63E-E70A7F87C998');
var loadQosScenarioName = 'BaseApplication.load';
var renderQosScenarioName = 'BaseApplication.render';
var unloadQosScenarioName = 'BaseApplication.unload';
var BaseApplication =  (function (_super) {
    tslib_1.__extends(BaseApplication, _super);
    function BaseApplication() {
        var _this = _super.call(this) || this;
        _this['__type'] = 'BaseApplication';
        return _this;
    }
    Object.defineProperty(BaseApplication.prototype, "domElement", {
        get: function () {
            Validate.isNotNullOrUndefined(this.context, 'context');
            Validate.isNotDisposed(this.context, 'context');
            Validate.isNotNullOrUndefined(this.context.chrome, 'chrome');
            Validate.isNotDisposed(this.context.chrome, 'chrome');
            return this.context.chrome.appDiv;
        },
        set: function (value) {
            throw new Error('The property cannot be assigned because it is read-only');
        },
        enumerable: true,
        configurable: true
    });
    BaseApplication.prototype._load = function (contextParameters) {
        var qosMonitor = new _QosMonitor(loadQosScenarioName);
        try {
            var applicationContext = this._getApplicationContext(contextParameters);
            this._initializeContext(applicationContext);
            var aadPlaceholderManager = new AadPlaceholderManager();
            aadPlaceholderManager.setUpTokenAcquistionFailurePlaceholder(this, this.context.serviceScope);
            return this.onLoad()
                .then(function () {
                qosMonitor.writeSuccess();
            }).catch(function (e) {
                qosMonitor.writeExpectedFailure('onLoadFailure', e);
                throw e;
            });
        }
        catch (error) {
            qosMonitor.writeUnexpectedFailure('SyncError', error);
            return Promise.reject(error);
        }
    };
    BaseApplication.prototype._render = function () {
        var qosMonitor = new _QosMonitor(renderQosScenarioName);
        try {
            loadThemedStyles.flush();
            this.context.chrome.show();
            var isFullPageLoad = _SPLoaderFlights._useNewBootSequence() && Flights.useNewChromeSequence() ?
                this.context.loadType === ApplicationLoadType.FullPageLoad :
                true;
            var suiteNavManager = new SuiteNavManager(this.context.chrome.suiteNavDiv, this.context.serviceScope, isFullPageLoad);
            this.context.initializeSuiteNavManager(suiteNavManager);
            if (!this.suiteNavConfiguration().isSuiteNavDisabled()) {
                if (_SPFlight.isEnabled(1309 )
                    && this.suiteNavConfiguration().isSuiteNavLoadingDeferred()) {
                    suiteNavManager.loadSuiteNavNewFlow(this.suiteNavConfiguration()); 
                }
                else {
                    suiteNavManager.loadSuiteNav(this.suiteNavConfiguration());
                }
            }
            this.onRender();
            qosMonitor.writeSuccess();
        }
        catch (error) {
            qosMonitor.writeExpectedFailure('onRenderError', error);
        }
    };
    BaseApplication.prototype._unload = function () {
        if (!_SPLoaderFlights._useNewBootSequence()) {
            return;
        }
        var qosMonitor = new _QosMonitor(unloadQosScenarioName);
        try {
            this.onUnload();
            qosMonitor.writeSuccess();
        }
        catch (error) {
            qosMonitor.writeExpectedFailure('onUnloadError', error);
        }
    };
    BaseApplication.prototype._loadTheme = function () {
        if (Killswitches.themeProviderChromelessKSActive()
            || !ApplicationManager._isChromelessApplication(this.componentId)) {
            var themeProvider = this._getThemeProvider();
            themeProvider.loadThemedStyles();
        }
    };
    BaseApplication.prototype.dispose = function () {
        if (_SPLoaderFlights._useNewBootSequence() && this.context.chrome) {
            this.context.chrome.hide();
        }
        this.onDispose();
        _super.prototype.dispose.call(this);
    };
    BaseApplication.prototype.getBrowserCompatibility = function () {
        return {
            supportLevel: 0 ,
            warning: undefined
        };
    };
    BaseApplication.prototype.suiteNavConfiguration = function () {
        var config = new SuiteNavManagerConfiguration(this._getSuiteNavManagerConfigurationData());
        config.updateSuiteNavHeight = this._suiteNavHeightHandler.bind(this);
        return config;
    };
    BaseApplication.prototype._getApplicationContext = function (contextParameters) {
        return new BaseApplicationContext(contextParameters);
    };
    BaseApplication.prototype._getThemeProvider = function () {
        return new SPThemeProvider(this.context.serviceScope);
    };
    BaseApplication.prototype.onLoad = function () {
        return Promise.resolve();
    };
    BaseApplication.prototype.onUnload = function () {
    };
    BaseApplication.prototype.onRender = function () {
    };
    BaseApplication.prototype.onDispose = function () {
    };
    BaseApplication.prototype._getSuiteNavManagerConfigurationData = function () {
        var pageContext = this.context.pageContext;
        var webTemplateId;
        if (!_SPKillSwitch.isActivated(SUITE_NAV_USE_SPO_BEHAVIOR_KILL_SWITCH, '9/08/2017', 'SuiteNavUseSPOBehaviors')) {
            webTemplateId = pageContext.legacyPageContext.webTemplateId;
        }
        return {
            currentUICultureName: pageContext.cultureInfo.currentUICultureName,
            disableSuiteNav: this._shouldDisableSuiteNav(),
            settingsData: pageContext.legacyPageContext.MenuData ?
                pageContext.legacyPageContext.MenuData.SettingsData :
                undefined,
            signoutUrl: pageContext.legacyPageContext.MenuData ?
                pageContext.legacyPageContext.MenuData.SignOutUrl :
                undefined,
            siteClientTag: pageContext.legacyPageContext.siteClientTag,
            systemUserKey: pageContext.legacyPageContext.systemUserKey,
            userDisplayName: pageContext.user.displayName,
            webServerRelativeUrl: pageContext.web.serverRelativeUrl,
            webTemplateId: webTemplateId
        };
    };
    BaseApplication.prototype._navigate = function (url, props) {
        var _this = this;
        Validate.isNonemptyString(url, 'url');
        return this.context.navigator.navigate(url, props).then(function (res) {
            if (!_SPLoaderFlights._useNewBootSequence()) {
                _this.context.navigator._loadApplicationCustomizers(res.preloadedData);
            }
            return res;
        });
    };
    BaseApplication.prototype._navigateToPreloadedData = function (preloadedData) {
        var _this = this;
        Validate.isNotNullOrUndefined(preloadedData, 'preloadedData');
        this.context.navigator.navigateToPreloadedData(preloadedData).then(function () {
            if (!_SPLoaderFlights._useNewBootSequence()) {
                _this.context.navigator._loadApplicationCustomizers(preloadedData);
            }
        });
    };
    BaseApplication.prototype._invalidate = function (url) {
        Validate.isNonemptyString(url, 'url');
        return this.context.navigator.invalidate(url);
    };
    BaseApplication.prototype.raiseLayoutChangedEvent = function () {
        _SPEventManager.instance.raiseEvent(BaseApplication._layoutChangedEventName, {});
    };
    BaseApplication.prototype._shouldDisableSuiteNav = function () {
        var urlQueryParams = new UrlQueryParameterCollection(window.location.href);
        return window.location.hostname === 'localhost' ||
            urlQueryParams.getValue('disableSuiteNav') === 'true' ?
            true :
            false;
    };
    BaseApplication.prototype._suiteNavHeightHandler = function (height) {
        if (this.context.chrome) {
            this.context.chrome.changeSuiteNavHeight(height);
        }
    };
    BaseApplication._navigatedEventName = 'application.navigatedEvent';
    BaseApplication._layoutChangedEventName = 'application.layoutChangedEvent';
    BaseApplication._prefetchedDataEventName = 'application.prefetchedDataEvent';
    tslib_1.__decorate([
        override
    ], BaseApplication.prototype, "dispose", null);
    tslib_1.__decorate([
        virtual
    ], BaseApplication.prototype, "getBrowserCompatibility", null);
    tslib_1.__decorate([
        virtual
    ], BaseApplication.prototype, "suiteNavConfiguration", null);
    tslib_1.__decorate([
        virtual
    ], BaseApplication.prototype, "_getApplicationContext", null);
    tslib_1.__decorate([
        virtual
    ], BaseApplication.prototype, "_getThemeProvider", null);
    tslib_1.__decorate([
        virtual
    ], BaseApplication.prototype, "onLoad", null);
    tslib_1.__decorate([
        virtual
    ], BaseApplication.prototype, "onUnload", null);
    tslib_1.__decorate([
        virtual
    ], BaseApplication.prototype, "onRender", null);
    tslib_1.__decorate([
        virtual
    ], BaseApplication.prototype, "onDispose", null);
    return BaseApplication;
}(BaseComponent));
export default BaseApplication;
