// @microsoft/sp-webpart-base/lib/SPPropertyPane.js
import { PropertyPaneCustomField as SPPropertyPane_PropertyPaneCustomField } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneCustomField/PropertyPaneCustomField';
import { PropertyPaneButton as SPPropertyPane_PropertyPaneButton } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneButton/PropertyPaneButton';
import { PropertyPaneCheckbox as SPPropertyPane_PropertyPaneCheckbox } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneCheckBox/PropertyPaneCheckbox';
import { PropertyPaneChoiceGroup as SPPropertyPane_PropertyPaneChoiceGroup } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneChoiceGroup/PropertyPaneChoiceGroup';
import { PropertyPaneDropdown as SPPropertyPane_PropertyPaneDropdown } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneDropdown/PropertyPaneDropdown';
import { PropertyPaneDynamicField as SPPropertyPane_PropertyPaneDynamicField } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneDynamicField/PropertyPaneDynamicField';
import { PropertyPaneDynamicFieldSet as SPPropertyPane_PropertyPaneDynamicFieldSet } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneDynamicFieldSet/PropertyPaneDynamicFieldSet';
import { PropertyPaneHorizontalRule as SPPropertyPane_PropertyPaneHorizontalRule } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneHorizontalRule/PropertyPaneHorizontalRule';
import { PropertyPaneLabel as SPPropertyPane_PropertyPaneLabel } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneLabel/PropertyPaneLabel';
import { PropertyPaneToggle as SPPropertyPane_PropertyPaneToggle } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneToggle/PropertyPaneToggle';
import { PropertyPaneDynamicTextField as SPPropertyPane_PropertyPaneDynamicTextField } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneDynamicTextField/PropertyPaneDynamicTextField';
import { PropertyPaneTextField as SPPropertyPane_PropertyPaneTextField } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneTextField/PropertyPaneTextField';
import { PropertyPaneLink as SPPropertyPane_PropertyPaneLink } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneLink/PropertyPaneLink';
import { PropertyPaneSlider as SPPropertyPane_PropertyPaneSlider } from '@microsoft/sp-property-pane/lib/propertyPaneFields/propertyPaneSlider/PropertyPaneSlider';
export function PropertyPaneCustomField(properties) {
    return SPPropertyPane_PropertyPaneCustomField(properties);
}
export function PropertyPaneButton(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneButton(targetProperty, properties);
}
export function PropertyPaneCheckbox(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneCheckbox(targetProperty, properties);
}
export function PropertyPaneChoiceGroup(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneChoiceGroup(targetProperty, properties);
}
export function PropertyPaneDropdown(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneDropdown(targetProperty, properties);
}
export function PropertyPaneDynamicField(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneDynamicField(targetProperty, properties);
}
export function PropertyPaneDynamicFieldSet(properties) {
    return SPPropertyPane_PropertyPaneDynamicFieldSet(properties);
}
export function PropertyPaneHorizontalRule() {
    return SPPropertyPane_PropertyPaneHorizontalRule();
}
export function PropertyPaneLabel(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneLabel(targetProperty, properties);
}
export function PropertyPaneLink(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneLink(targetProperty, properties);
}
export function PropertyPaneSlider(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneSlider(targetProperty, properties);
}
export function PropertyPaneTextField(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneTextField(targetProperty, properties);
}
export function PropertyPaneToggle(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneToggle(targetProperty, properties);
}
export function PropertyPaneDynamicTextField(targetProperty, properties) {
    return SPPropertyPane_PropertyPaneDynamicTextField(targetProperty, properties);
}
export var PropertyPaneFieldType;
(function (PropertyPaneFieldType) {
    PropertyPaneFieldType[PropertyPaneFieldType["Custom"] = 1] = "Custom";
    PropertyPaneFieldType[PropertyPaneFieldType["CheckBox"] = 2] = "CheckBox";
    PropertyPaneFieldType[PropertyPaneFieldType["TextField"] = 3] = "TextField";
    PropertyPaneFieldType[PropertyPaneFieldType["Toggle"] = 5] = "Toggle";
    PropertyPaneFieldType[PropertyPaneFieldType["Dropdown"] = 6] = "Dropdown";
    PropertyPaneFieldType[PropertyPaneFieldType["Label"] = 7] = "Label";
    PropertyPaneFieldType[PropertyPaneFieldType["Slider"] = 8] = "Slider";
    PropertyPaneFieldType[PropertyPaneFieldType["Heading"] = 9] = "Heading";
    PropertyPaneFieldType[PropertyPaneFieldType["ChoiceGroup"] = 10] = "ChoiceGroup";
    PropertyPaneFieldType[PropertyPaneFieldType["Button"] = 11] = "Button";
    PropertyPaneFieldType[PropertyPaneFieldType["HorizontalRule"] = 12] = "HorizontalRule";
    PropertyPaneFieldType[PropertyPaneFieldType["Link"] = 13] = "Link";
    PropertyPaneFieldType[PropertyPaneFieldType["DynamicField"] = 14] = "DynamicField";
    PropertyPaneFieldType[PropertyPaneFieldType["DynamicTextField"] = 15] = "DynamicTextField";
    PropertyPaneFieldType[PropertyPaneFieldType["DynamicFieldSet"] = 16] = "DynamicFieldSet";
    PropertyPaneFieldType[PropertyPaneFieldType["SpinButton"] = 17] = "SpinButton";
})(PropertyPaneFieldType || (PropertyPaneFieldType = {}));
export var PropertyPaneButtonType;
(function (PropertyPaneButtonType) {
    PropertyPaneButtonType[PropertyPaneButtonType["Normal"] = 0] = "Normal";
    PropertyPaneButtonType[PropertyPaneButtonType["Primary"] = 1] = "Primary";
    PropertyPaneButtonType[PropertyPaneButtonType["Hero"] = 2] = "Hero";
    PropertyPaneButtonType[PropertyPaneButtonType["Compound"] = 3] = "Compound";
    PropertyPaneButtonType[PropertyPaneButtonType["Command"] = 4] = "Command";
    PropertyPaneButtonType[PropertyPaneButtonType["Icon"] = 5] = "Icon";
})(PropertyPaneButtonType || (PropertyPaneButtonType = {}));
export var PropertyPaneDropdownOptionType;
(function (PropertyPaneDropdownOptionType) {
    PropertyPaneDropdownOptionType[PropertyPaneDropdownOptionType["Normal"] = 0] = "Normal";
    PropertyPaneDropdownOptionType[PropertyPaneDropdownOptionType["Divider"] = 1] = "Divider";
    PropertyPaneDropdownOptionType[PropertyPaneDropdownOptionType["Header"] = 2] = "Header";
})(PropertyPaneDropdownOptionType || (PropertyPaneDropdownOptionType = {}));
export var DynamicDataSharedDepth;
(function (DynamicDataSharedDepth) {
    DynamicDataSharedDepth[DynamicDataSharedDepth["None"] = 0] = "None";
    DynamicDataSharedDepth[DynamicDataSharedDepth["Source"] = 1] = "Source";
    DynamicDataSharedDepth[DynamicDataSharedDepth["Property"] = 2] = "Property";
})(DynamicDataSharedDepth || (DynamicDataSharedDepth = {}));
