// office-ui-fabric-react/lib-es2015/components/ContextualMenu/examples/ContextualMenu.CustomMenuItem.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { ContextualMenuItemType } from 'office-ui-fabric-react/lib/ContextualMenu';
import { DefaultButton } from 'office-ui-fabric-react/lib/Button';
import './ContextualMenuExample.scss';
var ContextualMenuWithCustomMenuItemExample = /** @class */ (function (_super) {
    tslib_1.__extends(ContextualMenuWithCustomMenuItemExample, _super);
    function ContextualMenuWithCustomMenuItemExample(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            showCallout: false
        };
        return _this;
    }
    ContextualMenuWithCustomMenuItemExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(DefaultButton, { id: 'ContextualMenuButton1', text: 'Click for ContextualMenu', menuProps: {
                    shouldFocusOnMount: true,
                    items: [
                        {
                            key: 'newItem',
                            name: 'New'
                        },
                        {
                            key: 'divider_1',
                            itemType: ContextualMenuItemType.Divider
                        },
                        {
                            key: 'rename',
                            name: 'Rename'
                        },
                        {
                            key: 'edit',
                            name: 'Edit'
                        },
                        {
                            key: 'properties',
                            name: 'Properties'
                        },
                        {
                            key: 'disabled',
                            name: 'Disabled item',
                            disabled: true
                        }
                    ],
                    contextualMenuItemAs: function (props) { return React.createElement("div", null,
                        "Custom rendered ",
                        props.item.name); }
                } })));
    };
    return ContextualMenuWithCustomMenuItemExample;
}(React.Component));
export { ContextualMenuWithCustomMenuItemExample };
