// @pnp/spfx-property-controls/lib/propertyFields/dateTimePicker/PropertyFieldDateTimePicker.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var DateValues_1 = require("office-ui-fabric-react/lib/utilities/dateValues/DateValues");
var React = require("react");
var ReactDom = require("react-dom");
var sp_webpart_base_1 = require("@microsoft/sp-webpart-base");
var PropertyFieldDateTimePickerHost_1 = require("./PropertyFieldDateTimePickerHost");
var IPropertyFieldDateTimePicker_1 = require("./IPropertyFieldDateTimePicker");
/**
 * Represents a PropertyFieldDateTimePicker object
 */
var PropertyFieldDateTimePickerBuilder = (function () {
    /**
     * Constructor
     */
    function PropertyFieldDateTimePickerBuilder(_targetProperty, _properties) {
        // Properties defined by IPropertyPaneField
        this.type = sp_webpart_base_1.PropertyPaneFieldType.Custom;
        this.disabled = false;
        this.deferredValidationTime = 200;
        this.render = this.render.bind(this);
        this.targetProperty = _properties.targetProperty;
        this.properties = _properties;
        this.label = _properties.label;
        this.initialDate = _properties.initialDate;
        this.properties.onDispose = this.dispose;
        this.properties.onRender = this.render;
        this.onPropertyChange = _properties.onPropertyChange;
        this.formatDate = _properties.formatDate;
        this.customProperties = _properties.properties;
        this.key = _properties.key;
        this.onGetErrorMessage = _properties.onGetErrorMessage;
        if (_properties.deferredValidationTime) {
            this.deferredValidationTime = _properties.deferredValidationTime;
        }
        if (typeof _properties.disabled !== 'undefined') {
            this.disabled = _properties.disabled;
        }
        if (typeof _properties.dateConvention !== 'undefined') {
            this.dateConvention = _properties.dateConvention;
        }
        else {
            this.dateConvention = IPropertyFieldDateTimePicker_1.DateConvention.DateTime;
        }
        if (typeof _properties.timeConvention !== 'undefined') {
            this.timeConvention = _properties.timeConvention;
        }
        else {
            this.timeConvention = IPropertyFieldDateTimePicker_1.TimeConvention.Hours24;
        }
        if (typeof _properties.firstDayOfWeek !== 'undefined') {
            this.firstDayOfWeek = _properties.firstDayOfWeek;
        }
        else {
            this.firstDayOfWeek = DateValues_1.DayOfWeek.Sunday;
        }
    }
    /**
     * Renders the DatePicker field content
     */
    PropertyFieldDateTimePickerBuilder.prototype.render = function (elem, ctx, changeCallback) {
        // Construct the JSX properties
        var element = React.createElement(PropertyFieldDateTimePickerHost_1.default, {
            label: this.label,
            disabled: this.disabled,
            initialDate: this.initialDate,
            targetProperty: this.targetProperty,
            formatDate: this.formatDate,
            dateConvention: this.dateConvention,
            timeConvention: this.timeConvention,
            firstDayOfWeek: this.firstDayOfWeek,
            onDispose: this.dispose,
            onRender: this.render,
            onPropertyChange: this.onPropertyChange,
            onChange: changeCallback,
            properties: this.customProperties,
            key: this.key,
            onGetErrorMessage: this.onGetErrorMessage,
            deferredValidationTime: this.deferredValidationTime
        });
        // Calls the REACT content generator
        ReactDom.render(element, elem);
    };
    /**
     * Disposes the current object
     */
    PropertyFieldDateTimePickerBuilder.prototype.dispose = function (elem) {
    };
    return PropertyFieldDateTimePickerBuilder;
}());
/**
 * Helper method to create the customer field on the PropertyPane.
 * @param targetProperty - Target property the custom field is associated to.
 * @param properties - Strongly typed custom field properties.
 */
function PropertyFieldDateTimePicker(targetProperty, properties) {
    // Calls the PropertyFieldDateTimePicker builder object
    // This object will simulate a PropertyFieldCustom to manage his rendering process
    return new PropertyFieldDateTimePickerBuilder(targetProperty, __assign({}, properties, { targetProperty: targetProperty, onDispose: null, onRender: null }));
}
exports.PropertyFieldDateTimePicker = PropertyFieldDateTimePicker;
