// office-ui-fabric-react/lib/components/Button/examples/Button.Compound.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { css, classNamesFunction, CompoundButton, Label } from 'office-ui-fabric-react';
var exampleStyles = {
    example: [
        'ms-BasicButtonsExample',
        {
            selectors: {
                '.ms-Button': {
                    margin: '10px 0'
                }
            }
        }
    ],
    twoup: [
        'ms-BasicButtonsTwoUp',
        {
            display: 'flex',
            selectors: {
                '& > *': {
                    flexGrow: 1
                },
                '.ms-Label': {
                    marginBottom: '10px'
                }
            }
        }
    ]
};
var getClassNames = classNamesFunction();
var classNames = getClassNames(exampleStyles, {});
var ButtonCompoundExample = /** @class */ (function (_super) {
    tslib_1.__extends(ButtonCompoundExample, _super);
    function ButtonCompoundExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ButtonCompoundExample.prototype.render = function () {
        var _a = this.props, disabled = _a.disabled, checked = _a.checked;
        return (React.createElement("div", { className: css(classNames.example, classNames.twoup) },
            React.createElement("div", null,
                React.createElement(Label, null, "Standard"),
                React.createElement(CompoundButton, { secondaryText: "You can create a new account here.", disabled: disabled, checked: checked }, "Create account")),
            React.createElement("div", null,
                React.createElement(Label, null, "Primary"),
                React.createElement(CompoundButton, { primary: true, secondaryText: "You can create a new account here.", disabled: disabled, checked: checked }, "Create account"))));
    };
    return ButtonCompoundExample;
}(React.Component));
export { ButtonCompoundExample };
