// office-ui-fabric-react/lib/components/Button/examples/Button.Default.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { css, classNamesFunction, DefaultButton, Label, PrimaryButton } from 'office-ui-fabric-react';
var exampleStyles = {
    twoup: [
        'ms-BasicButtonsTwoUp',
        {
            display: 'flex',
            selectors: {
                '& > *': {
                    flexGrow: 1
                },
                '.ms-Label': {
                    marginBottom: '10px'
                }
            }
        }
    ]
};
var getClassNames = classNamesFunction();
var classNames = getClassNames(exampleStyles, {});
var ButtonDefaultExample = /** @class */ (function (_super) {
    tslib_1.__extends(ButtonDefaultExample, _super);
    function ButtonDefaultExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ButtonDefaultExample.prototype.render = function () {
        var _a = this.props, disabled = _a.disabled, checked = _a.checked;
        return (React.createElement("div", { className: css(classNames.twoup) },
            React.createElement("div", null,
                React.createElement(Label, null, "Standard"),
                React.createElement(DefaultButton, { "data-automation-id": "test", allowDisabledFocus: true, disabled: disabled, checked: checked, text: "Button", onClick: this._alertClicked })),
            React.createElement("div", null,
                React.createElement(Label, null, "Primary"),
                React.createElement(PrimaryButton, { "data-automation-id": "test", disabled: disabled, checked: checked, text: "Button", onClick: this._alertClicked, allowDisabledFocus: true }))));
    };
    ButtonDefaultExample.prototype._alertClicked = function () {
        alert('Clicked');
    };
    return ButtonDefaultExample;
}(React.Component));
export { ButtonDefaultExample };
