// office-ui-fabric-react/lib-amd/components/ScrollablePane/examples/ScrollablePane.Default.Example.js
define(["require", "exports", "tslib", "react", "office-ui-fabric-react/lib/ScrollablePane", "office-ui-fabric-react/lib/Sticky", "@uifabric/example-app-base", "./ScrollablePane.Example.scss"], function (require, exports, tslib_1, React, ScrollablePane_1, Sticky_1, example_app_base_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var colors = [
        '#eaeaea',
        '#dadada',
        '#d0d0d0',
        '#c8c8c8',
        '#a6a6a6',
        '#c7e0f4',
        '#71afe5',
        '#eff6fc',
        '#deecf9'
    ];
    var ScrollablePaneDefaultExample = /** @class */ (function (_super) {
        tslib_1.__extends(ScrollablePaneDefaultExample, _super);
        function ScrollablePaneDefaultExample() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        ScrollablePaneDefaultExample.prototype.render = function () {
            var contentAreas = [];
            for (var i = 0; i < 5; i++) {
                contentAreas.push(this._createContentArea(i));
            }
            return (React.createElement("div", { style: {
                    height: '900px',
                    position: 'relative',
                    maxHeight: 'inherit'
                } },
                React.createElement(ScrollablePane_1.ScrollablePane, { className: 'scrollablePaneDefaultExample' }, contentAreas.map(function (ele) {
                    return ele;
                }))));
        };
        ScrollablePaneDefaultExample.prototype._createContentArea = function (index) {
            var color = colors.splice(Math.floor(Math.random() * colors.length), 1)[0];
            return (React.createElement("div", { key: index, style: {
                    backgroundColor: color
                } },
                React.createElement(Sticky_1.Sticky, { stickyPosition: Sticky_1.StickyPositionType.Both },
                    React.createElement("div", { className: 'sticky' },
                        "Sticky Component #",
                        index + 1)),
                React.createElement("div", { className: 'textContent' }, example_app_base_1.lorem(200))));
        };
        return ScrollablePaneDefaultExample;
    }(React.Component));
    exports.ScrollablePaneDefaultExample = ScrollablePaneDefaultExample;
});
