// office-ui-fabric-react/lib-amd/demo/ComponentStatus/AllComponents.checklist.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AllComponentsStatus = {
        ActivityItem: require('../../components/ActivityItem/ActivityItem.checklist').ActivityItemStatus,
        Breadcrumb: require('../../components/Breadcrumb/Breadcrumb.checklist').BreadcrumbStatus,
        Button: require('../../components/Button/Button.checklist').ButtonStatus,
        Calendar: require('../../components/Calendar/Calendar.checklist').CalendarStatus,
        Callout: require('../../components/Callout/Callout.checklist').CalloutStatus,
        Checkbox: require('../../components/Checkbox/Checkbox.checklist').CheckboxStatus,
        ChoiceGroup: require('../../components/ChoiceGroup/ChoiceGroup.checklist').ChoiceGroupStatus,
        ColorPicker: require('../../components/ColorPicker/ColorPicker.checklist').ColorPickerStatus,
        ComboBox: require('../../components/ComboBox/ComboBox.checklist').ComboBoxStatus,
        CommandBar: require('../../components/CommandBar/CommandBar.checklist').CommandBarStatus,
        ContextualMenu: require('../../components/ContextualMenu/ContextualMenu.checklist').ContextualMenuStatus,
        DatePicker: require('../../components/DatePicker/DatePicker.checklist').DatePickerStatus,
        DetailsList: require('../../components/DetailsList/DetailsList.checklist').DetailsListStatus,
        Dialog: require('../../components/Dialog/Dialog.checklist').DialogStatus,
        DocumentCard: require('../../components/Calendar/Calendar.checklist').CalendarStatus,
        Dropdown: require('../../components/Dropdown/Dropdown.checklist').DropdownStatus,
        Facepile: require('../../components/Facepile/Facepile.checklist').FacepileStatus,
        GroupedList: require('../../components/GroupedList/GroupedList.checklist').GroupedListStatus,
        HoverCard: require('../../components/HoverCard/HoverCard.checklist').HoverCardStatus,
        Icon: require('../../components/Icon/Icon.checklist').IconStatus,
        Image: require('../../components/Image/Image.checklist').ImageStatus,
        Label: require('../../components/Label/Label.checklist').LabelStatus,
        Layer: require('../../components/Layer/Layer.checklist').LayerStatus,
        Link: require('../../components/Link/Link.checklist').LinkStatus,
        List: require('../../components/List/List.checklist').ListStatus,
        MessageBar: require('../../components/MessageBar/MessageBar.checklist').MessageBarStatus,
        Modal: require('../../components/Modal/Modal.checklist').ModalStatus,
        Nav: require('../../components/Nav/Nav.checklist').NavStatus,
        Overlay: require('../../components/Overlay/Overlay.checklist').OverlayStatus,
        OverflowSet: require('../../components/OverflowSet/OverflowSet.checklist').OverflowSetStatus,
        Panel: require('../../components/Panel/Panel.checklist').PanelStatus,
        Persona: require('../../components/Persona/Persona.checklist').PersonaStatus,
        Pickers: require('../../components/pickers/Pickers.checklist').PickersStatus,
        PeoplePicker: require('../../components/pickers/PeoplePicker/PeoplePicker.checklist').PeoplePickerStatus,
        Pivot: require('../../components/Pivot/Pivot.checklist').PivotStatus,
        ProgressIndicator: require('../../components/ProgressIndicator/ProgressIndicator.checklist').ProgressIndicatorStatus,
        Rating: require('../../components/Rating/Rating.checklist').RatingStatus,
        ResizeGroup: require('../../components/ResizeGroup/ResizeGroup.checklist').ResizeGroupStatus,
        ScrollablePane: require('../../components/ScrollablePane/ScrollablePane.checklist').ScrollablePaneStatus,
        SearchBox: require('../../components/SearchBox/SearchBox.checklist').SearchBoxStatus,
        Shimmer: require('../../components/Shimmer/Shimmer.checklist').ShimmerStatus,
        Slider: require('../../components/Slider/Slider.checklist').SliderStatus,
        Spinner: require('../../components/Spinner/Spinner.checklist').SpinnerStatus,
        SpinButton: require('../../components/SpinButton/SpinButton.checklist').SpinButtonStatus,
        SwatchColorPicker: require('../../components/SwatchColorPicker/SwatchColorPicker.checklist').SwatchColorPickerStatus,
        TeachingBubble: require('../../components/TeachingBubble/TeachingBubble.checklist').TeachingBubbleStatus,
        TextField: require('../../components/TextField/TextField.checklist').TextFieldStatus,
        Toggle: require('../../components/Toggle/Toggle.checklist').ToggleStatus,
        Tooltip: require('../../components/Tooltip/Tooltip.checklist').TooltipStatus,
    };
});
