// office-ui-fabric-react/lib-commonjs/components/Stack/examples/Stack.Horizontal.Shrink.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var Slider_1 = require("office-ui-fabric-react/lib/Slider");
var Stack_1 = require("../Stack");
var Styling_1 = require("office-ui-fabric-react/lib/Styling");
var HorizontalStackShrinkExample = /** @class */ (function (_super) {
    tslib_1.__extends(HorizontalStackShrinkExample, _super);
    function HorizontalStackShrinkExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onWidthChange = function (value) {
            _this.setState({ stackWidth: value });
        };
        _this.state = {
            stackWidth: 100
        };
        return _this;
    }
    HorizontalStackShrinkExample.prototype.render = function () {
        var stackStyles = {
            root: {
                background: Styling_1.DefaultPalette.themeTertiary,
                overflow: 'hidden',
                width: this.state.stackWidth + "%"
            }
        };
        var stackItemStyles = {
            root: {
                alignItems: 'center',
                background: Styling_1.DefaultPalette.themePrimary,
                color: Styling_1.DefaultPalette.white,
                display: 'flex',
                height: 50,
                justifyContent: 'center',
                overflow: 'hidden'
            }
        };
        var nonShrinkingStackItemStyles = {
            root: {
                alignItems: 'center',
                background: Styling_1.DefaultPalette.themePrimary,
                color: Styling_1.DefaultPalette.white,
                display: 'flex',
                height: 50,
                justifyContent: 'center',
                overflow: 'hidden',
                width: 500
            }
        };
        var outerStackTokens = { childrenGap: 5 };
        var innerStackTokens = {
            childrenGap: 5,
            padding: 10
        };
        return (React.createElement(Stack_1.Stack, { tokens: outerStackTokens },
            React.createElement(Slider_1.Slider, { label: "Change the stack width to see how child items shrink:", min: 1, max: 100, step: 1, defaultValue: 100, showValue: true, onChange: this._onWidthChange }),
            React.createElement(Stack_1.Stack, { horizontal: true, styles: stackStyles, tokens: innerStackTokens },
                React.createElement(Stack_1.Stack.Item, { grow: true, styles: stackItemStyles }, "I shrink"),
                React.createElement(Stack_1.Stack.Item, { grow: true, styles: stackItemStyles }, "I shrink"),
                React.createElement(Stack_1.Stack.Item, { grow: true, disableShrink: true, styles: nonShrinkingStackItemStyles }, "I don't shrink"),
                React.createElement(Stack_1.Stack.Item, { grow: true, styles: stackItemStyles }, "I shrink"))));
    };
    return HorizontalStackShrinkExample;
}(React.Component));
exports.HorizontalStackShrinkExample = HorizontalStackShrinkExample;
