// office-ui-fabric-react/lib-amd/components/Dialog/examples/Dialog.Basic.Example.js
define(["require", "exports", "tslib", "react", "office-ui-fabric-react/lib/Dialog", "office-ui-fabric-react/lib/Button", "./Dialog.Basic.Example.scss"], function (require, exports, tslib_1, React, Dialog_1, Button_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DialogBasicExample = /** @class */ (function (_super) {
        tslib_1.__extends(DialogBasicExample, _super);
        function DialogBasicExample(props) {
            var _this = _super.call(this, props) || this;
            _this._showDialog = function () {
                _this.setState({ hideDialog: false });
            };
            _this._closeDialog = function () {
                _this.setState({ hideDialog: true });
            };
            _this.state = {
                hideDialog: true
            };
            return _this;
        }
        DialogBasicExample.prototype.render = function () {
            return (React.createElement("div", null,
                React.createElement(Button_1.DefaultButton, { secondaryText: 'Opens the Sample Dialog', onClick: this._showDialog, text: 'Open Dialog' }),
                React.createElement("label", { id: 'myLabelId', className: 'screenReaderOnly' }, "My sample Label"),
                React.createElement("label", { id: 'mySubTextId', className: 'screenReaderOnly' }, "My Sample description"),
                React.createElement(Dialog_1.Dialog, { hidden: this.state.hideDialog, onDismiss: this._closeDialog, dialogContentProps: {
                        type: Dialog_1.DialogType.normal,
                        title: 'All emails together',
                        subText: 'Your Inbox has changed. No longer does it include favorites, it is a singular destination for your emails.'
                    }, modalProps: {
                        titleAriaId: 'myLabelId',
                        subtitleAriaId: 'mySubTextId',
                        isBlocking: false,
                        containerClassName: 'ms-dialogMainOverride'
                    } },
                    null /** You can also include null values as the result of conditionals */,
                    React.createElement(Dialog_1.DialogFooter, null,
                        React.createElement(Button_1.PrimaryButton, { onClick: this._closeDialog, text: 'Save' }),
                        React.createElement(Button_1.DefaultButton, { onClick: this._closeDialog, text: 'Cancel' })))));
        };
        return DialogBasicExample;
    }(React.Component));
    exports.DialogBasicExample = DialogBasicExample;
});
