// @uifabric/fluent-theme/lib-commonjs/fluent/styles/CompoundButton.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Styling_1 = require("office-ui-fabric-react/lib/Styling");
exports.CompoundButtonStyles = function (props) {
    var theme = props.theme;
    if (!theme) {
        throw new Error('Theme is undefined or null.');
    }
    var palette = theme.palette, effects = theme.effects;
    return {
        root: tslib_1.__assign({}, Styling_1.getFocusStyle(theme, { inset: 2 }), { backgroundColor: palette.white, border: "1px solid " + palette.neutralSecondaryAlt, borderRadius: effects.roundedCorner2, padding: '16px 12px', 
            // Primary styles require targeting a selector for now.
            // @todo: These selectors override the focus style above. Need to fix this.
            selectors: {
                '&.ms-Button--compoundPrimary': {
                    backgroundColor: palette.themePrimary,
                    borderColor: palette.themePrimary
                }
            } }),
        rootPressed: {
            backgroundColor: palette.neutralQuaternaryAlt,
            // Primary styles require targeting a selector for now.
            selectors: {
                '&.ms-Button--compoundPrimary:active': {
                    backgroundColor: palette.themeDark
                }
            }
        },
        rootChecked: {
            backgroundColor: palette.neutralQuaternaryAlt,
            // Primary styles require targeting a selector for now.
            selectors: {
                '&.ms-Button--compoundPrimary': {
                    backgroundColor: palette.themeDark,
                    borderColor: palette.themeDark
                }
            }
        },
        rootDisabled: {
            borderColor: palette.neutralLighter,
            selectors: {
                '&.ms-Button--compoundPrimary': {
                    backgroundColor: palette.neutralLighter,
                    borderColor: palette.neutralLighter
                }
            }
        }
    };
};
