// @uifabric/fluent-theme/lib-commonjs/fluent/styles/Dialog.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var styling_1 = require("@uifabric/styling");
var FluentType_1 = require("../FluentType");
exports.DialogContentStyles = function (props) {
    var theme = props.theme;
    var palette = theme.palette, effects = theme.effects;
    return {
        title: {
            fontSize: FluentType_1.FontSizes.size20,
            fontWeight: styling_1.FontWeights.semibold,
            padding: '16px 46px 24px 24px',
            lineHeight: 'normal'
        },
        topButton: {
            padding: '14px 14px 0 0',
            selectors: {
                '.ms-Dialog-button': {
                    color: palette.neutralSecondary
                },
                '.ms-Dialog-button:hover': {
                    color: palette.neutralDark,
                    borderRadius: effects.roundedCorner2
                }
            }
        },
        inner: {
            padding: '0 24px 24px 24px'
        },
        subText: {
            fontWeight: styling_1.FontWeights.regular
        }
    };
};
exports.DialogFooterStyles = function (props) {
    return {
        actions: {
            margin: '16px 0 0'
        }
    };
};
