// @pnp/spfx-controls-react/lib/controls/listItemAttachments/ListItemAttachments.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
// Joao Mendes November 2018, SPFx reusable Control ListItemAttachments
var React = require("react");
var Dialog_1 = require("office-ui-fabric-react/lib/Dialog");
var Button_1 = require("office-ui-fabric-react/lib/Button");
var Label_1 = require("office-ui-fabric-react/lib/Label");
var strings = require("ControlStrings");
var ListItemAttachments_module_scss_1 = require("./ListItemAttachments.module.scss");
var UploadAttachment_1 = require("./UploadAttachment");
var DocumentCard_1 = require("office-ui-fabric-react/lib/DocumentCard");
var Image_1 = require("office-ui-fabric-react/lib/Image");
var SPService_1 = require("../../services/SPService");
var Tooltip_1 = require("office-ui-fabric-react/lib/Tooltip");
var Spinner_1 = require("office-ui-fabric-react/lib/Spinner");
var utilities_1 = require("./utilities");
var placeholder_1 = require("../placeholder");
var ListItemAttachments = (function (_super) {
    __extends(ListItemAttachments, _super);
    function ListItemAttachments(props) {
        var _this = _super.call(this, props) || this;
        /**
         * Close the dialog
         */
        _this._closeDialog = function () {
            _this.setState({
                hideDialog: true,
                dialogMessage: '',
                file: null,
                deleteAttachment: false,
            });
            _this.loadAttachments();
        };
        /**
         * Attachment uploaded event handler
         */
        _this._onAttachmentUpload = function () {
            // load Attachments
            _this.loadAttachments();
        };
        /**
         * On delete attachment event handler
         *
         * @param file
         */
        _this.onDeleteAttachment = function (file) {
            _this.setState({
                hideDialog: false,
                deleteAttachment: true,
                file: file,
                dialogMessage: strings.ListItemAttachmentsconfirmDelete.replace('{0}', file.FileName),
            });
        };
        /**
         * Delete the attachment once it was confirmed
         */
        _this.onConfirmedDeleteAttachment = function () { return __awaiter(_this, void 0, void 0, function () {
            var file, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        file = this.state.file;
                        this.setState({
                            disableButton: true,
                        });
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this._spservice.deleteAttachment(file.FileName, this.props.listId, this.props.itemId, this.props.webUrl)];
                    case 2:
                        _a.sent();
                        this.setState({
                            hideDialog: false,
                            deleteAttachment: false,
                            disableButton: false,
                            file: null,
                            dialogMessage: strings.ListItemAttachmentsfileDeletedMsg.replace('{0}', file.FileName),
                        });
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _a.sent();
                        this.setState({
                            hideDialog: false,
                            file: null,
                            deleteAttachment: false,
                            dialogMessage: strings.ListItemAttachmentsfileDeleteError.replace('{0}', file.FileName).replace('{1}', error_1.message)
                        });
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        }); };
        _this.state = {
            file: null,
            hideDialog: true,
            dialogMessage: '',
            attachments: [],
            deleteAttachment: false,
            disableButton: false,
            showPlaceHolder: false,
            fireUpload: false
        };
        // Get SPService Factory
        _this._spservice = new SPService_1.default(_this.props.context);
        _this._utilities = new utilities_1.default();
        return _this;
    }
    /**
     * componentDidMount lifecycle hook
     */
    ListItemAttachments.prototype.componentDidMount = function () {
        this.loadAttachments();
    };
    /**
     * Load Item Attachments
     */
    ListItemAttachments.prototype.loadAttachments = function () {
        return __awaiter(this, void 0, void 0, function () {
            var files, _i, files_1, file, _previewImage, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.previewImages = [];
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 7, , 8]);
                        return [4 /*yield*/, this._spservice.getListItemAttachments(this.props.listId, this.props.itemId)];
                    case 2:
                        files = _a.sent();
                        _i = 0, files_1 = files;
                        _a.label = 3;
                    case 3:
                        if (!(_i < files_1.length)) return [3 /*break*/, 6];
                        file = files_1[_i];
                        return [4 /*yield*/, this._utilities.GetFileImageUrl(file)];
                    case 4:
                        _previewImage = _a.sent();
                        this.previewImages.push({
                            name: file.FileName,
                            previewImageSrc: _previewImage,
                            iconSrc: '',
                            imageFit: Image_1.ImageFit.center,
                            width: 187,
                            height: 130,
                        });
                        _a.label = 5;
                    case 5:
                        _i++;
                        return [3 /*break*/, 3];
                    case 6:
                        this.setState({
                            hideDialog: true,
                            dialogMessage: '',
                            attachments: files,
                            showPlaceHolder: files.length === 0 ? true : false
                        });
                        return [3 /*break*/, 8];
                    case 7:
                        error_2 = _a.sent();
                        this.setState({
                            hideDialog: true,
                            dialogMessage: strings.ListItemAttachmentserrorLoadAttachments.replace('{0}', error_2.message)
                        });
                        return [3 /*break*/, 8];
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Default React render method
     */
    ListItemAttachments.prototype.render = function () {
        var _this = this;
        return (React.createElement("div", { className: ListItemAttachments_module_scss_1.default.ListItemAttachments },
            React.createElement(UploadAttachment_1.UploadAttachment, { listId: this.props.listId, itemId: this.props.itemId, disabled: this.props.disabled, context: this.props.context, onAttachmentUpload: this._onAttachmentUpload, fireUpload: this.state.fireUpload }),
            this.state.showPlaceHolder ?
                React.createElement(placeholder_1.Placeholder, { iconName: 'Upload', iconText: strings.ListItemAttachmentslPlaceHolderIconText, description: strings.ListItemAttachmentslPlaceHolderDescription, buttonLabel: strings.ListItemAttachmentslPlaceHolderButtonLabel, onConfigure: function () { return _this.setState({ fireUpload: true }); } })
                :
                    this.state.attachments.map(function (file, i) {
                        return (React.createElement("div", { className: ListItemAttachments_module_scss_1.default.documentCardWrapper },
                            React.createElement(Tooltip_1.TooltipHost, { content: file.FileName, calloutProps: { gapSpace: 0, isBeakVisible: true }, closeDelay: 200, directionalHint: 12 /* rightCenter */ },
                                React.createElement(DocumentCard_1.DocumentCard, { onClickHref: file.ServerRelativeUrl + "?web=1", className: ListItemAttachments_module_scss_1.default.documentCard },
                                    React.createElement(DocumentCard_1.DocumentCardPreview, { previewImages: [_this.previewImages[i]] }),
                                    React.createElement(Label_1.Label, { className: ListItemAttachments_module_scss_1.default.fileLabel }, file.FileName),
                                    React.createElement(DocumentCard_1.DocumentCardActions, { actions: [
                                            {
                                                iconProps: {
                                                    iconName: 'Delete',
                                                    title: strings.ListItemAttachmentsActionDeleteIconTitle,
                                                },
                                                title: strings.ListItemAttachmentsactionDeleteTitle,
                                                text: strings.ListItemAttachmentsactionDeleteTitle,
                                                disabled: _this.props.disabled,
                                                onClick: function (ev) {
                                                    ev.preventDefault();
                                                    ev.stopPropagation();
                                                    _this.onDeleteAttachment(file);
                                                }
                                            },
                                        ] })))));
                    }),
            React.createElement(Dialog_1.Dialog, { hidden: this.state.hideDialog, type: Dialog_1.DialogType.normal, onDismiss: this._closeDialog, dialogContentProps: {
                    type: Dialog_1.DialogType.normal,
                    title: strings.ListItemAttachmentsdialogTitle,
                    subText: this.state.dialogMessage
                }, modalProps: {
                    isBlocking: true,
                    containerClassName: 'ms-dialogMainOverride'
                } },
                React.createElement(Dialog_1.DialogFooter, null,
                    React.createElement("div", { style: { marginBottom: 7 } }, this.state.disableButton ? React.createElement(Spinner_1.Spinner, { size: Spinner_1.SpinnerSize.medium }) : null),
                    this.state.deleteAttachment ? (React.createElement(Button_1.PrimaryButton, { disabled: this.state.disableButton, onClick: this.onConfirmedDeleteAttachment }, strings.ListItemAttachmentsdialogOKbuttonLabelOnDelete)) : null,
                    this.state.deleteAttachment ? (React.createElement(Button_1.DefaultButton, { disabled: this.state.disableButton, onClick: this._closeDialog }, strings.ListItemAttachmentsdialogCancelButtonLabel))
                        : React.createElement(Button_1.PrimaryButton, { onClick: this._closeDialog }, strings.ListItemAttachmentsdialogOKbuttonLabel)))));
    };
    return ListItemAttachments;
}(React.Component));
exports.ListItemAttachments = ListItemAttachments;
