// office-ui-fabric-react/lib/components/Stack/examples/Stack.Vertical.Grow.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Stack } from '../Stack';
import { DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var VerticalStackGrowExample = /** @class */ (function (_super) {
    tslib_1.__extends(VerticalStackGrowExample, _super);
    function VerticalStackGrowExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    VerticalStackGrowExample.prototype.render = function () {
        var stackStyles = {
            root: {
                background: DefaultPalette.themeTertiary,
                height: 250
            }
        };
        var stackItemStyles = {
            root: {
                alignItems: 'center',
                background: DefaultPalette.themePrimary,
                color: DefaultPalette.white,
                display: 'flex',
                justifyContent: 'center'
            }
        };
        var outerStackTokens = { childrenGap: 5 };
        var innerStackTokens = {
            childrenGap: 5,
            padding: 10
        };
        return (React.createElement(Stack, { tokens: outerStackTokens },
            React.createElement(Stack, { styles: stackStyles, tokens: innerStackTokens },
                React.createElement(Stack.Item, { grow: 3, styles: stackItemStyles }, "Grow is 3"),
                React.createElement(Stack.Item, { grow: 2, styles: stackItemStyles }, "Grow is 2"),
                React.createElement(Stack.Item, { grow: true, styles: stackItemStyles }, "Grow is 1"))));
    };
    return VerticalStackGrowExample;
}(React.Component));
export { VerticalStackGrowExample };
