// office-ui-fabric-react/lib-amd/components/Button/examples/Button.Split.Example.js
define(["require", "exports", "tslib", "react", "office-ui-fabric-react"], function (require, exports, tslib_1, React, office_ui_fabric_react_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var exampleStyles = {
        twoup: [
            'ms-BasicButtonsTwoUp',
            {
                display: 'flex',
                selectors: {
                    '& > *': {
                        flexGrow: 1
                    },
                    '.ms-Label': {
                        marginBottom: '10px'
                    }
                }
            }
        ]
    };
    var customSplitButtonStyles = {
        splitButtonMenuButton: { backgroundColor: 'white', width: '10px' },
        splitButtonMenuIcon: { fontSize: '7px' },
        splitButtonDivider: { borderLeft: '1px solid #c8c8c8', right: 17 },
        splitButtonContainer: {
            selectors: (_a = {},
                _a[office_ui_fabric_react_1.HighContrastSelector] = {
                    border: 'none'
                },
                _a)
        }
    };
    var alertClicked = function () {
        alert('Clicked');
    };
    var getClassNames = office_ui_fabric_react_1.classNamesFunction();
    var classNames = getClassNames(exampleStyles, {});
    var ButtonSplitExample = /** @class */ (function (_super) {
        tslib_1.__extends(ButtonSplitExample, _super);
        function ButtonSplitExample() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        ButtonSplitExample.prototype.render = function () {
            var _a = this.props, disabled = _a.disabled, checked = _a.checked;
            return (React.createElement("div", { className: office_ui_fabric_react_1.css(classNames.twoup) },
                React.createElement("div", null,
                    React.createElement(office_ui_fabric_react_1.Label, null, "Standard"),
                    React.createElement(office_ui_fabric_react_1.DefaultButton, { "data-automation-id": "test", disabled: disabled, checked: checked, text: "Create account", onClick: alertClicked, split: true, splitButtonAriaLabel: 'See 2 sample options', "aria-roledescription": 'split button', style: { height: '35px' }, menuProps: {
                            items: [
                                {
                                    key: 'emailMessage',
                                    text: 'Email message',
                                    iconProps: { iconName: 'Mail' }
                                },
                                {
                                    key: 'calendarEvent',
                                    text: 'Calendar event',
                                    iconProps: { iconName: 'Calendar' }
                                }
                            ]
                        } })),
                React.createElement("div", null,
                    React.createElement(office_ui_fabric_react_1.Label, null, "Primary"),
                    React.createElement(office_ui_fabric_react_1.DefaultButton, { primary: true, "data-automation-id": "test", disabled: disabled, checked: checked, text: "Create account", onClick: alertClicked, split: true, "aria-roledescription": 'split button', style: { height: '35px' }, menuProps: {
                            items: [
                                {
                                    key: 'emailMessage',
                                    text: 'Email message',
                                    iconProps: { iconName: 'Mail' }
                                },
                                {
                                    key: 'calendarEvent',
                                    text: 'Calendar event',
                                    iconProps: { iconName: 'Calendar' }
                                }
                            ]
                        } })),
                React.createElement("div", null,
                    React.createElement(office_ui_fabric_react_1.Label, null, "Primary Action Disabled"),
                    React.createElement(office_ui_fabric_react_1.DefaultButton, { primary: true, "data-automation-id": "test", disabled: disabled, primaryDisabled: true, checked: checked, text: "Create account", onClick: alertClicked, split: true, "aria-roledescription": 'split button', style: { height: '35px' }, menuProps: {
                            items: [
                                {
                                    key: 'emailMessage',
                                    text: 'Email message',
                                    iconProps: { iconName: 'Mail' }
                                },
                                {
                                    key: 'calendarEvent',
                                    text: 'Calendar event',
                                    iconProps: { iconName: 'Calendar' }
                                }
                            ]
                        } })),
                React.createElement("div", null,
                    React.createElement(office_ui_fabric_react_1.Label, null, "Button Disabled"),
                    React.createElement(office_ui_fabric_react_1.DefaultButton, { primary: true, "data-automation-id": "test", disabled: true, allowDisabledFocus: true, checked: checked, text: "Create account", onClick: alertClicked, onKeyPress: alertClicked, onKeyDown: alertClicked, onKeyUp: alertClicked, onMouseDown: alertClicked, onMouseUp: alertClicked, split: true, "aria-roledescription": 'split button', style: { height: '35px' }, menuProps: {
                            items: [
                                {
                                    key: 'emailMessage',
                                    text: 'Email message',
                                    iconProps: { iconName: 'Mail' }
                                },
                                {
                                    key: 'calendarEvent',
                                    text: 'Calendar event',
                                    iconProps: { iconName: 'Calendar' }
                                }
                            ]
                        } }))));
        };
        return ButtonSplitExample;
    }(React.Component));
    exports.ButtonSplitExample = ButtonSplitExample;
    var ButtonSplitCustomExample = /** @class */ (function (_super) {
        tslib_1.__extends(ButtonSplitCustomExample, _super);
        function ButtonSplitCustomExample() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        ButtonSplitCustomExample.prototype.render = function () {
            var _a = this.props, disabled = _a.disabled, checked = _a.checked;
            return (React.createElement("div", null,
                React.createElement(office_ui_fabric_react_1.Label, null, "Split button with icon and custom styles"),
                React.createElement(office_ui_fabric_react_1.IconButton, { "data-automation-id": "test", disabled: disabled, checked: checked, iconProps: { iconName: 'Upload' }, text: "Create account", onClick: alertClicked, split: true, "aria-roledescription": 'split button', styles: customSplitButtonStyles, menuProps: {
                        items: [
                            {
                                key: 'emailMessage',
                                text: 'Email message',
                                iconProps: { iconName: 'Mail' }
                            },
                            {
                                key: 'calendarEvent',
                                text: 'Calendar event',
                                iconProps: { iconName: 'Calendar' }
                            }
                        ]
                    } })));
        };
        return ButtonSplitCustomExample;
    }(React.Component));
    exports.ButtonSplitCustomExample = ButtonSplitCustomExample;
    var _a;
});
