// office-ui-fabric-react/lib-amd/components/Label/Label.classNames.js
define(["require", "exports", "../../Utilities", "../../Styling"], function (require, exports, Utilities_1, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getLabelClassNames = Utilities_1.memoizeFunction(function (theme, className, disabled, required) {
        return Styling_1.mergeStyleSets({
            root: ['ms-Label',
                {
                    color: theme.semanticColors.bodyText,
                    boxSizing: 'border-box',
                    boxShadow: 'none',
                    margin: 0,
                    display: 'block',
                    padding: '5px 0',
                    wordWrap: 'break-word',
                    overflowWrap: 'break-word',
                },
                disabled && {
                    color: theme.semanticColors.disabledBodyText,
                    selectors: (_a = {},
                        _a[Styling_1.HighContrastSelector] = {
                            color: 'GrayText'
                        },
                        _a)
                },
                required && {
                    selectors: {
                        '::after': {
                            content: "' *'",
                            color: theme.semanticColors.errorText,
                            paddingRight: 12
                        }
                    }
                },
                className]
        });
        var _a;
    });
});
