// @pnp/spfx-controls-react/lib/controls/richText/RichText.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./RichText.module.css');
var styles = {
    richtext: 'richtext_c360855e',
    insertLinkDialog: 'insertLinkDialog_c360855e',
    actions: 'actions_c360855e',
    actionsRight: 'actionsRight_c360855e',
    action: 'action_c360855e',
    unlinkButton: 'unlinkButton_c360855e',
    headerDropDown: 'headerDropDown_c360855e',
    toolbarDropDown: 'toolbarDropDown_c360855e',
    toolbarDropDownTitleIcon: 'toolbarDropDownTitleIcon_c360855e',
    toolbarDropDownIcon: 'toolbarDropDownIcon_c360855e',
    toolbarDropDownOption: 'toolbarDropDownOption_c360855e',
    toolbarButtonH2: 'toolbarButtonH2_c360855e',
    toolbarButtonH3: 'toolbarButtonH3_c360855e',
    toolbarButtonH4: 'toolbarButtonH4_c360855e',
    toolbarButtonBlockQuote: 'toolbarButtonBlockQuote_c360855e',
    toolbarSubmenuCaret: 'toolbarSubmenuCaret_c360855e',
    toolbarSubmenuDisplayButton: 'toolbarSubmenuDisplayButton_c360855e',
};
exports.default = styles;
/* tslint:enable */
