// @pnp/spfx-controls-react/lib/controls/richText/RichTextPropertyPane.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var strings = require("ControlStrings");
var RichTextPropertyPane_module_scss_1 = require("./RichTextPropertyPane.module.scss");
var RteColorPicker_1 = require("./RteColorPicker");
var Button_1 = require("office-ui-fabric-react/lib/Button");
var Panel_1 = require("office-ui-fabric-react/lib/Panel");
var Tooltip_1 = require("office-ui-fabric-react/lib/Tooltip");
var Dropdown_1 = require("office-ui-fabric-react/lib/Dropdown");
var ThemeColorHelper_1 = require("../../common/utilities/ThemeColorHelper");
var RichTextPropertyPane = (function (_super) {
    __extends(RichTextPropertyPane, _super);
    function RichTextPropertyPane(props) {
        var _this = _super.call(this, props) || this;
        /**
         * On selection changed event handler
         */
        _this.onChangeSelection = function (range, oldRange, source) {
            var quill = _this.props.editor;
            if (quill === undefined || range === undefined) {
                return;
            }
            if (range) {
                var formats = quill.getFormat(range);
                _this.setState({
                    formats: formats
                });
            }
        };
        /**
         * Render the actions group
         */
        _this.renderActionsGroup = function () {
            return (React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupField },
                React.createElement("div", { className: "ms-CustomFieldHost" },
                    React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.controlsInOneRow },
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.UndoTitle, id: "undo-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { onClick: _this.handleUndo, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "undo-propertyPaneButton", iconProps: {
                                    iconName: 'Undo',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.RedoTitle, id: "redo-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { onClick: _this.handleRedo, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "redo-propertyPaneButton", iconProps: {
                                    iconName: 'Redo',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.ClearFormattingTitle, id: "clearFormatting-button-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { onClick: _this.handleClearFormatting, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "clearFormatting-button-propertyPaneButton", iconProps: {
                                    iconName: 'ClearFormatting',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } }))))));
        };
        /**
         * Render font styles group
         */
        _this.renderFontStylesGroup = function () {
            var selectedHeader = _this.state.formats.header ? _this.state.formats.header : 0;
            return (React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupField },
                React.createElement(Dropdown_1.Dropdown, { label: strings.FontStyleTitle, ariaLabel: strings.FontStyleTitle, defaultSelectedKey: 0, selectedKey: selectedHeader, options: [
                        { key: 0, text: strings.HeaderNormalText },
                        { key: 2, text: strings.HeaderH2 },
                        { key: 3, text: strings.HeaderH3 },
                        { key: 4, text: strings.HeaderH4 },
                        { key: 7, text: strings.HeaderBlockQuote }
                    ], onChanged: _this.onChangeHeading })));
        };
        /**
         * Render font size group
         */
        _this.renderFontSizesGroup = function () {
            // get the selected header
            var selectedSize = _this.state.formats.size ? _this.state.formats.size : 'large';
            return (React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupField },
                React.createElement(Dropdown_1.Dropdown, { label: strings.FontSizeTitle, ariaLabel: strings.FontSizeTitle, defaultSelectedKey: 'large', selectedKey: selectedSize, options: [
                        { key: 'small', text: '12' },
                        { key: 'medium', text: '14' },
                        { key: 'mediumplus', text: '15' },
                        { key: 'large', text: '17' },
                        { key: 'xlarge', text: '21' },
                        { key: 'xlargeplus', text: '24' },
                        { key: 'xxlarge', text: '28' },
                        { key: 'xxxlarge', text: '32' },
                        { key: 'xxlargeplus', text: '36' },
                        { key: 'super', text: '42' },
                    ], onChanged: _this.onChangeSize })));
        };
        /**
         * Render inline styles group
         */
        _this.renderInlineStylesGroup = function () {
            return (React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupField },
                React.createElement("div", { className: "ms-CustomFieldHost" },
                    React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.controlsInOneRow },
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.BoldTitle, id: "bold-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.bold, onClick: function () { return _this.applyFormat('bold', !_this.state.formats.bold); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "bold-propertyPaneButton", iconProps: {
                                    iconName: 'Bold',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.ItalicTitle, id: "italic-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.italic, onClick: function () { return _this.applyFormat('italic', !_this.state.formats.italic); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "italic-propertyPaneButton", iconProps: {
                                    iconName: 'Italic',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.UnderlineTitle, id: "underline-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.underline, onClick: function () { return _this.applyFormat('underline', !_this.state.formats.underline); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "underline-propertyPaneButton", iconProps: {
                                    iconName: 'Underline',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.StrikethroughTitle, id: "strikethrough-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.strike, onClick: function () { return _this.applyFormat('strike', !_this.state.formats.strike); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "strikethrough-propertyPaneButton", iconProps: {
                                    iconName: 'Strikethrough',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.SuperscriptTitle, id: "superscript-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.script === 'super', onClick: function () { return _this.applyFormat('script', _this.state.formats.script === 'super' ? '' : 'super'); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "superscript-propertyPaneButton", iconProps: {
                                    iconName: 'Superscript',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.SubscriptTitle, id: "subscript-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.script === 'sub', onClick: function () { return _this.applyFormat('script', _this.state.formats.script === 'sub' ? '' : 'sub'); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "subscript-propertyPaneButton", iconProps: {
                                    iconName: 'Subscript',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } }))))));
        };
        /**
         * Render color styles group
         */
        _this.renderColorStylesGroup = function () {
            var color = _this.state.formats.color || ThemeColorHelper_1.ThemeColorHelper.GetThemeColor(RichTextPropertyPane_module_scss_1.default.NeutralPrimary);
            var backgroundColor = _this.state.formats.background || "rgba(0, 0, 0, 0)";
            return (React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupField },
                React.createElement("div", { className: "ms-CustomFieldHost" },
                    React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.controlsInOneRow },
                        React.createElement(RteColorPicker_1.default, { colorPickerGroups: [
                                "themeColors",
                                "standardColors"
                            ], buttonLabel: strings.FontColorLabel, id: "fontColor-propertyPaneButton", defaultButtonLabel: strings.AutomaticFontColor, fillThemeColor: true, previewColor: color, selectedColor: color, onColorChanged: _this.handleFillColorChanged, switchToDefaultColor: function () { return _this.handleFillColorChanged(undefined); } }),
                        React.createElement(RteColorPicker_1.default, { buttonLabel: strings.HighlightColorLabel, colorPickerGroups: [
                                "highlightColors"
                            ], fillThemeColor: false, onColorChanged: _this.handleHighlightColorChanged, switchToDefaultColor: function () { return _this.handleHighlightColorChanged(undefined); }, previewColor: backgroundColor, defaultButtonLabel: strings.NoColorHighlightColor, selectedColor: backgroundColor, id: "highlightColor-propertyPaneButton" })))));
        };
        /**
         * Render alignment style groups
         */
        _this.renderAlignmentStylesGroup = function () {
            return (React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupField },
                React.createElement("div", { className: "ms-CustomFieldHost" },
                    React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.controlsInOneRow },
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.AlignLeft, id: "left-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.align === undefined, onClick: function () { return _this.applyFormat('align', 'left'); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "left-propertyPaneButton", iconProps: {
                                    iconName: 'AlignLeft',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.AlignCenter, id: "center-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.align === 'center', onClick: function () { return _this.applyFormat('align', 'center'); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "center-propertyPaneButton", iconProps: {
                                    iconName: 'AlignCenter',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.AlignRight, id: "right-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.align === 'right', onClick: function () { return _this.applyFormat('align', 'right'); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "right-propertyPaneButton", iconProps: {
                                    iconName: 'AlignRight',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.AlignJustify, id: "justify-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.align === 'justify', onClick: function () { return _this.applyFormat('align', 'justify'); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "justify-propertyPaneButton", iconProps: {
                                    iconName: 'AlignJustify',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.IncreaseIndentTitle, id: "increaseindent-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.script === 'super', onClick: function () { return _this.applyFormat('script', _this.state.formats.script === 'super' ? '' : 'super'); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "increaseindent-propertyPaneButton", iconProps: {
                                    iconName: 'IncreaseIndentLegacy',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.DecreaseIndentTitle, id: "decreaseindent-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.script === 'sub', onClick: function () { return _this.applyFormat('script', _this.state.formats.script === 'sub' ? '' : 'sub'); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "decreaseindent-propertyPaneButton", iconProps: {
                                    iconName: 'DecreaseIndentLegacy',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } }))))));
        };
        /**
         * Render list styles group
         */
        _this.renderListStylesGroup = function () {
            return React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupField },
                React.createElement("div", { className: "ms-CustomFieldHost" },
                    React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.controlsInOneRow },
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.ListBullet, id: "bullet-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.list === 'bullet', onClick: function () { return _this.applyFormat('list', 'bullet'); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "bullet-propertyPaneButton", iconProps: {
                                    iconName: 'BulletedList',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.ListNumbered, id: "ordered-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.list === 'ordered', onClick: function () { return _this.applyFormat('list', 'ordered'); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "ordered-propertyPaneButton", iconProps: {
                                    iconName: 'NumberedList',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })))));
        };
        /**
         * Render hyperlink styles group
         */
        _this.renderHyperlinkStylesGroup = function () {
            return (React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupField },
                React.createElement("div", { className: "ms-CustomFieldHost" },
                    React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.controlsInOneRow },
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.LinkTitle, id: "link-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { checked: _this.state.formats.link !== undefined, onClick: function () { return _this.props.onLink(); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "link-propertyPaneButton", iconProps: {
                                    iconName: 'Link',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } })),
                        React.createElement(Tooltip_1.TooltipHost, { content: strings.RemoveLinkLabel, id: "unlink-propertyPaneButton", calloutProps: { gapSpace: 0 } },
                            React.createElement(Button_1.IconButton, { disabled: _this.state.formats.link === undefined, onClick: function () { return _this.applyFormat('link', false); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneButton, "aria-describedby": "unlink-propertyPaneButton", iconProps: {
                                    iconName: 'RemoveLink',
                                    style: {
                                        fontSize: '20px'
                                    }
                                } }))))));
        };
        /**
         * Handle fill color change
         */
        _this.handleFillColorChanged = function (color) {
            _this.applyFormat('color', color);
        };
        /**
         * Handle the hightlight color change
         */
        _this.handleHighlightColorChanged = function (color) {
            _this.applyFormat('background', color);
        };
        /**
         * On heading change
         */
        _this.onChangeHeading = function (item) {
            var newHeadingValue = item.key === 0 ? '' : item.key.toString();
            _this.applyFormat("header", newHeadingValue);
        };
        /**
         * On size change
         */
        _this.onChangeSize = function (item) {
            var newSizeValue = item.key === 0 ? '' : item.key.toString();
            _this.applyFormat("size", newSizeValue);
        };
        /**
         * Handle the undo action
         */
        _this.handleUndo = function () {
            var quill = _this.props.editor;
            quill.getModule("history").undo();
            setTimeout(function () {
                _this.onChangeSelection(quill.getSelection());
            }, 100);
        };
        /**
         * Handle the clear formatting action
         */
        _this.handleClearFormatting = function () {
            var quill = _this.props.editor;
            var range = quill.getSelection();
            if (range) {
                if (range.length > 0) {
                    quill.removeFormat(range.index, range.length);
                    setTimeout(function () {
                        _this.onChangeSelection(quill.getSelection());
                    }, 100);
                }
            }
        };
        /**
         * Handle the redo action
         */
        _this.handleRedo = function () {
            var quill = _this.props.editor;
            quill.getModule("history").redo();
            setTimeout(function () {
                _this.onChangeSelection(quill.getSelection());
            }, 100);
        };
        /**
         * Navigation render
         */
        _this.handleRenderNavigation = function () {
            return (React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.formattingPaneTitle, "aria-hidden": "true" },
                strings.FormattingPaneTitle,
                React.createElement(Button_1.IconButton, { onClick: function () { return _this.props.onClose(); }, className: RichTextPropertyPane_module_scss_1.default.propertyPaneClose, iconProps: {
                        iconName: 'Cancel'
                    }, title: strings.CloseButton, ariaLabel: strings.CloseButton })));
        };
        _this.state = {
            formats: {}
        };
        return _this;
    }
    /**
     * componentDidUpdate lifecycle hook
     *
     * @param prevProps
     * @param prevState
     */
    RichTextPropertyPane.prototype.componentDidUpdate = function (prevProps, prevState) {
        // if we're just opening, sync the format choices with the current selection
        if (!prevProps.isOpen && this.props.isOpen) {
            var quill = this.props.editor;
            if (quill === undefined) {
                return;
            }
            var range = quill.getSelection();
            this.onChangeSelection(range, undefined, undefined);
        }
    };
    /**
     * Default React render method
     */
    RichTextPropertyPane.prototype.render = function () {
        return (React.createElement(Panel_1.Panel, { className: RichTextPropertyPane_module_scss_1.default.richTextPropertyPane, isBlocking: false, isOpen: this.props.isOpen, type: Panel_1.PanelType.smallFixedFar, onDismiss: this.props.onClose, closeButtonAriaLabel: strings.CloseButton, onRenderNavigation: this.handleRenderNavigation, focusTrapZoneProps: {
                forceFocusInsideTrap: false,
                isClickableOutsideFocusTrap: true
            } },
            React.createElement("div", null,
                React.createElement("div", null,
                    React.createElement("div", null,
                        React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupContent },
                            this.renderActionsGroup(),
                            this.renderFontStylesGroup(),
                            this.renderFontSizesGroup(),
                            this.renderInlineStylesGroup(),
                            this.renderColorStylesGroup())),
                    React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupContent },
                        React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupHeaderNoAccordion }, strings.ParagraphSectionTitle),
                        this.renderAlignmentStylesGroup(),
                        this.renderListStylesGroup()),
                    React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupContent },
                        React.createElement("div", { className: RichTextPropertyPane_module_scss_1.default.propertyPaneGroupHeaderNoAccordion }, strings.HyperlinkSectionTitle),
                        this.renderHyperlinkStylesGroup())))));
    };
    /**
     * Apply the new format
     *
     * @param name
     * @param value
     */
    RichTextPropertyPane.prototype.applyFormat = function (name, value) {
        var _this = this;
        var quill = this.props.editor;
        quill.format(name, value);
        setTimeout(function () {
            _this.onChangeSelection(quill.getSelection());
        }, 100);
    };
    return RichTextPropertyPane;
}(React.Component));
exports.default = RichTextPropertyPane;
