// @pnp/spfx-controls-react/lib/controls/richText/RichText.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var strings = require("ControlStrings");
require("react-quill/dist/quill.snow.css");
var RichTextPropertyPane_1 = require("./RichTextPropertyPane");
var react_quill_1 = require("react-quill");
var RichText_module_scss_1 = require("./RichText.module.scss");
var Button_1 = require("office-ui-fabric-react/lib/Button");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var Tooltip_1 = require("office-ui-fabric-react/lib/Tooltip");
var Dialog_1 = require("office-ui-fabric-react/lib/Dialog");
var TextField_1 = require("office-ui-fabric-react/lib/TextField");
var Link_1 = require("office-ui-fabric-react/lib/Link");
var Button_2 = require("office-ui-fabric-react/lib/Button");
var Dropdown_1 = require("office-ui-fabric-react/lib/Dropdown");
var Icon_1 = require("office-ui-fabric-react/lib/Icon");
var Utilities_1 = require("office-ui-fabric-react/lib/Utilities");
var telemetry = require("../../common/telemetry");
var Icons_1 = require("office-ui-fabric-react/lib/Icons");
var TOOLBARPADDING = 28;
/**
 * Creates a rich text editing control that mimics the out-of-the-box
 * SharePoint Rich Text control.
 * NOTE:
 * Quill.js has a few quirks that we can't work around
 * - Block quotes only work on single lines. This is a frequently-requested feature with Quill that isn't available yet.
 * - Tables aren't supported yet. I'll gladly add table formatting support if users request it.
 */
Icons_1.initializeIcons();
var RichText = (function (_super) {
    __extends(RichText, _super);
    function RichText(props) {
        var _this = _super.call(this, props) || this;
        _this._quillElem = undefined;
        _this._wrapperRef = undefined;
        _this._propertyPaneRef = undefined;
        _this._toolbarId = undefined;
        _this.ddStyleOpts = [{
                key: 0,
                text: strings.HeaderNormalText,
                data: {}
            }, {
                key: 2,
                text: strings.HeaderH2,
                data: { className: RichText_module_scss_1.default.toolbarButtonH2 }
            }, {
                key: 3,
                text: strings.HeaderH3,
                data: { className: RichText_module_scss_1.default.toolbarButtonH3 }
            }, {
                key: 4,
                text: strings.HeaderH4,
                data: { className: RichText_module_scss_1.default.toolbarButtonH4 }
            }, {
                key: 7,
                text: strings.HeaderBlockQuote,
                data: { className: RichText_module_scss_1.default.toolbarButtonBlockQuote }
            }];
        _this.ddAlignOpts = [{
                key: 'left',
                text: strings.AlignLeft,
                data: { icon: 'AlignLeft' }
            }, {
                key: 'center',
                text: strings.AlignCenter,
                data: { icon: 'AlignCenter' }
            }, {
                key: 'right',
                text: strings.AlignRight,
                data: { icon: 'AlignRight' }
            }];
        _this.ddListOpts = [{
                key: 'bullet',
                text: strings.ListBullet,
                data: { icon: 'BulletedList' }
            }, {
                key: 'ordered',
                text: strings.ListNumbered,
                data: { icon: 'NumberedList' }
            }];
        /**
         * Returns a handle to the Quill editor
         */
        _this.getEditor = function () {
            try {
                return _this._quillElem.getEditor();
            }
            catch (error) {
                return undefined;
            }
        };
        /**
         * Renders the "Insert Link" dialog
         */
        _this.renderLinkDialog = function () {
            return (React.createElement(Dialog_1.Dialog, { hidden: _this.state.hideDialog, onDismiss: _this.closeDialog, dialogContentProps: {
                    type: Dialog_1.DialogType.normal,
                    title: strings.InsertLinkTitle,
                }, modalProps: {
                    className: RichText_module_scss_1.default.insertLinkDialog,
                    isBlocking: true,
                    containerClassName: 'ms-dialogMainOverride'
                } },
                React.createElement(TextField_1.TextField, { label: strings.AddressFieldLabel, value: _this.state.insertUrl !== undefined ? _this.state.insertUrl : "https://", onChanged: function (newValue) {
                        _this.setState({
                            insertUrl: newValue
                        });
                    } }),
                React.createElement(TextField_1.TextField, { label: strings.TextToDisplayLabel, value: _this.state.insertUrlText || _this.state.insertUrl, onChanged: function (newValue) {
                        if (newValue !== _this.state.insertUrl) {
                            _this.setState({
                                insertUrlText: newValue
                            });
                        }
                    } }),
                React.createElement(Dialog_1.DialogFooter, { className: RichText_module_scss_1.default.actions },
                    React.createElement("div", { className: "ms-Dialog-actionsRight " + RichText_module_scss_1.default.actionsRight },
                        _this.state.selectedUrl && (React.createElement(Link_1.Link, { className: RichText_module_scss_1.default.action + " " + RichText_module_scss_1.default.unlinkButton, onClick: _this.handleRemoveLink }, strings.RemoveLinkLabel)),
                        React.createElement(Button_2.PrimaryButton, { className: RichText_module_scss_1.default.action, onClick: _this.handleCreateLink, text: strings.SaveButtonLabel }),
                        React.createElement(Button_2.DefaultButton, { className: RichText_module_scss_1.default.action, onClick: _this.closeDialog, text: strings.CancelButtonLabel })))));
        };
        /**
         * Style trigger events
         */
        _this.onChangeBold = function () {
            var newBoldValue = !_this.state.formats.bold;
            _this.applyFormat("bold", newBoldValue);
        };
        _this.onChangeItalic = function () {
            var newValue = !_this.state.formats.italic;
            _this.applyFormat("italic", newValue);
        };
        _this.onChangeUnderline = function () {
            var newValue = !_this.state.formats.underline;
            _this.applyFormat("underline", newValue);
        };
        _this.onChangeHeading = function (item) {
            var newHeadingValue = item.key === 0 ? '' : item.key.toString();
            _this.applyFormat("header", newHeadingValue);
        };
        _this.onChangeAlign = function (item) {
            var newAlignValue = item.key === 'left' ? false : item.key.toString();
            _this.applyFormat("align", newAlignValue);
        };
        _this.onChangeList = function (item) {
            // if we're already in list mode, toggle off
            var key = item.key;
            var newAlignValue = (key === 'bullet' && _this.state.formats.list === 'bullet') || (key === 'numbered' && _this.state.formats.list === 'numbered') ? false : key;
            _this.applyFormat("list", newAlignValue);
        };
        /**
         * Displays the insert link dialog
         */
        _this.showInsertLinkDialog = function () {
            var quill = _this.getEditor();
            var range = quill.getSelection();
            var linkText = _this.state.selectedText;
            if (_this.state.selectedUrl !== undefined && _this.state.selectedText === "") {
                var text = _this.state.text;
                var urlStartIndex = text.indexOf(_this.state.selectedUrl);
                var startTextIndex = text.indexOf(">", urlStartIndex) + 1;
                var endTextIndex = text.indexOf("<", startTextIndex);
                var realLength = endTextIndex - startTextIndex;
                linkText = text.substr(startTextIndex, realLength);
                //Find where the link text starts and select that
                var editorText = quill.getText();
                var linkStart = editorText.indexOf(linkText);
                range.index = linkStart;
                range.length = linkText.length;
            }
            _this.setState({
                hideDialog: false,
                insertUrlText: linkText,
                insertUrl: _this.state.selectedUrl,
                selectedRange: range
            });
        };
        /**
         * Hides the insert link dialog
         */
        _this.closeDialog = function () {
            _this.setState({ hideDialog: true });
        };
        /**
         * When user enters the richtext editor, displays the border
         */
        _this.handleOnFocus = function (range, source, editor) {
            if (!_this.state.editing) {
                _this.setState({ editing: true });
            }
        };
        /**
         * Called when user removes the link
         */
        _this.handleRemoveLink = function () {
            var quill = _this.getEditor();
            quill.format('link', false);
            _this.closeDialog();
        };
        /**
         * Called when user creates a new link
         */
        _this.handleCreateLink = function () {
            var quill = _this.getEditor();
            var range = _this.state.selectedRange;
            var cursorPosition = range.index;
            if (range.length > 0) {
                quill.deleteText(range.index, range.length);
            }
            if (cursorPosition) {
                var textToInsert = _this.state.insertUrlText !== undefined ? _this.state.insertUrlText : _this.state.insertUrl;
                var urlToInsert = _this.state.insertUrl;
                quill.insertText(cursorPosition, textToInsert);
                quill.setSelection(cursorPosition, textToInsert.length);
                quill.formatText(cursorPosition, textToInsert.length, 'link', urlToInsert);
            }
            _this.setState({
                hideDialog: true,
                insertUrl: undefined,
                insertUrlText: undefined
            });
        };
        /**
         * Called when richtext selection changes
         */
        _this.handleChangeSelection = function (range, oldRange, source) {
            var quill = _this.getEditor();
            try {
                if (quill) {
                    // Get the selected text
                    var selectedText = quill.getText(range);
                    // Get the current format
                    var formats = quill.getFormat(range);
                    // Get the currently selected url
                    var selectedUrl = formats.link ? formats.link : undefined;
                    _this.setState({
                        selectedText: selectedText,
                        selectedUrl: selectedUrl,
                        formats: formats
                    });
                    if (_this._propertyPaneRef && _this.state.morePaneVisible) {
                        _this._propertyPaneRef.onChangeSelection(range, oldRange, source);
                    }
                }
            }
            catch (error) {
            }
        };
        /**
         * Called when user clicks on the close icon
         */
        _this.handleClosePanel = function () {
            _this.closePanel();
        };
        /**
         * Closes the panel
         */
        _this.closePanel = function () {
            _this.setState({ morePaneVisible: false });
        };
        /**
         * Called when user clicks on the more button
         */
        _this.handleShowMore = function () {
            _this.setState({
                morePaneVisible: !_this.state.morePaneVisible
            }, function () {
                _this.getEditor().focus();
            });
        };
        /**
         * Called when user changes the text of the editor
         */
        _this.handleChange = function (value) {
            var onChange = _this.props.onChange;
            // do we need to pass this to a handler?
            if (onChange) {
                // yes, get the changed text from the handler
                var newText = onChange(value);
                _this.setState({ text: newText });
            }
            else {
                // no, write the text to the state
                _this.setState({ text: value });
            }
        };
        /**
         * Keeps track of whether we clicked outside the element
         */
        _this.handleClickOutside = function (event) {
            var outside = !Utilities_1.elementContains(_this._wrapperRef, event.target);
            // Did we click outside?
            if (outside) {
                // If we are currently editing, stop editing
                // -- unless we're using the property pane or the dialog
                if (_this.state.editing) {
                    _this.setState({
                        editing: false
                    });
                }
            }
            else {
                // We clicked inside
                if (!_this.state.editing) {
                    // if we aren't currently editing, start editing
                    _this.setState({ editing: true });
                }
            }
        };
        /**
         * Links to the quill reference
         */
        _this.linkQuill = function (e) {
            _this._quillElem = e;
        };
        /**
         * Links to the property pane element
         */
        _this.linkPropertyPane = function (e) {
            _this._propertyPaneRef = e;
        };
        telemetry.track('ReactRichText', {
            className: !!props.className
        });
        _this.state = {
            text: _this.props.value,
            editing: false,
            morePaneVisible: false,
            hideDialog: true,
            formats: {},
            insertUrl: undefined,
            insertUrlText: undefined,
            selectedText: undefined,
            selectedUrl: undefined,
            wrapperTop: 0
        };
        // Get a unique toolbar id
        _this._toolbarId = "toolbar_" + sp_core_library_1.Guid.newGuid().toString();
        return _this;
    }
    /**
     * Attaches to mouse down events to determine if we clicked outside
     */
    RichText.prototype.componentDidMount = function () {
        // If we're in edit mode, attach the mouse down event
        if (this.props.isEditMode) {
            document.addEventListener('click', this.handleClickOutside);
            document.addEventListener('focus', this.handleClickOutside);
        }
        var clientRect = this._wrapperRef.getBoundingClientRect();
        var parentClientRect = this._wrapperRef.parentElement.getBoundingClientRect();
        var toolbarTop = clientRect.top - parentClientRect.top - TOOLBARPADDING;
        this.setState({
            wrapperTop: toolbarTop
        });
    };
    /**
     * Removes the mouse down event
     */
    RichText.prototype.componentWillUnmount = function () {
        // If we were in edit mode, remove the mouse down handler
        if (this.props.isEditMode) {
            document.removeEventListener('click', this.handleClickOutside);
            document.removeEventListener('focus', this.handleClickOutside);
        }
    };
    /**
     * If we're switching from non-edit mode to edit mode, attach mouse down event
     * If we're going from edit mode to non-edit mode, remove mouse down event
     * @param prevProps
     * @param prevState
     */
    RichText.prototype.componentDidUpdate = function (prevProps, prevState) {
        // If we're going from non-edit to edit mode
        if (this.props.isEditMode && !prevProps.isEditMode) {
            document.addEventListener('click', this.handleClickOutside);
            document.addEventListener('focus', this.handleClickOutside);
        }
        // If we're going from edit mode to non-edit mode
        if (!this.props.isEditMode && prevProps.isEditMode) {
            document.removeEventListener('click', this.handleClickOutside);
            document.removeEventListener('focus', this.handleClickOutside);
        }
    };
    /**
     * Render style option
     *
     * @param option
     */
    RichText.prototype.onRenderStyleOption = function (option) {
        return (React.createElement(Tooltip_1.TooltipHost, { content: option.text, id: option.text + "-toolbarButton", calloutProps: { gapSpace: 0 } },
            React.createElement("div", { className: RichText_module_scss_1.default.toolbarDropDownOption + " " + (option.data.className ? option.data.className : ""), "aria-describedby": option.text + "-toolbarButton" },
                React.createElement("span", null, option.text))));
    };
    /**
     * Render the title of the style dropdown
     *
     * @param options
     */
    RichText.prototype.onRenderStyleTitle = function (options) {
        var option = options[0];
        return (React.createElement(Tooltip_1.TooltipHost, { content: option.text, id: option.text + "-dropDownTitle", calloutProps: { gapSpace: 0 } },
            React.createElement("div", { className: RichText_module_scss_1.default.toolbarSubmenuDisplayButton, "aria-describedby": option.text + "-dropDownTitle" },
                React.createElement("span", null, option.text))));
    };
    /**
     * Render align option
     *
     * @param option
     */
    RichText.prototype.onRenderAlignOption = function (option) {
        return (React.createElement(Tooltip_1.TooltipHost, { content: option.text, id: option.text + "-toolbarButton", calloutProps: { gapSpace: 0 } },
            React.createElement("div", { className: RichText_module_scss_1.default.toolbarDropDownOption + " " + (option.data.className ? option.data.className : ""), "aria-describedby": option.text + "-toolbarButton" },
                React.createElement(Icon_1.Icon, { className: RichText_module_scss_1.default.toolbarDropDownIcon, iconName: option.data.icon, "aria-hidden": "true" }))));
    };
    /**
     * Render the list dropdown title
     *
     * @param options
     */
    RichText.prototype.onRenderListTitle = function (options) {
        var option = options[0];
        return (React.createElement(Tooltip_1.TooltipHost, { content: option.text, id: option.text + "-dropDownTitle", calloutProps: { gapSpace: 0 } },
            React.createElement("div", { className: RichText_module_scss_1.default.toolbarSubmenuDisplayButton, "aria-describedby": option.text + "-dropDownTitle" },
                React.createElement(Icon_1.Icon, { className: RichText_module_scss_1.default.toolbarDropDownTitleIcon, iconName: option.data.icon, "aria-hidden": "true" }))));
    };
    /**
     * Render the title of the align dropdown
     *
     * @param options
     */
    RichText.prototype.onRenderAlignTitle = function (options) {
        var option = options[0];
        return (React.createElement(Tooltip_1.TooltipHost, { content: option.text, id: option.text + "-dropDownTitle", calloutProps: { gapSpace: 0 } },
            React.createElement("div", { className: RichText_module_scss_1.default.toolbarSubmenuDisplayButton, "aria-describedby": option.text + "-dropDownTitle" },
                React.createElement(Icon_1.Icon, { className: RichText_module_scss_1.default.toolbarDropDownTitleIcon, iconName: option.data.icon, "aria-hidden": "true" }))));
    };
    /**
     * Render list dropdown option
     *
     * @param option
     */
    RichText.prototype.onRenderListOption = function (option) {
        return (React.createElement(Tooltip_1.TooltipHost, { content: option.text, id: option.text + "-toolbarButton", calloutProps: { gapSpace: 0 } },
            React.createElement("div", { className: RichText_module_scss_1.default.toolbarDropDownOption + " " + (option.data.className ? option.data.className : ""), "aria-describedby": option.text + "-toolbarButton" },
                React.createElement(Icon_1.Icon, { className: RichText_module_scss_1.default.toolbarDropDownIcon, iconName: option.data.icon, "aria-hidden": "true" }))));
    };
    /**
     * Render the list dropdown placeholder
     */
    RichText.prototype.onRenderListPlaceholder = function () {
        return (React.createElement(Tooltip_1.TooltipHost, { content: "Placeholder", id: "Placeholder-dropDownTitle", calloutProps: { gapSpace: 0 } },
            React.createElement("div", { className: RichText_module_scss_1.default.toolbarSubmenuDisplayButton, "aria-describedby": "Placeholder-dropDownTitle" },
                React.createElement(Icon_1.Icon, { className: RichText_module_scss_1.default.toolbarDropDownTitleIcon, iconName: 'BulletedList', "aria-hidden": "true" }))));
    };
    /**
     * Renders the Rich Text Editor
     */
    RichText.prototype.render = function () {
        var _this = this;
        var text = this.state.text;
        var isEditMode = this.props.isEditMode;
        // If we're not in edit mode, display read-only version of the html
        if (!isEditMode) {
            return (React.createElement("div", { className: "ql-editor " + RichText_module_scss_1.default.richtext + " " + this.props.className, dangerouslySetInnerHTML: { __html: text } }));
        }
        // Okay, we're in edit mode.
        var _a = this.props, placeholder = _a.placeholder, _b = _a.styleOptions, showStyles = _b.showStyles, showBold = _b.showBold, showItalic = _b.showItalic, showUnderline = _b.showUnderline, showAlign = _b.showAlign, showList = _b.showList, showLink = _b.showLink, showMore = _b.showMore;
        // Get a unique id for the toolbar
        var modules = {
            toolbar: {
                container: "#" + this._toolbarId,
                handlers: [
                    "link" // disable the link handler so we can add our own
                ]
            },
            clipboard: {
                matchVisual: false // prevents weird bug that inserts blank lines when loading stored text
            }
        };
        // Remove fonts and set Segoe UI as the main font
        var font = react_quill_1.Quill.import('formats/font');
        font.whitelist = ['Segoe UI'];
        react_quill_1.Quill.register(font, true);
        // Set headers and add blockquote capability
        var header = react_quill_1.Quill.import('formats/header');
        header.tagName = [
            'H1',
            'H2',
            'H3',
            'H4',
            'H5',
            'H6',
            'blockquote'
        ];
        react_quill_1.Quill.register(header, true);
        // Add the SharePoint font sizes
        var SizeClass = react_quill_1.Quill.import('formats/size');
        SizeClass.whitelist = [
            'small',
            'medium',
            'mediumplus',
            'large',
            'xlarge',
            'xlargeplus',
            'xxlarge',
            'xxxlarge',
            'xxlargeplus',
            'super'
        ];
        react_quill_1.Quill.register(SizeClass, true);
        return (React.createElement("div", { ref: function (ref) { return _this._wrapperRef = ref; }, className: (RichText_module_scss_1.default.richtext && this.state.editing ? 'ql-active' : '') + " " + this.props.className },
            React.createElement("div", { id: this._toolbarId, style: { top: this.state.wrapperTop } },
                showStyles && (React.createElement(Dropdown_1.Dropdown, { id: "DropDownStyles", className: RichText_module_scss_1.default.headerDropDown + " " + RichText_module_scss_1.default.toolbarDropDown, onRenderCaretDown: function () { return React.createElement(Icon_1.Icon, { className: RichText_module_scss_1.default.toolbarSubmenuCaret, iconName: "CaretDownSolid8" }); }, selectedKey: this.state.formats.header || 0, options: this.ddStyleOpts, onChanged: this.onChangeHeading, onRenderOption: this.onRenderStyleOption, onRenderTitle: this.onRenderStyleTitle })),
                showBold && (React.createElement(Tooltip_1.TooltipHost, { content: strings.BoldTitle, id: "bold-richtextbutton", calloutProps: { gapSpace: 0 } },
                    React.createElement(Button_1.IconButton, { iconProps: { iconName: 'Bold' }, "aria-describedby": "bold-richtextbutton", checked: this.state.formats.bold, onClick: this.onChangeBold }))),
                showItalic && (React.createElement(Tooltip_1.TooltipHost, { content: strings.ItalicTitle, id: "italic-richtextbutton", calloutProps: { gapSpace: 0 } },
                    React.createElement(Button_1.IconButton, { iconProps: { iconName: 'Italic' }, "aria-describedby": "italic-richtextbutton", checked: this.state.formats.italic, onClick: this.onChangeItalic }))),
                showUnderline && (React.createElement(Tooltip_1.TooltipHost, { content: strings.UnderlineTitle, id: "underline-richtextbutton", calloutProps: { gapSpace: 0 } },
                    React.createElement(Button_1.IconButton, { iconProps: { iconName: 'Underline' }, "aria-describedby": "underline-richtextbutton", checked: this.state.formats.underline, onClick: this.onChangeUnderline }))),
                showAlign && (React.createElement(Dropdown_1.Dropdown, { className: "" + RichText_module_scss_1.default.toolbarDropDown, id: "DropDownAlign", onRenderCaretDown: function () { return React.createElement(Icon_1.Icon, { className: RichText_module_scss_1.default.toolbarSubmenuCaret, iconName: "CaretDownSolid8" }); }, selectedKey: this.state.formats.align || 'left', options: this.ddAlignOpts, onChanged: this.onChangeAlign, onRenderOption: this.onRenderAlignOption, onRenderTitle: this.onRenderAlignTitle })),
                showList && (React.createElement(Dropdown_1.Dropdown, { className: RichText_module_scss_1.default.toolbarDropDown, id: "DropDownLists", onRenderCaretDown: function () { return React.createElement(Icon_1.Icon, { className: RichText_module_scss_1.default.toolbarSubmenuCaret, iconName: "CaretDownSolid8" }); }, selectedKey: this.state.formats.list, options: this.ddListOpts, 
                    // this option is not available yet
                    // notifyOnReselect={true} // allows re-selecting selected item to turn it off
                    onChanged: this.onChangeList, onRenderOption: this.onRenderListOption, onRenderTitle: this.onRenderListTitle, onRenderPlaceHolder: this.onRenderListPlaceholder })),
                showLink && (React.createElement(Tooltip_1.TooltipHost, { content: strings.LinkTitle, id: "link-richtextbutton", calloutProps: { gapSpace: 0 } },
                    React.createElement(Button_1.IconButton, { checked: this.state.formats.link !== undefined, onClick: this.showInsertLinkDialog, "aria-describedby": "link-richtextbutton", iconProps: {
                            iconName: 'Link'
                        } }))),
                showMore && (React.createElement(Tooltip_1.TooltipHost, { content: strings.MoreTitle, id: "more-richtextbutton", calloutProps: { gapSpace: 0 } },
                    React.createElement(Button_1.IconButton, { iconProps: { iconName: 'More' }, "aria-describedby": "more-richtextbutton", onClick: this.handleShowMore })))),
            React.createElement(react_quill_1.default, { ref: this.linkQuill, placeholder: placeholder, modules: modules, value: text || '', onChange: this.handleChange, onChangeSelection: this.handleChangeSelection, onFocus: this.handleOnFocus }),
            React.createElement(RichTextPropertyPane_1.default, { ref: this.linkPropertyPane, editor: this.getEditor(), isOpen: this.state.morePaneVisible, onClose: this.handleClosePanel, onLink: this.showInsertLinkDialog }),
            this.renderLinkDialog()));
    };
    /**
     * Applies a format to the selection
     * @param name format name
     * @param value format value, or false to unset format
     */
    RichText.prototype.applyFormat = function (name, value) {
        var _this = this;
        var quill = this.getEditor();
        quill.format(name, value);
        // We use a timeout to ensure that format has been applied and buttons are updated
        setTimeout(function () {
            _this.handleChangeSelection(quill.getSelection(), undefined, undefined);
        }, 100);
    };
    /**
     * Sets default properties
     */
    RichText.defaultProps = {
        isEditMode: true,
        styleOptions: {
            showStyles: true,
            showBold: true,
            showItalic: true,
            showUnderline: true,
            showAlign: true,
            showList: true,
            showLink: true,
            showMore: true
        }
    };
    return RichText;
}(React.Component));
exports.RichText = RichText;
