// @pnp/spfx-controls-react/lib/controls/richText/RteColorPicker.module.scss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable */
require('./RteColorPicker.module.css');
var styles = {
    ThemeColorDarker: '#004578',
    ThemeColorDark: '#005a9e',
    ThemeColorDarkAlt: '#106ebe',
    ThemeColorPrimary: '#0078d4',
    ThemeColorSecondary: '#2b88d8',
    ThemeColorTertiary: '#71afe5',
    ThemeColorNeutralSecondary: '#666666',
    ThemeColorNeutralPrimaryAlt: '#3C3C3C',
    ThemeColorNeutralPrimary: '#333333',
    ThemeColorNeutralDark: '#212121',
    colorPickerButton: 'colorPickerButton_3c377881',
    previewSvg: 'previewSvg_3c377881',
    border: 'border_3c377881',
    previewRectangle: 'previewRectangle_3c377881',
    buttonLabel: 'buttonLabel_3c377881',
    previewIcon: 'previewIcon_3c377881',
    pickerCallout: 'pickerCallout_3c377881',
    defaultSvg: 'defaultSvg_3c377881',
    fillThemeColor: 'fillThemeColor_3c377881',
    fillDefaultColor: 'fillDefaultColor_3c377881',
    pickerLabel: 'pickerLabel_3c377881',
    focusedContainer: 'focusedContainer_3c377881',
    tableRoot: 'tableRoot_3c377881',
    tableCell: 'tableCell_3c377881',
    colorCell: 'colorCell_3c377881',
    svg: 'svg_3c377881',
    selected: 'selected_3c377881',
};
exports.default = styles;
/* tslint:enable */
