// @uifabric/fluent-theme/lib-amd/fluent/FluentStyles.js
define(["require", "exports", "./styles/BasePicker.styles", "./styles/Breadcrumb.styles", "./styles/DetailsList.styles", "./styles/Callout.styles", "./styles/Checkbox.styles", "./styles/ChoiceGroupOption.styles", "./styles/ColorPickerGridCell.styles", "./styles/ColorPicker.styles", "./styles/ComboBox.styles", "./styles/CommandBar.styles", "./styles/CommandBarButton.styles", "./styles/CompoundButton.styles", "./styles/ContextualMenu.styles", "./styles/DatePicker.styles", "./styles/DefaultButton.styles", "./styles/DetailsColumn.styles", "./styles/Dialog.styles", "./styles/Dropdown.styles", "./styles/HoverCard.styles", "./styles/Facepile.styles", "./styles/IconButton.styles", "./styles/Label.styles", "./styles/Link.styles", "./styles/Modal.styles", "./styles/Panel.styles", "./styles/PeoplePicker.styles", "./styles/Persona.styles", "./styles/Pivot.styles", "./styles/PrimaryButton.styles", "./styles/Rating.styles", "./styles/SearchBox.styles", "./styles/Slider.styles", "./styles/SpinButton.styles", "./styles/PickerSuggestions.styles", "./styles/TagPicker.styles", "./styles/TeachingBubble.styles", "./styles/TextField.styles", "./styles/Toggle.styles"], function (require, exports, BasePicker_styles_1, Breadcrumb_styles_1, DetailsList_styles_1, Callout_styles_1, Checkbox_styles_1, ChoiceGroupOption_styles_1, ColorPickerGridCell_styles_1, ColorPicker_styles_1, ComboBox_styles_1, CommandBar_styles_1, CommandBarButton_styles_1, CompoundButton_styles_1, ContextualMenu_styles_1, DatePicker_styles_1, DefaultButton_styles_1, DetailsColumn_styles_1, Dialog_styles_1, Dropdown_styles_1, HoverCard_styles_1, Facepile_styles_1, IconButton_styles_1, Label_styles_1, Link_styles_1, Modal_styles_1, Panel_styles_1, PeoplePicker_styles_1, Persona_styles_1, Pivot_styles_1, PrimaryButton_styles_1, Rating_styles_1, SearchBox_styles_1, Slider_styles_1, SpinButton_styles_1, PickerSuggestions_styles_1, TagPicker_styles_1, TeachingBubble_styles_1, TextField_styles_1, Toggle_styles_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    // Roll up all style overrides in a single "Fluent theme" object
    // TODO: "any" is used here to get around "is using xxx but cannot be named" TS error. Should be able to remove
    //        this 'any' once we upgrade to TS3.1+
    // tslint:disable-next-line:no-any
    exports.FluentStyles = {
        Breadcrumb: {
            styles: Breadcrumb_styles_1.BreadcrumbStyles
        },
        CalloutContent: {
            styles: Callout_styles_1.CalloutContentStyles
        },
        ColorPicker: {
            styles: ColorPicker_styles_1.ColorPickerStyles
        },
        ColorRectangle: {
            styles: ColorPicker_styles_1.ColorRectangleStyles
        },
        ColorSlider: {
            styles: ColorPicker_styles_1.ColorSliderStyles
        },
        CommandBar: {
            styles: CommandBar_styles_1.CommandBarStyles
        },
        CommandBarButton: {
            styles: CommandBarButton_styles_1.CommandBarButtonStyles
        },
        CompoundButton: {
            styles: CompoundButton_styles_1.CompoundButtonStyles
        },
        Check: {
            styles: DetailsList_styles_1.CheckStyles
        },
        Checkbox: {
            styles: Checkbox_styles_1.CheckboxStyles
        },
        ChoiceGroupOption: {
            styles: ChoiceGroupOption_styles_1.ChoiceGroupOptionStyles
        },
        ColorPickerGridCell: {
            styles: ColorPickerGridCell_styles_1.ColorPickerGridCellStyles
        },
        ComboBox: {
            styles: ComboBox_styles_1.ComboBoxStyles
        },
        CompactPeoplePicker: {
            styles: BasePicker_styles_1.BasePickerStyles
        },
        ContextualMenu: {
            styles: ContextualMenu_styles_1.ContextualMenuStyles
        },
        DatePicker: {
            styles: DatePicker_styles_1.DatePickerStyles
        },
        DefaultButton: {
            styles: DefaultButton_styles_1.DefaultButtonStyles
        },
        DetailsColumn: {
            styles: DetailsColumn_styles_1.DetailsColumnStyles
        },
        DetailsRow: {
            styles: DetailsList_styles_1.DetailsRowStyles
        },
        DialogContent: {
            styles: Dialog_styles_1.DialogContentStyles
        },
        DialogFooter: {
            styles: Dialog_styles_1.DialogFooterStyles
        },
        Dropdown: {
            styles: Dropdown_styles_1.DropdownStyles
        },
        ExpandingCard: {
            styles: HoverCard_styles_1.ExpandingCardStyles
        },
        Facepile: {
            styles: Facepile_styles_1.FacepileStyles
        },
        IconButton: {
            styles: IconButton_styles_1.IconButtonStyles
        },
        Label: {
            styles: Label_styles_1.LabelStyles
        },
        Link: {
            styles: Link_styles_1.LinkStyles
        },
        ListPeoplePickerBase: {
            styles: BasePicker_styles_1.BasePickerStyles
        },
        Modal: {
            styles: Modal_styles_1.ModalStyles
        },
        Panel: {
            styles: Panel_styles_1.PanelStyles
        },
        NormalPeoplePicker: {
            styles: BasePicker_styles_1.BasePickerStyles
        },
        PeoplePickerItem: {
            styles: PeoplePicker_styles_1.PeoplePickerItemStyles
        },
        Persona: {
            styles: Persona_styles_1.PersonaStyles
        },
        Pivot: {
            styles: Pivot_styles_1.PivotStyles
        },
        PlainCard: {
            styles: HoverCard_styles_1.PlainCardStyles
        },
        PrimaryButton: {
            styles: PrimaryButton_styles_1.PrimaryButtonStyles
        },
        Rating: {
            styles: Rating_styles_1.RatingStyles
        },
        SearchBox: {
            styles: SearchBox_styles_1.SearchBoxStyles
        },
        Slider: {
            styles: Slider_styles_1.SliderStyles
        },
        SpinButton: {
            styles: SpinButton_styles_1.SpinButtonStyles
        },
        Suggestions: {
            styles: PickerSuggestions_styles_1.SuggestionsStyles
        },
        SuggestionItem: {
            styles: PickerSuggestions_styles_1.SuggestionItemStyles
        },
        TagItem: {
            styles: TagPicker_styles_1.TagItemStyles
        },
        TagPicker: {
            styles: BasePicker_styles_1.BasePickerStyles
        },
        TeachingBubble: {
            styles: TeachingBubble_styles_1.TeachingBubbleStyles
        },
        TeachingBubbleContent: {
            styles: TeachingBubble_styles_1.TeachingBubbleStyles
        },
        TextField: {
            styles: TextField_styles_1.TextFieldStyles
        },
        Toggle: {
            styles: Toggle_styles_1.ToggleStyles
        }
    };
});
