// @microsoft/sp-webpart-base/lib/test-api/MockWebPartContext.js
import * as tslib_1 from "tslib";
import { SPEvent, ServiceScope } from '@microsoft/sp-core-library';
import { WebPartContext } from '../index'; 
import MockWebPartManifestInstance from './MockWebPartManifestInstance';
var MockWebPartContextParameters =  (function () {
    function MockWebPartContextParameters() {
        this.domElement = document.createElement('div');
        this.manifest = new MockWebPartManifestInstance();
        this.instanceId = 'wpInstanceId';
        this.webPartTag = 'WebPart.daf0b71c-6de8-4ef7-b511-faae7c388708.0ae4973a-70e3-451e-9f1f-3f25e5455a60';
        this.loggingTag = 'WebPart.daf0b71c-6de8-4ef7-b511-faae7c388708.0ae4973a-70e3-451e-9f1f-3f25e5455a60';
        this.httpClient = {};
        this.spHttpClient = {};
        this.graphHttpClient = {};
        this.parentServiceScope = ServiceScope.startNewRoot();
        this.statusRenderer = {};
        this.host = {};
        this.formFactor = 0 ;
        this.propertyPane = {
            open: function () { },
            refresh: function () { }
        };
        this._dataUpdatedEvent = new SPEvent('webPart_exampleInstanceId_dataUpdated');
        this.isPropertyPaneRenderedByWebPart = function () { return true; };
        this.isPropertyPaneOpen = function () { return false; };
        this.requestPropertyPaneAction = function () { }; 
    }
    return MockWebPartContextParameters;
}());
export { MockWebPartContextParameters };
var MockWebPartContext =  (function (_super) {
    tslib_1.__extends(MockWebPartContext, _super);
    function MockWebPartContext(parameters) {
        return _super.call(this, parameters) || this;
    }
    MockWebPartContext.prototype.initialize = function (webPart) {
        var parentServiceScope = this.serviceScope.getParent();
        if (parentServiceScope) {
            parentServiceScope.finish();
        }
        webPart._internalInitialize(this);
        return webPart;
    };
    return MockWebPartContext;
}(WebPartContext));
export default MockWebPartContext;
