// office-ui-fabric-react/lib/components/Button/examples/Button.CustomSplit.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { HighContrastSelector } from '../../../Styling';
import { IconButton } from 'office-ui-fabric-react/lib/Button';
import { Label } from 'office-ui-fabric-react/lib/Label';
var alertClicked = function () {
    alert('Clicked');
};
var ButtonSplitCustomExample = /** @class */ (function (_super) {
    tslib_1.__extends(ButtonSplitCustomExample, _super);
    function ButtonSplitCustomExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ButtonSplitCustomExample.prototype.render = function () {
        var _a = this.props, disabled = _a.disabled, checked = _a.checked;
        var customSplitButtonStyles = function () {
            var _a;
            return {
                splitButtonMenuButton: { backgroundColor: 'white', width: '10px', border: 'none' },
                splitButtonMenuIcon: { fontSize: '7px' },
                splitButtonDivider: { borderLeft: '1px solid #c8c8c8', right: 17 },
                splitButtonContainer: {
                    selectors: (_a = {},
                        _a[HighContrastSelector] = {
                            border: 'none'
                        },
                        _a)
                }
            };
        };
        return (React.createElement("div", null,
            React.createElement(Label, null, "Split button with icon and custom styles"),
            React.createElement(IconButton, { "data-automation-id": "test", disabled: disabled, checked: checked, iconProps: { iconName: 'Upload' }, text: "Create account", onClick: alertClicked, split: true, "aria-roledescription": 'split button', styles: customSplitButtonStyles(), menuProps: {
                    items: [
                        {
                            key: 'emailMessage',
                            text: 'Email message',
                            iconProps: { iconName: 'Mail' }
                        },
                        {
                            key: 'calendarEvent',
                            text: 'Calendar event',
                            iconProps: { iconName: 'Calendar' }
                        }
                    ]
                } })));
    };
    return ButtonSplitCustomExample;
}(React.Component));
export { ButtonSplitCustomExample };
