// office-ui-fabric-react/lib-commonjs/components/Shimmer/examples/Shimmer.Styling.Example.js
"use strict";
// @codepen
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var Shimmer_1 = require("office-ui-fabric-react/lib/Shimmer");
var Styling_1 = require("office-ui-fabric-react/lib/Styling");
var wrapperClass = Styling_1.mergeStyles({
    padding: 2,
    selectors: {
        '& > *': {
            margin: '10px 0'
        }
    }
});
var ShimmerStylingExample = /** @class */ (function (_super) {
    tslib_1.__extends(ShimmerStylingExample, _super);
    function ShimmerStylingExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._getShimmerStyles = function (props) {
            return {
                shimmerWrapper: [
                    {
                        backgroundColor: '#deecf9',
                        backgroundImage: 'linear-gradient(to right, rgba(255, 255, 255, 0) 0%, #c7e0f4 50%, rgba(255, 255, 255, 0) 100%)'
                    }
                ]
            };
        };
        return _this;
    }
    ShimmerStylingExample.prototype.render = function () {
        return (React.createElement("div", { className: wrapperClass },
            React.createElement(Shimmer_1.Shimmer, { width: "75%", styles: this._getShimmerStyles }),
            React.createElement(Shimmer_1.Shimmer, { width: "75%", styles: this._getShimmerStyles }),
            React.createElement(Shimmer_1.Shimmer, { width: "75%", styles: this._getShimmerStyles }),
            React.createElement(Shimmer_1.Shimmer, { width: "75%", styles: this._getShimmerStyles }),
            React.createElement(Shimmer_1.Shimmer, { width: "75%", styles: this._getShimmerStyles })));
    };
    return ShimmerStylingExample;
}(React.Component));
exports.ShimmerStylingExample = ShimmerStylingExample;
