// @microsoft/sp-application-base/lib/navigator/Navigator.js
import * as tslib_1 from "tslib";
import { Guid, Validate, _SPEventManager, _SPKillSwitch } from '@microsoft/sp-core-library';
import { _SPLoaderFlights } from '@microsoft/sp-loader';
import { _QosMonitor } from '@microsoft/sp-diagnostics';
import BaseApplication from '../BaseApplication';
import NavigationDataProvider from './NavigationDataProvider';
import NavigationOrchestrator from './NavigationOrchestrator';
import PrefetchedDataEventArgs from './PrefetchedDataEventArgs';
import { Killswitches } from '../common/Killswitches';
var navigateQosScenarioName = 'Navigator.navigate';
var navigateToPreloadedDataQosScenarioName = 'Navigator.navigateToPreloadedData';
var prefetchNavigationDataQosScenarioName = 'Navigator.prefetch';
var NAVIGATE_TO_URL_ON_ERROR = Guid.parse('9078878d-3ba1-41e4-8faf-fb4e2e69f3dc');
var Navigator =  (function () {
    function Navigator(serviceScope, applicationManager) {
        var _this = this;
        Validate.isNotNullOrUndefined(serviceScope, 'serviceScope');
        this._navigationDataProvider = new NavigationDataProvider(serviceScope, function (data) {
            _this._validatePreloadedData(data);
        });
        this._navigationOrchestrator = new NavigationOrchestrator(serviceScope, applicationManager, this);
    }
    Object.defineProperty(Navigator.prototype, "preloadedData", {
        get: function () {
            Validate.isNotNullOrUndefined(this._preloadedData, 'preloadedData');
            return this._preloadedData;
        },
        enumerable: true,
        configurable: true
    });
    Navigator.prototype.navigate = function (url, props) {
        var _this = this;
        Validate.isNonemptyString(url, 'url');
        var qosMonitor = new _QosMonitor(navigateQosScenarioName);
        return this._navigationDataProvider.getData(url, props).then(function (response) {
            return response.preloadedData.then(function (preloadedData) {
                _this._processDataPrefetch(url, preloadedData, response.prefetchedData, true);
                return _this.navigateToPreloadedData(preloadedData);
            }).then(function (navigationResult) {
                qosMonitor.writeSuccess();
                return navigationResult;
            });
        }).catch(function (error) {
            qosMonitor.writeUnexpectedFailure(undefined, error);
            if (_SPKillSwitch.isActivated(NAVIGATE_TO_URL_ON_ERROR, '04/16/2018', 'Navigate to url on error')) {
                location.href = url;
            }
            throw error;
        });
    };
    Navigator.prototype.prefetch = function (url, props) {
        var _this = this;
        if (!_SPLoaderFlights._useNewBootSequence()) {
            return Promise.resolve();
        }
        var prefetchProps = Killswitches.usePrefetchHeader() ? tslib_1.__assign({}, props, { isPrefetchRequest: true }) : props;
        Validate.isNonemptyString(url, 'url');
        var qosMonitor = new _QosMonitor(prefetchNavigationDataQosScenarioName);
        return new Promise(function (resolve, reject) {
            _this._navigationDataProvider.getData(url, prefetchProps).then(function (response) {
                response.preloadedData.then(function (preloadedData) {
                    _this._processDataPrefetch(url, preloadedData, response.prefetchedData, false);
                    qosMonitor.writeSuccess();
                    resolve();
                });
            }).catch(function (error) {
                qosMonitor.writeUnexpectedFailure(undefined, error);
                reject(error);
            });
        });
    };
    Navigator.prototype.navigateToPreloadedData = function (preloadedData) {
        var _this = this;
        var qosMonitor = new _QosMonitor(navigateToPreloadedDataQosScenarioName);
        try {
            this._validatePreloadedData(preloadedData);
            this._preloadedData = preloadedData;
            return this._navigationOrchestrator.navigate(preloadedData).then(function (navigationResult) {
                _this._preloadedData = navigationResult.preloadedData;
                qosMonitor.writeSuccess();
                return navigationResult;
            }).catch(function (error) {
                qosMonitor.writeUnexpectedFailure('AsyncError', error);
                throw error;
            });
        }
        catch (error) {
            qosMonitor.writeUnexpectedFailure('SyncError', error);
            return Promise.reject(error);
        }
    };
    Navigator.prototype.navigateToApplication = function (preloadedData) {
        var _this = this;
        var qosMonitor = new _QosMonitor(navigateToPreloadedDataQosScenarioName);
        try {
            this._validatePreloadedData(preloadedData);
            this._preloadedData = preloadedData;
            return this._navigationOrchestrator.navigate(preloadedData).then(function (navigationResult) {
                _this._preloadedData = navigationResult.preloadedData;
                qosMonitor.writeSuccess();
                return Promise.resolve(navigationResult.application);
            });
        }
        catch (error) {
            qosMonitor.writeUnexpectedFailure(error);
            throw error;
        }
    };
    Navigator.prototype._loadApplicationCustomizers = function (preloadedData) {
        return this._navigationOrchestrator._loadApplicationCustomizers(preloadedData);
    };
    Navigator.prototype.invalidate = function (url) {
        return this._navigationDataProvider.invalidate(url);
    };
    Navigator.prototype._validatePreloadedData = function (preloadedData) {
        Validate.isNotNullOrUndefined(preloadedData, 'preloadedData');
        Validate.isNotNullOrUndefined(preloadedData.spPageContextInfo, 'preloadedData.spPageContextInfo');
        Validate.isNonemptyString(preloadedData.clientSideApplicationId, 'preloadedData.clientSideApplicationId');
    };
    Navigator.prototype._isDataPrefetchSupported = function (componentId) {
        return componentId === 'b6917cb1-93a0-4b97-a84d-7cf49975d4ec';
    };
    Navigator.prototype._processDataPrefetch = function (url, preloadedData, prefetchData, 
    shouldRaiseEvent) {
        if (prefetchData &&
            this._isDataPrefetchSupported(preloadedData.clientSideApplicationId)) {
            prefetchData.then(function (data) {
                if (shouldRaiseEvent && data) {
                    _SPEventManager.instance.raiseStickyEvent(BaseApplication._prefetchedDataEventName, new PrefetchedDataEventArgs(preloadedData.clientSideApplicationId, url, data));
                }
            });
        }
    };
    return Navigator;
}());
export default Navigator;
