// office-ui-fabric-react/lib/components/Stack/examples/Stack.Horizontal.Wrap.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { mergeStyleSets, DefaultPalette, Slider, Stack } from 'office-ui-fabric-react';
var HorizontalStackWrapExample = /** @class */ (function (_super) {
    tslib_1.__extends(HorizontalStackWrapExample, _super);
    function HorizontalStackWrapExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onWidthChange = function (value) {
            _this.setState({ stackWidth: value });
        };
        _this.state = {
            stackWidth: 100
        };
        return _this;
    }
    HorizontalStackWrapExample.prototype.render = function () {
        var styles = mergeStyleSets({
            root: {
                background: DefaultPalette.themeTertiary,
                width: this.state.stackWidth + "%"
            },
            item: {
                width: 50,
                height: 50,
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'center',
                background: DefaultPalette.themePrimary,
                color: DefaultPalette.white
            }
        });
        var tokens = {
            sectionStack: {
                childrenGap: 10
            },
            wrapStack: {
                childrenGap: 30
            }
        };
        return (React.createElement(Stack, { tokens: tokens.sectionStack },
            React.createElement(Slider, { label: "Change the stack width to see how child items wrap onto multiple rows:", min: 1, max: 100, step: 1, defaultValue: 100, showValue: true, onChange: this._onWidthChange }),
            React.createElement(Stack, { horizontal: true, wrap: true, tokens: tokens.wrapStack, className: styles.root },
                React.createElement("span", { className: styles.item }, "1"),
                React.createElement("span", { className: styles.item }, "2"),
                React.createElement("span", { className: styles.item }, "3"),
                React.createElement("span", { className: styles.item }, "4"),
                React.createElement("span", { className: styles.item }, "5"),
                React.createElement("span", { className: styles.item }, "6"),
                React.createElement("span", { className: styles.item }, "7"),
                React.createElement("span", { className: styles.item }, "8"),
                React.createElement("span", { className: styles.item }, "9"),
                React.createElement("span", { className: styles.item }, "10"))));
    };
    return HorizontalStackWrapExample;
}(React.Component));
export { HorizontalStackWrapExample };
