// office-ui-fabric-react/lib/components/DetailsList/examples/DetailsList.CustomGroupHeaders.Example.js
// @codepen
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Link } from 'office-ui-fabric-react/lib/Link';
import { DetailsList } from 'office-ui-fabric-react/lib/DetailsList';
import { createListItems, createGroups } from 'office-ui-fabric-react/lib/utilities/exampleData';
import { getTheme, mergeStyleSets } from 'office-ui-fabric-react/lib/Styling';
var theme = getTheme();
var classNames = mergeStyleSets({
    headerAndFooter: {
        borderTop: '1px solid ' + theme.palette.neutralQuaternary,
        borderBottom: '1px solid ' + theme.palette.neutralQuaternary,
        padding: '8px',
        margin: '8px 0',
        background: theme.palette.neutralLighterAlt,
        // Overlay the sizer bars
        position: 'relative',
        zIndex: 100
    },
    headerTitle: [
        theme.fonts.xLarge,
        {
            padding: '4px 0'
        }
    ],
    headerLinkSet: {
        margin: '4px -8px'
    },
    headerLink: {
        margin: '0 8px'
    }
});
var ITEMS_PER_GROUP = 20;
var GROUP_COUNT = 20;
var DetailsListCustomGroupHeadersExample = /** @class */ (function (_super) {
    tslib_1.__extends(DetailsListCustomGroupHeadersExample, _super);
    function DetailsListCustomGroupHeadersExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onRenderGroupHeader = function (props) {
            return (React.createElement("div", { className: classNames.headerAndFooter },
                React.createElement("div", { className: classNames.headerTitle }, "Custom header for " + props.group.name),
                React.createElement("div", { className: classNames.headerLinkSet },
                    React.createElement(Link, { className: classNames.headerLink, onClick: _this._onToggleSelectGroup(props) }, props.isSelected ? 'Remove selection' : 'Select group'),
                    React.createElement(Link, { className: classNames.headerLink, onClick: _this._onToggleCollapse(props) }, props.group.isCollapsed ? 'Expand group' : 'Collapse group'))));
        };
        _this._onRenderGroupFooter = function (props) {
            return (React.createElement("div", { className: classNames.headerAndFooter },
                React.createElement("em", null, "Custom footer for " + props.group.name)));
        };
        _this._items = createListItems(500);
        _this._groups = createGroups(GROUP_COUNT, 1, 0, ITEMS_PER_GROUP);
        return _this;
    }
    DetailsListCustomGroupHeadersExample.prototype.render = function () {
        return (React.createElement(DetailsList, { items: this._items, groups: this._groups, groupProps: {
                onRenderHeader: this._onRenderGroupHeader,
                onRenderFooter: this._onRenderGroupFooter
            }, ariaLabelForSelectionColumn: "Toggle selection", ariaLabelForSelectAllCheckbox: "Toggle selection for all items" }));
    };
    DetailsListCustomGroupHeadersExample.prototype._onToggleSelectGroup = function (props) {
        return function () {
            props.onToggleSelectGroup(props.group);
        };
    };
    DetailsListCustomGroupHeadersExample.prototype._onToggleCollapse = function (props) {
        return function () {
            props.onToggleCollapse(props.group);
        };
    };
    return DetailsListCustomGroupHeadersExample;
}(React.Component));
export { DetailsListCustomGroupHeadersExample };
