// office-ui-fabric-react/lib-commonjs/components/Stack/examples/Stack.Horizontal.Basic.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var Stack_1 = require("../Stack");
var Styling_1 = require("office-ui-fabric-react/lib/Styling");
var HorizontalStackBasicExample = /** @class */ (function (_super) {
    tslib_1.__extends(HorizontalStackBasicExample, _super);
    function HorizontalStackBasicExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._onClick = function () {
            alert('Clicked Stack');
        };
        return _this;
    }
    HorizontalStackBasicExample.prototype.render = function () {
        var stackStyles = {
            root: {
                background: Styling_1.DefaultPalette.themeTertiary
            }
        };
        var stackItemStyles = {
            root: {
                background: Styling_1.DefaultPalette.themePrimary,
                color: Styling_1.DefaultPalette.white,
                padding: 5
            }
        };
        var itemAlignmentStackStyles = {
            root: {
                background: Styling_1.DefaultPalette.themeTertiary,
                height: 100
            }
        };
        var containerStackTokens = { childrenGap: 5 };
        var horizontalGapStackTokens = {
            childrenGap: 10,
            padding: 10
        };
        var itemAlignmentStackTokens = {
            childrenGap: 5,
            padding: 10
        };
        var clickableStackTokens = {
            padding: 10
        };
        return (React.createElement(Stack_1.Stack, { tokens: containerStackTokens },
            React.createElement("span", null, "Default horizontal stack"),
            React.createElement(Stack_1.Stack, { horizontal: true, disableShrink: true, styles: stackStyles },
                React.createElement("span", null, "Item One"),
                React.createElement("span", null, "Item Two"),
                React.createElement("span", null, "Item Three")),
            React.createElement("span", null, "Ordered stack"),
            React.createElement(Stack_1.Stack, { horizontal: true, disableShrink: true, styles: stackStyles },
                React.createElement(Stack_1.Stack.Item, { order: 2 },
                    React.createElement("span", null, "Item One")),
                React.createElement(Stack_1.Stack.Item, { order: 3 },
                    React.createElement("span", null, "Item Two")),
                React.createElement(Stack_1.Stack.Item, { order: 1 },
                    React.createElement("span", null, "Item Three"))),
            React.createElement("span", null, "Horizontal gap between items"),
            React.createElement(Stack_1.Stack, { horizontal: true, disableShrink: true, styles: stackStyles, tokens: horizontalGapStackTokens },
                React.createElement("span", null, "Item One"),
                React.createElement("span", null, "Item Two"),
                React.createElement("span", null, "Item Three")),
            React.createElement("span", null, "Item alignments"),
            React.createElement(Stack_1.Stack, { horizontal: true, disableShrink: true, styles: itemAlignmentStackStyles, tokens: itemAlignmentStackTokens },
                React.createElement(Stack_1.Stack.Item, { align: "auto", styles: stackItemStyles },
                    React.createElement("span", null, "Auto-aligned item")),
                React.createElement(Stack_1.Stack.Item, { align: "stretch", styles: stackItemStyles },
                    React.createElement("span", null, "Stretch-aligned item")),
                React.createElement(Stack_1.Stack.Item, { align: "baseline", styles: stackItemStyles },
                    React.createElement("span", null, "Baseline-aligned item")),
                React.createElement(Stack_1.Stack.Item, { align: "start", styles: stackItemStyles },
                    React.createElement("span", null, "Start-aligned item")),
                React.createElement(Stack_1.Stack.Item, { align: "center", styles: stackItemStyles },
                    React.createElement("span", null, "Center-aligned item")),
                React.createElement(Stack_1.Stack.Item, { align: "end", styles: stackItemStyles },
                    React.createElement("span", null, "End-aligned item"))),
            React.createElement("span", null, "Clickable stack"),
            React.createElement(Stack_1.Stack, { horizontal: true, disableShrink: true, onClick: this._onClick, styles: stackStyles, tokens: clickableStackTokens },
                React.createElement("span", null, "Click inside this box"))));
    };
    return HorizontalStackBasicExample;
}(React.Component));
exports.HorizontalStackBasicExample = HorizontalStackBasicExample;
