// office-ui-fabric-react/lib-commonjs/components/Stack/examples/Stack.Horizontal.Grow.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var Stack_1 = require("../Stack");
var Styling_1 = require("office-ui-fabric-react/lib/Styling");
var HorizontalStackGrowExample = /** @class */ (function (_super) {
    tslib_1.__extends(HorizontalStackGrowExample, _super);
    function HorizontalStackGrowExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HorizontalStackGrowExample.prototype.render = function () {
        var stackStyles = {
            root: {
                background: Styling_1.DefaultPalette.themeTertiary
            }
        };
        var stackItemStyles = {
            root: {
                alignItems: 'center',
                background: Styling_1.DefaultPalette.themePrimary,
                color: Styling_1.DefaultPalette.white,
                display: 'flex',
                height: 50,
                justifyContent: 'center'
            }
        };
        var stackTokens = {
            childrenGap: 5,
            padding: 10
        };
        return (React.createElement(Stack_1.Stack, { horizontal: true, styles: stackStyles, tokens: stackTokens },
            React.createElement(Stack_1.Stack.Item, { grow: 3, styles: stackItemStyles }, "Grow is 3"),
            React.createElement(Stack_1.Stack.Item, { grow: 2, styles: stackItemStyles }, "Grow is 2"),
            React.createElement(Stack_1.Stack.Item, { grow: true, styles: stackItemStyles }, "Grow is 1")));
    };
    return HorizontalStackGrowExample;
}(React.Component));
exports.HorizontalStackGrowExample = HorizontalStackGrowExample;
