// @microsoft/api-extractor-model/lib/mixins/ApiItemContainerMixin.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.s
Object.defineProperty(exports, "__esModule", { value: true });
const ApiItem_1 = require("../items/ApiItem");
const ApiNameMixin_1 = require("./ApiNameMixin");
const _members = Symbol('ApiItemContainerMixin._members');
const _membersSorted = Symbol('ApiItemContainerMixin._membersSorted');
const _membersByCanonicalReference = Symbol('ApiItemContainerMixin._membersByCanonicalReference');
const _membersByName = Symbol('ApiItemContainerMixin._membersByName');
/**
 * Mixin function for {@link (ApiDeclaredItem:interface)}.
 *
 * @param baseClass - The base class to be extended
 * @returns A child class that extends baseClass, adding the {@link (ApiItemContainerMixin:interface)} functionality.
 *
 * @public
 */
function ApiItemContainerMixin(baseClass) {
    class MixedClass extends baseClass {
        // tslint:disable-next-line:no-any
        constructor(...args) {
            super(...args);
            const options = args[0];
            this[_members] = [];
            this[_membersByCanonicalReference] = new Map();
            if (options.members) {
                for (const member of options.members) {
                    this.addMember(member);
                }
            }
        }
        /** @override */
        static onDeserializeInto(options, jsonObject) {
            baseClass.onDeserializeInto(options, jsonObject);
            options.members = [];
            for (const memberObject of jsonObject.members) {
                options.members.push(ApiItem_1.ApiItem.deserialize(memberObject));
            }
        }
        get members() {
            if (!this[_membersSorted]) {
                this[_members].sort((x, y) => x.getSortKey().localeCompare(y.getSortKey()));
                this[_membersSorted] = true;
            }
            return this[_members];
        }
        addMember(member) {
            if (this[_membersByCanonicalReference].has(member.canonicalReference)) {
                throw new Error('Another member has already been added with the same name and canonicalReference');
            }
            const existingParent = member[ApiItem_1.ApiItem_parent];
            if (existingParent !== undefined) {
                throw new Error(`This item has already been added to another container: "${existingParent.displayName}"`);
            }
            this[_members].push(member);
            this[_membersByName] = undefined; // invalidate the lookup
            this[_membersSorted] = false;
            this[_membersByCanonicalReference].set(member.canonicalReference, member);
            member[ApiItem_1.ApiItem_parent] = this;
        }
        tryGetMember(canonicalReference) {
            return this[_membersByCanonicalReference].get(canonicalReference);
        }
        findMembersByName(name) {
            // Build the lookup on demand
            if (this[_membersByName] === undefined) {
                const map = new Map();
                for (const member of this[_members]) {
                    if (ApiNameMixin_1.ApiNameMixin.isBaseClassOf(member)) {
                        let list = map.get(member.name);
                        if (list === undefined) {
                            list = [];
                            map.set(member.name, list);
                        }
                        list.push(member);
                    }
                }
                this[_membersByName] = map;
            }
            return this[_membersByName].get(name) || [];
        }
        /** @override */
        serializeInto(jsonObject) {
            super.serializeInto(jsonObject);
            const memberObjects = [];
            for (const member of this.members) {
                const memberJsonObject = {};
                member.serializeInto(memberJsonObject);
                memberObjects.push(memberJsonObject);
            }
            jsonObject.members = memberObjects;
        }
    }
    return MixedClass;
}
exports.ApiItemContainerMixin = ApiItemContainerMixin;
/**
 * Static members for {@link (ApiItemContainerMixin:interface)}.
 * @public
 */
(function (ApiItemContainerMixin) {
    /**
     * A type guard that tests whether the specified `ApiItem` subclass extends the `ApiItemContainerMixin` mixin.
     *
     * @remarks
     *
     * The JavaScript `instanceof` operator cannot be used to test for mixin inheritance, because each invocation of
     * the mixin function produces a different subclass.  (This could be mitigated by `Symbol.hasInstance`, however
     * the TypeScript type system cannot invoke a runtime test.)
     */
    function isBaseClassOf(apiItem) {
        return apiItem.hasOwnProperty(_members);
    }
    ApiItemContainerMixin.isBaseClassOf = isBaseClassOf;
})(ApiItemContainerMixin = exports.ApiItemContainerMixin || (exports.ApiItemContainerMixin = {}));
