// @microsoft/sp-application-base/lib/pageChrome/SPPageChrome.js
import { _SPFlight, Validate } from '@microsoft/sp-core-library';
import { _SPLoaderFlights } from '@microsoft/sp-loader';
import { Flights } from '../common/Flights';
import SPThemeProvider from './SPThemeProvider';
import SPGlobalErrorHandler from '../error/SPGlobalErrorHandler';
import DialogManagerConfiguration from '../extensibility/DialogManagerConfiguration';
import { _PerformanceLogger } from '@ms/sp-telemetry';
import styles from './SPPageChrome.module.scss';
var SPPageChrome =  (function () {
    function SPPageChrome(serviceScope, parentChrome) {
        this._isDisposed = false;
        this._masterChrome = parentChrome;
        this._createPageChrome();
        DialogManagerConfiguration.instance.initialize(this.dialogContainerDiv);
        this._spThemeProvider = new SPThemeProvider(serviceScope);
    }
    SPPageChrome.prototype.changeSuiteNavHeight = function (pixels) {
        var top = parseInt(pixels, 10) + 'px';
        this.suiteNavDiv.style.height = top;
    };
    Object.defineProperty(SPPageChrome.prototype, "themeProvider", {
        get: function () {
            return this._spThemeProvider;
        },
        enumerable: true,
        configurable: true
    });
    SPPageChrome.prototype.show = function () {
        if (!this._isDisposed) {
            if (Flights.useNewChromeSequence()) {
                this._masterChrome.show();
                this._masterChrome.chromeDiv.appendChild(this.appAndPropertyPaneDiv);
                this._masterChrome.chromeDiv.appendChild(this.dialogContainerDiv);
            }
            else {
                document.body.appendChild(this._chromeDiv);
            }
        }
    };
    SPPageChrome.prototype.hide = function () {
        if (!this._isDisposed) {
            if (Flights.useNewChromeSequence()) {
                this.appAndPropertyPaneDiv = this._masterChrome.chromeDiv.removeChild(this.appAndPropertyPaneDiv);
                this.dialogContainerDiv = this._masterChrome.chromeDiv.removeChild(this.dialogContainerDiv);
            }
            else {
                this._chromeDiv = document.body.removeChild(this._chromeDiv);
            }
        }
    };
    SPPageChrome.prototype.dispose = function () {
        if (_SPLoaderFlights._useNewBootSequence() && !this._isDisposed) {
            delete this.appDiv;
            delete this.propertyPaneContainerDiv;
            if (this.appAndPropertyPaneDiv) {
                delete this.appAndPropertyPaneDiv;
            }
            delete this.dialogContainerDiv;
            if (!Flights.useNewChromeSequence()) {
                delete this.suiteNavDiv;
                delete this._chromeDiv;
            }
            this._isDisposed = true;
        }
    };
    Object.defineProperty(SPPageChrome.prototype, "isDisposed", {
        get: function () {
            return this._isDisposed;
        },
        enumerable: true,
        configurable: true
    });
    SPPageChrome.prototype._createPageChrome = function () {
        _PerformanceLogger.devMark('SPPageChrome.createPageChrome');
        if (Flights.useNewChromeSequence()) {
            Validate.isNotNullOrUndefined(this._masterChrome, 'masterChrome');
            Validate.isNotDisposed(this._masterChrome, 'masterChrome');
            Validate.isNotNullOrUndefined(this._masterChrome.chromeDiv, 'chromeDiv');
            Validate.isNotNullOrUndefined(this._masterChrome.suiteNavDiv, 'suiteNavDiv');
            this.suiteNavDiv = this._masterChrome.suiteNavDiv;
        }
        else {
            document.body.className = 'ms-font-m';
            this._chromeDiv = document.createElement('div');
            this._chromeDiv.setAttribute('class', 'SPPageChrome');
            this.suiteNavDiv = document.createElement('div');
            this.suiteNavDiv.setAttribute('id', 'SuiteNavPlaceHolder');
            this._chromeDiv.appendChild(this.suiteNavDiv);
        }
        this.appDiv = document.createElement('div');
        this.appDiv.setAttribute('class', "SPPageChrome-app");
        this.appDiv.setAttribute('id', 'spPageChromeAppDiv');
        this.appDiv.setAttribute('data-sp-feature-tag', 'Page Chrome');
        this.appDiv.setAttribute('data-sp-feature-instance-id', '_Page Chrome');
        this.appDiv.setAttribute('data-sp-a11y-navigatebyhierarchy', 'true');
        this.propertyPaneContainerDiv = document.createElement('div');
        this.propertyPaneContainerDiv.setAttribute('class', "spPropertyPaneContainer " + styles.spPageContentTransition);
        this.propertyPaneContainerDiv.setAttribute('id', 'spPropertyPaneContainer');
        this.appAndPropertyPaneDiv = document.createElement('div');
        this.appAndPropertyPaneDiv.setAttribute('class', _SPFlight.isEnabled(175) ? 
            'spAppAndPropertyPanelContainer spAppAndPropertyPanelNavBelowHeader' :
            'spAppAndPropertyPanelContainer');
        this.appAndPropertyPaneDiv.appendChild(this.appDiv);
        this.appAndPropertyPaneDiv.appendChild(this.propertyPaneContainerDiv);
        this.dialogContainerDiv = document.createElement('div');
        this.dialogContainerDiv.setAttribute('id', 'spDialogContainerDiv');
        if (!Flights.useNewChromeSequence()) {
            this._chromeDiv.appendChild(this.appAndPropertyPaneDiv);
            this._chromeDiv.appendChild(this.dialogContainerDiv);
            if (!window.onerror) {
                window.onerror = SPGlobalErrorHandler.basicTracingHandler;
            }
        }
    };
    return SPPageChrome;
}());
export default SPPageChrome;
