// @microsoft/sp-property-pane/lib/propertyPaneDynamicData/dynamicDataWidgetEntry/DynamicDataWidgetEntry.js
import * as tslib_1 from "tslib";
import { autobind } from '@microsoft/office-ui-fabric-react-bundle';
import { DynamicDataProvider } from '@microsoft/sp-component-base';
import { Log, Text } from '@microsoft/sp-core-library';
import { Dropdown } from 'office-ui-fabric-react/lib/components/Dropdown/Dropdown';
import * as React from 'react';
import Strings from '../../loc/Strings.resx';
var MAX_PROPERTY_VALUE_DEPTH = 2;
var DynamicDataWidgetEntry =  (function (_super) {
    tslib_1.__extends(DynamicDataWidgetEntry, _super);
    function DynamicDataWidgetEntry(props) {
        var _this = _super.call(this, props) || this;
        _this._selectedSubPropertyText = '';
        _this.state = {
            selectedSubPropertyPath: _this.props.value._getPropertyPath(),
            selectedSubPropertyText: undefined
        };
        _this._updateReference();
        return _this;
    }
    DynamicDataWidgetEntry.prototype.componentWillReceiveProps = function (newProps) {
        var _this = this;
        if (this.props.selectedSource.id !== newProps.selectedSource.id ||
            this.props.selectedPropertyId !== newProps.selectedPropertyId) {
            this.setState({
                selectedSubPropertyText: '',
                selectedSubPropertyPath: ''
            }, function () {
                _this._updateReference();
            });
        }
    };
    DynamicDataWidgetEntry.prototype.render = function () {
        var selectedPropertyId = this.props.selectedPropertyId;
        var selectedSubPropertyPath = this.state.selectedSubPropertyPath;
        var dropDowns = [];
        if (selectedPropertyId) {
            var level = 0;
            var currentPropertyPath = selectedPropertyId;
            var subProperties = selectedSubPropertyPath ? selectedSubPropertyPath.split('.') : [];
            var maxDepth = this._getMaxSubPropertyDepth();
            while (level < maxDepth && currentPropertyPath && subProperties) {
                dropDowns.push(this._getDropdownForSubProperty(level, currentPropertyPath, subProperties[level]));
                currentPropertyPath = subProperties[level] ? currentPropertyPath + '.' + subProperties[level] : '';
                level++;
            }
        }
        return (React.createElement("div", null, dropDowns));
    };
    DynamicDataWidgetEntry.prototype._getDropdownForSubProperty = function (level, propertyPath, selectedKey) {
        var key = this.props.parentKey + 'sub-property' + level;
        selectedKey = selectedKey && selectedKey.split(DynamicDataProvider._jsonPathArrayRegex)[0];
        var options = this._getDropdownOptions(this._getSubPropertyValueObject(propertyPath), propertyPath, false, selectedKey);
        var selectedText = (level === 0) ?
            this.props.selectedPropertyText :
            (this.state.selectedSubPropertyText || this._selectedSubPropertyText);
        return (React.createElement("div", { key: key }, (options.length > 0) &&
            React.createElement(Dropdown, { label: Text.format(Strings.DynamicDataPropertiesDropdownLabel, selectedText), onChanged: this._onSubPropertySelectionChange.bind(this, level), options: options, selectedKey: selectedKey || '' })));
    };
    DynamicDataWidgetEntry.prototype._onSubPropertySelectionChange = function (level, option) {
        var _this = this;
        var _a = this.state, selectedSubPropertyPath = _a.selectedSubPropertyPath, selectedSubPropertyText = _a.selectedSubPropertyText;
        var selectedKey = option.key;
        var paths = selectedSubPropertyPath ? selectedSubPropertyPath.split('.') : [];
        selectedSubPropertyPath = '';
        for (var i = 0; i < level; i++) {
            selectedSubPropertyPath += paths[i] + '.';
        }
        selectedSubPropertyPath = selectedKey ?
            selectedSubPropertyPath + selectedKey : selectedSubPropertyPath.slice(0, -1); 
        selectedSubPropertyText = level === 0 ?
            option.text : selectedSubPropertyText;
        this.setState({
            selectedSubPropertyText: selectedSubPropertyText,
            selectedSubPropertyPath: selectedSubPropertyPath
        }, function () {
            _this._updateReference();
        });
    };
    DynamicDataWidgetEntry.prototype._getDropdownOptions = function (propValue, 
    propertyPath, recursiveCall, selectedKey) {
        var _this = this;
        var options = [];
        if (typeof propValue === 'object') {
            if (Array.isArray(propValue)) { 
                options = this._getDropdownOptions(propValue[0], propertyPath, true, selectedKey);
            }
            else if (propValue instanceof Set) {
            }
            else if (propValue instanceof Map) {
                propValue.forEach(function (value, key) {
                    options.push(_this._getDropdownOption(propertyPath, key, selectedKey));
                });
            }
            else { 
                Object.keys(propValue).forEach(function (key) {
                    options.push(_this._getDropdownOption(propertyPath, key, selectedKey));
                });
            }
            if (!recursiveCall && options.length > 1) {
                options.unshift({ key: '', text: '' });
            }
        }
        return options;
    };
    DynamicDataWidgetEntry.prototype._getDropdownOption = function (propertyPath, key, selectedKey) {
        if (key === selectedKey) {
            this._selectedSubPropertyText = key;
        }
        return {
            key: key,
            text: this._getOptionText(propertyPath, key) || key
        };
    };
    DynamicDataWidgetEntry.prototype._getOptionText = function (propertyPath, key) {
        if (this._currentAnnotatedPropertyValue && this._currentAnnotatedPropertyValue.metadata) {
            var subPaths = propertyPath.split('.').slice(1);
            var metadata = this._currentAnnotatedPropertyValue.metadata;
            for (var _i = 0, subPaths_1 = subPaths; _i < subPaths_1.length; _i++) {
                var subPath = subPaths_1[_i];
                if (metadata && metadata.hasOwnProperty(subPath)) {
                    if (key === subPath) {
                        return metadata[subPath].title;
                    }
                    metadata = metadata[subPath].metadata;
                }
                else {
                    return undefined;
                }
            }
            if (metadata && metadata.hasOwnProperty(key)) {
                return metadata[key].title;
            }
        }
        else {
            return undefined;
        }
    };
    DynamicDataWidgetEntry.prototype._updateReference = function () {
        if (this.props.selectedPropertyId) {
            var reference = this.props.selectedSource.id + ':' + this.props.selectedPropertyId;
            if (this._isSelectedPropertyValueAnArray()) {
                reference += '[*]';
            }
            if (this.state.selectedSubPropertyPath) {
                var subPropertyPath = this._getSubPropertyPath();
                if (subPropertyPath) {
                    reference += ':' + subPropertyPath;
                }
            }
            this.props.value.setReference(reference);
            this.props.onChange(this.props.targetProperty, this.props.value);
        }
    };
    DynamicDataWidgetEntry.prototype._getMaxSubPropertyDepth = function () {
        var propertyValueDepth = this.props.propertyValueDepth;
        var maxDepth = MAX_PROPERTY_VALUE_DEPTH;
        if (propertyValueDepth !== undefined) {
            if (propertyValueDepth >= 0 && propertyValueDepth < 2) {
                maxDepth = propertyValueDepth;
            }
            else {
                Log.warn('propertyValueDepth', Text.format(Strings.InvalidPropertyValueDepthWarning, 'propertyValueDepth', 0, 2));
            }
        }
        return maxDepth;
    };
    DynamicDataWidgetEntry.prototype._getSubPropertyValueObject = function (propertyPath) {
        var subPropertyValueObject; 
        var subPaths = propertyPath.split('.');
        var property = subPaths.shift();
        this._currentAnnotatedPropertyValue =
            this.props.selectedSource.getAnnotatedPropertyValue(property); 
        subPropertyValueObject = this._currentAnnotatedPropertyValue.sampleValue;
        if (typeof subPropertyValueObject === 'object') {
            for (var _i = 0, subPaths_2 = subPaths; _i < subPaths_2.length; _i++) {
                var subPath = subPaths_2[_i];
                subPath = subPath.split(DynamicDataProvider._jsonPathArrayRegex)[0];
                while (Array.isArray(subPropertyValueObject)) {
                    subPropertyValueObject = subPropertyValueObject[0];
                }
                if (typeof subPropertyValueObject === 'object') {
                    if (subPropertyValueObject.hasOwnProperty(subPath)) {
                        subPropertyValueObject = subPropertyValueObject[subPath];
                    }
                }
            }
        }
        return subPropertyValueObject;
    };
    DynamicDataWidgetEntry.prototype._isSelectedPropertyValueAnArray = function () {
        var _a = this.props, selectedSource = _a.selectedSource, selectedPropertyId = _a.selectedPropertyId;
        var sampleValue = selectedSource.getAnnotatedPropertyValue(selectedPropertyId).sampleValue;
        return Array.isArray(sampleValue);
    };
    DynamicDataWidgetEntry.prototype._getSubPropertyPath = function () {
        var _a = this.props, selectedSource = _a.selectedSource, selectedPropertyId = _a.selectedPropertyId;
        var subPropertyReferencePath = '';
        var subPropertyPaths = this.state.selectedSubPropertyPath.split('.');
        var propertyValueObject = selectedSource.getAnnotatedPropertyValue(selectedPropertyId).sampleValue;
        subPropertyPaths.forEach(function (subPropertyPath) {
            subPropertyPath = subPropertyPath.split(DynamicDataProvider._jsonPathArrayRegex)[0];
            if (propertyValueObject) {
                while (Array.isArray(propertyValueObject) && propertyValueObject.length) {
                    propertyValueObject = propertyValueObject[0];
                }
                if (propertyValueObject.hasOwnProperty(subPropertyPath)) { 
                    subPropertyReferencePath += subPropertyPath;
                    var subPropertyValueObject = propertyValueObject[subPropertyPath]; 
                    if (Array.isArray(subPropertyValueObject)) {
                        subPropertyReferencePath += '[*]';
                    }
                    subPropertyReferencePath += '.';
                    propertyValueObject = subPropertyValueObject;
                }
                else {
                    propertyValueObject = undefined;
                }
            }
        });
        return subPropertyReferencePath.replace(/\.$/, '');
    };
    tslib_1.__decorate([
        autobind
    ], DynamicDataWidgetEntry.prototype, "_onSubPropertySelectionChange", null);
    tslib_1.__decorate([
        autobind
    ], DynamicDataWidgetEntry.prototype, "_getDropdownOption", null);
    tslib_1.__decorate([
        autobind
    ], DynamicDataWidgetEntry.prototype, "_getOptionText", null);
    tslib_1.__decorate([
        autobind
    ], DynamicDataWidgetEntry.prototype, "_updateReference", null);
    tslib_1.__decorate([
        autobind
    ], DynamicDataWidgetEntry.prototype, "_getMaxSubPropertyDepth", null);
    return DynamicDataWidgetEntry;
}(React.Component));
export default DynamicDataWidgetEntry;
