// @uifabric/fluent-theme/lib-commonjs/fluent/FluentStyles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var BasePicker_styles_1 = require("./styles/BasePicker.styles");
var Breadcrumb_styles_1 = require("./styles/Breadcrumb.styles");
var DetailsList_styles_1 = require("./styles/DetailsList.styles");
var Callout_styles_1 = require("./styles/Callout.styles");
var Checkbox_styles_1 = require("./styles/Checkbox.styles");
var ChoiceGroupOption_styles_1 = require("./styles/ChoiceGroupOption.styles");
var ColorPickerGridCell_styles_1 = require("./styles/ColorPickerGridCell.styles");
var ColorPicker_styles_1 = require("./styles/ColorPicker.styles");
var ComboBox_styles_1 = require("./styles/ComboBox.styles");
var CommandBar_styles_1 = require("./styles/CommandBar.styles");
var CommandBarButton_styles_1 = require("./styles/CommandBarButton.styles");
var CompoundButton_styles_1 = require("./styles/CompoundButton.styles");
var ContextualMenu_styles_1 = require("./styles/ContextualMenu.styles");
var DatePicker_styles_1 = require("./styles/DatePicker.styles");
var DefaultButton_styles_1 = require("./styles/DefaultButton.styles");
var DetailsColumn_styles_1 = require("./styles/DetailsColumn.styles");
var Dialog_styles_1 = require("./styles/Dialog.styles");
var Dropdown_styles_1 = require("./styles/Dropdown.styles");
var HoverCard_styles_1 = require("./styles/HoverCard.styles");
var Facepile_styles_1 = require("./styles/Facepile.styles");
var IconButton_styles_1 = require("./styles/IconButton.styles");
var Label_styles_1 = require("./styles/Label.styles");
var Link_styles_1 = require("./styles/Link.styles");
var Modal_styles_1 = require("./styles/Modal.styles");
var Panel_styles_1 = require("./styles/Panel.styles");
var PeoplePicker_styles_1 = require("./styles/PeoplePicker.styles");
var Persona_styles_1 = require("./styles/Persona.styles");
var Pivot_styles_1 = require("./styles/Pivot.styles");
var PrimaryButton_styles_1 = require("./styles/PrimaryButton.styles");
var Rating_styles_1 = require("./styles/Rating.styles");
var SearchBox_styles_1 = require("./styles/SearchBox.styles");
var Slider_styles_1 = require("./styles/Slider.styles");
var SpinButton_styles_1 = require("./styles/SpinButton.styles");
var PickerSuggestions_styles_1 = require("./styles/PickerSuggestions.styles");
var TagPicker_styles_1 = require("./styles/TagPicker.styles");
var TeachingBubble_styles_1 = require("./styles/TeachingBubble.styles");
var TextField_styles_1 = require("./styles/TextField.styles");
var Toggle_styles_1 = require("./styles/Toggle.styles");
// Roll up all style overrides in a single "Fluent theme" object
// TODO: "any" is used here to get around "is using xxx but cannot be named" TS error. Should be able to remove
//        this 'any' once we upgrade to TS3.1+
// tslint:disable-next-line:no-any
exports.FluentStyles = {
    Breadcrumb: {
        styles: Breadcrumb_styles_1.BreadcrumbStyles
    },
    CalloutContent: {
        styles: Callout_styles_1.CalloutContentStyles
    },
    ColorPicker: {
        styles: ColorPicker_styles_1.ColorPickerStyles
    },
    ColorRectangle: {
        styles: ColorPicker_styles_1.ColorRectangleStyles
    },
    ColorSlider: {
        styles: ColorPicker_styles_1.ColorSliderStyles
    },
    CommandBar: {
        styles: CommandBar_styles_1.CommandBarStyles
    },
    CommandBarButton: {
        styles: CommandBarButton_styles_1.CommandBarButtonStyles
    },
    CompoundButton: {
        styles: CompoundButton_styles_1.CompoundButtonStyles
    },
    Check: {
        styles: DetailsList_styles_1.CheckStyles
    },
    Checkbox: {
        styles: Checkbox_styles_1.CheckboxStyles
    },
    ChoiceGroupOption: {
        styles: ChoiceGroupOption_styles_1.ChoiceGroupOptionStyles
    },
    ColorPickerGridCell: {
        styles: ColorPickerGridCell_styles_1.ColorPickerGridCellStyles
    },
    ComboBox: {
        styles: ComboBox_styles_1.ComboBoxStyles
    },
    CompactPeoplePicker: {
        styles: BasePicker_styles_1.BasePickerStyles
    },
    ContextualMenu: {
        styles: ContextualMenu_styles_1.ContextualMenuStyles
    },
    DatePicker: {
        styles: DatePicker_styles_1.DatePickerStyles
    },
    DefaultButton: {
        styles: DefaultButton_styles_1.DefaultButtonStyles
    },
    DetailsColumn: {
        styles: DetailsColumn_styles_1.DetailsColumnStyles
    },
    DetailsRow: {
        styles: DetailsList_styles_1.DetailsRowStyles
    },
    DialogContent: {
        styles: Dialog_styles_1.DialogContentStyles
    },
    DialogFooter: {
        styles: Dialog_styles_1.DialogFooterStyles
    },
    Dropdown: {
        styles: Dropdown_styles_1.DropdownStyles
    },
    ExpandingCard: {
        styles: HoverCard_styles_1.ExpandingCardStyles
    },
    Facepile: {
        styles: Facepile_styles_1.FacepileStyles
    },
    IconButton: {
        styles: IconButton_styles_1.IconButtonStyles
    },
    Label: {
        styles: Label_styles_1.LabelStyles
    },
    Link: {
        styles: Link_styles_1.LinkStyles
    },
    ListPeoplePickerBase: {
        styles: BasePicker_styles_1.BasePickerStyles
    },
    Modal: {
        styles: Modal_styles_1.ModalStyles
    },
    Panel: {
        styles: Panel_styles_1.PanelStyles
    },
    NormalPeoplePicker: {
        styles: BasePicker_styles_1.BasePickerStyles
    },
    PeoplePickerItem: {
        styles: PeoplePicker_styles_1.PeoplePickerItemStyles
    },
    Persona: {
        styles: Persona_styles_1.PersonaStyles
    },
    Pivot: {
        styles: Pivot_styles_1.PivotStyles
    },
    PlainCard: {
        styles: HoverCard_styles_1.PlainCardStyles
    },
    PrimaryButton: {
        styles: PrimaryButton_styles_1.PrimaryButtonStyles
    },
    Rating: {
        styles: Rating_styles_1.RatingStyles
    },
    SearchBox: {
        styles: SearchBox_styles_1.SearchBoxStyles
    },
    Slider: {
        styles: Slider_styles_1.SliderStyles
    },
    SpinButton: {
        styles: SpinButton_styles_1.SpinButtonStyles
    },
    Suggestions: {
        styles: PickerSuggestions_styles_1.SuggestionsStyles
    },
    SuggestionItem: {
        styles: PickerSuggestions_styles_1.SuggestionItemStyles
    },
    TagItem: {
        styles: TagPicker_styles_1.TagItemStyles
    },
    TagPicker: {
        styles: BasePicker_styles_1.BasePickerStyles
    },
    TeachingBubble: {
        styles: TeachingBubble_styles_1.TeachingBubbleStyles
    },
    TeachingBubbleContent: {
        styles: TeachingBubble_styles_1.TeachingBubbleStyles
    },
    TextField: {
        styles: TextField_styles_1.TextFieldStyles
    },
    Toggle: {
        styles: Toggle_styles_1.ToggleStyles
    }
};
