// @uifabric/fluent-theme/lib-commonjs/fluent/FluentTheme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var styling_1 = require("@uifabric/styling");
var FluentColors_1 = require("./FluentColors");
var FluentDepths_1 = require("./FluentDepths");
exports.FluentTheme = styling_1.createTheme({
    palette: {
        neutralDark: FluentColors_1.NeutralColors.gray190,
        neutralPrimary: FluentColors_1.NeutralColors.gray160,
        neutralPrimaryAlt: FluentColors_1.NeutralColors.gray150,
        neutralSecondary: FluentColors_1.NeutralColors.gray130,
        neutralSecondaryAlt: FluentColors_1.NeutralColors.gray110,
        neutralTertiary: FluentColors_1.NeutralColors.gray90,
        neutralTertiaryAlt: FluentColors_1.NeutralColors.gray60,
        neutralQuaternary: FluentColors_1.NeutralColors.gray50,
        neutralQuaternaryAlt: FluentColors_1.NeutralColors.gray40,
        neutralLight: FluentColors_1.NeutralColors.gray30,
        neutralLighter: FluentColors_1.NeutralColors.gray20,
        neutralLighterAlt: FluentColors_1.NeutralColors.gray10,
        // Shared Colors
        red: FluentColors_1.SharedColors.red10,
        redDark: FluentColors_1.SharedColors.red20
    },
    effects: {
        roundedCorner2: '2px',
        elevation4: FluentDepths_1.Depths.depth4,
        elevation8: FluentDepths_1.Depths.depth8,
        elevation16: FluentDepths_1.Depths.depth16,
        elevation64: FluentDepths_1.Depths.depth64
    }
});
exports.default = exports.FluentTheme;
