// @pnp/spfx-controls-react/lib/controls/iFrameDialog/IFrameDialog.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Dialog_1 = require("office-ui-fabric-react/lib/Dialog");
var IFrameDialogContent_1 = require("./IFrameDialogContent");
var telemetry = require("../../common/telemetry");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var omit = require("lodash/omit");
/**
 * Dialog component to display content in iframe
 */
var IFrameDialog = (function (_super) {
    __extends(IFrameDialog, _super);
    function IFrameDialog(props, state) {
        var _this = _super.call(this, props, state) || this;
        telemetry.track('IFrameDialog', {});
        _this.state = {
            dialogId: null
        };
        return _this;
    }
    /**
     * componentWillMount lifecycle hook
     */
    IFrameDialog.prototype.componentWillMount = function () {
        this.setState({
            dialogId: "dialog-" + sp_core_library_1.Guid.newGuid().toString()
        });
    };
    /**
     * componentDidMount lifecycle hook
     */
    IFrameDialog.prototype.componentDidMount = function () {
        this.setDialogStyling();
    };
    IFrameDialog.prototype.componentDidUpdate = function (prevProps, prevState) {
        this.setDialogStyling();
    };
    IFrameDialog.prototype.render = function () {
        var _a = this.props, iframeOnLoad = _a.iframeOnLoad, height = _a.height, allowFullScreen = _a.allowFullScreen, allowTransparency = _a.allowTransparency, marginHeight = _a.marginHeight, marginWidth = _a.marginWidth, name = _a.name, sandbox = _a.sandbox, scrolling = _a.scrolling, seamless = _a.seamless;
        return (React.createElement(Dialog_1.Dialog, __assign({ className: this.state.dialogId + " " + this.props.className }, omit(this.props, 'className')),
            React.createElement(IFrameDialogContent_1.IFrameDialogContent, { src: this.props.url, iframeOnLoad: iframeOnLoad, close: this.props.onDismiss, height: height, allowFullScreen: allowFullScreen, allowTransparency: allowTransparency, marginHeight: marginHeight, marginWidth: marginWidth, name: name, sandbox: sandbox, scrolling: scrolling, seamless: seamless })));
    };
    /**
     * Set the dialog style
     */
    IFrameDialog.prototype.setDialogStyling = function () {
        if (!this.props.hidden && this.state.dialogId) {
            var element = document.querySelector("." + this.state.dialogId + " .ms-Dialog-main");
            if (element && this.props.width) {
                element.style.width = this.props.width;
                element.style.minWidth = this.props.width;
                element.style.maxWidth = this.props.width;
            }
        }
    };
    return IFrameDialog;
}(React.Component));
exports.IFrameDialog = IFrameDialog;
