// @pnp/spfx-controls-react/lib/controls/dateTimePicker/DateTimePicker.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var sp_lodash_subset_1 = require("@microsoft/sp-lodash-subset");
var DateTimeConventions_1 = require("./DateTimeConventions");
var DatePicker_1 = require("office-ui-fabric-react/lib/DatePicker");
var Label_1 = require("office-ui-fabric-react/lib/Label");
var strings = require("ControlStrings");
var ErrorMessage_1 = require("../errorMessage/ErrorMessage");
var DateTimePicker_module_scss_1 = require("./DateTimePicker.module.scss");
var HoursComponent_1 = require("./HoursComponent");
var MinutesComponent_1 = require("./MinutesComponent");
var SecondsComponent_1 = require("./SecondsComponent");
var telemetry = require("../../common/telemetry");
var Utilities_1 = require("office-ui-fabric-react/lib/Utilities");
var _1 = require(".");
var TimeHelper_1 = require("./TimeHelper");
/**
 * Renders the controls for DateTimePicker component
 */
var DateTimePicker = (function (_super) {
    __extends(DateTimePicker, _super);
    /**
     * Constructor
     */
    function DateTimePicker(props) {
        var _this = _super.call(this, props) || this;
        _this._latestValidateValue = NaN;
        /**
         * Function called when the DatePicker Office UI Fabric component selected date changed
         */
        _this.onSelectDate = function (date) {
            if (!TimeHelper_1.TimeHelper.isValidDate(date)) {
                return;
            }
            // Get hours, minutes and seconds from state or default to zero
            var _a = _this.state, _b = _a.hours, hours = _b === void 0 ? 0 : _b, _c = _a.minutes, minutes = _c === void 0 ? 0 : _c, _d = _a.seconds, seconds = _d === void 0 ? 0 : _d;
            var day = TimeHelper_1.TimeHelper.cloneDate(date);
            day.setHours(hours);
            day.setMinutes(minutes);
            day.setSeconds(seconds);
            _this.setState({ day: day }, function () { return _this.delayedValidate(_this.state.day); });
        };
        /**
         * Function called when hours value have been changed
         * @param element Hours dropdown value
         */
        _this.dropdownHoursChanged = function (value) {
            _this.setState(function (_a) {
                var day = _a.day;
                var hoursSplit = value.split(" ");
                var hoursValue = hoursSplit[0].length > 2 ? hoursSplit[0].substring(0, 2) : hoursSplit[0];
                var hours = parseInt(hoursValue);
                if (isNaN(hours)) {
                    return;
                }
                if (_this.props.timeConvention !== DateTimeConventions_1.TimeConvention.Hours24) {
                    if (hoursSplit[1] && hoursSplit[1].toLowerCase().indexOf("pm") !== -1) {
                        hours += 12;
                        if (hours === 24) {
                            hours = 0;
                        }
                    }
                }
                if (hours > 23) {
                    return;
                }
                var state = { hours: hours };
                var date = TimeHelper_1.TimeHelper.cloneDate(day);
                if (date) {
                    date.setHours(hours);
                    state.day = date;
                }
                return state;
            }, function () { return _this.delayedValidate(_this.state.day); });
        };
        /**
         * Function called when minutes value have been changed
         * @param element Minutes dropdown value
         */
        _this.dropdownMinutesChanged = function (value) {
            _this.setState(function (_a) {
                var day = _a.day;
                var minutes = parseInt(value.length > 2 ? value.substring(0, 2) : value);
                var state = { minutes: minutes };
                var date = TimeHelper_1.TimeHelper.cloneDate(day);
                if (date) {
                    date.setMinutes(minutes);
                    state.day = date;
                }
                return state;
            }, function () { return _this.delayedValidate(_this.state.day); });
        };
        /**
         * Function called when seconds value have been changed
         * @param element Seconds dropdown value
         */
        _this.dropdownSecondsChanged = function (value) {
            _this.setState(function (_a) {
                var day = _a.day;
                var seconds = parseInt(value.length > 2 ? value.substring(0, 2) : value);
                var state = { seconds: seconds };
                var date = TimeHelper_1.TimeHelper.cloneDate(day);
                if (date) {
                    date.setSeconds(seconds);
                    state.day = date;
                }
                return state;
            }, function () { return _this.delayedValidate(_this.state.day); });
        };
        /**
         * Validates the new custom field value
         */
        _this.validate = function (dateVal) {
            if (typeof _this.props.onGetErrorMessage === 'undefined' || _this.props.onGetErrorMessage === null) {
                _this.notifyAfterValidate(_this.props.value, dateVal);
                return;
            }
            var timestamp = dateVal.getTime();
            if (_this._latestValidateValue === timestamp) {
                return;
            }
            _this._latestValidateValue = timestamp;
            var result = _this.props.onGetErrorMessage(dateVal);
            if (typeof result !== 'undefined') {
                if (typeof result === 'string') {
                    if (result === '') {
                        _this.notifyAfterValidate(_this.props.value, dateVal);
                    }
                    _this.setState({
                        errorMessage: result
                    });
                }
                else {
                    result.then(function (errorMessage) {
                        if (typeof errorMessage === 'undefined' || errorMessage === '') {
                            _this.notifyAfterValidate(_this.props.value, dateVal);
                        }
                        _this.setState({
                            errorMessage: errorMessage
                        });
                    });
                }
            }
            else {
                _this.notifyAfterValidate(_this.props.value, dateVal);
            }
        };
        /**
         * Notifies the parent Web Part of a property value change
         */
        _this.notifyAfterValidate = function (oldValue, newValue) {
            if (typeof _this.props.onChange !== 'undefined' && _this.props.onChange !== null && newValue !== null) {
                _this.props.onChange(newValue);
            }
        };
        telemetry.track('ReactDateTimePicker', {
            dateConvention: props.dateConvention ? DateTimeConventions_1.DateConvention[props.dateConvention] : '',
            formatDate: !!props.formatDate,
            timeConvention: props.timeConvention ? DateTimeConventions_1.TimeConvention[props.timeConvention] : '',
            disabled: props.disabled
        });
        // Get the current date/time values
        var _a = _this.props.value, value = _a === void 0 ? null : _a;
        var _b = DateTimePicker.getDateComponents(value, props.dateConvention), day = _b.day, hours = _b.hours, minutes = _b.minutes, seconds = _b.seconds;
        // Set the current state
        _this.state = {
            day: day,
            hours: hours,
            minutes: minutes,
            seconds: seconds,
            errorMessage: ''
        };
        _this.async = new Utilities_1.Async(_this);
        _this.delayedValidate = _this.async.debounce(_this.validate, props.deferredValidationTime);
        return _this;
    }
    /**
     * Called when the component will unmount
     */
    DateTimePicker.prototype.componentWillUnmount = function () {
        this.async.dispose();
    };
    /**
     * Called before the component receives new props, used for matching state with new props.
     */
    DateTimePicker.prototype.componentWillReceiveProps = function (nextProps) {
        if (nextProps.value != null && !sp_lodash_subset_1.isEqual(nextProps.value, this.props.value)) {
            var _a = DateTimePicker.getDateComponents(nextProps.value, this.props.dateConvention), day = _a.day, hours = _a.hours, minutes = _a.minutes, seconds = _a.seconds;
            this.setState({ day: day, hours: hours, minutes: minutes, seconds: seconds });
        }
    };
    /**
     * Get the components of a Date object matching the DateConvention settings
     * @param date Date to extract components from
     * @param dateConvention DateConvention to follow when extracting date components
     */
    DateTimePicker.getDateComponents = function (date, dateConvention) {
        var day = TimeHelper_1.TimeHelper.cloneDate(date);
        var hours = dateConvention !== DateTimeConventions_1.DateConvention.Date && day !== null ? day.getHours() : 0;
        var minutes = dateConvention !== DateTimeConventions_1.DateConvention.Date && day !== null ? day.getMinutes() : 0;
        var seconds = dateConvention !== DateTimeConventions_1.DateConvention.Date && day !== null ? day.getSeconds() : 0;
        if (day !== null) {
            day.setHours(hours);
            day.setMinutes(minutes);
            day.setSeconds(seconds);
        }
        return { day: day, hours: hours, minutes: minutes, seconds: seconds };
    };
    /**
     * Renders the control
     */
    DateTimePicker.prototype.render = function () {
        // use date value from props if it exists, otherwise use internal state
        // if no dateConvention is set, default is DateConvention.DateTime
        var _a = this.props, label = _a.label, disabled = _a.disabled, timeConvention = _a.timeConvention, _b = _a.dateConvention, dateConvention = _b === void 0 ? DateTimeConventions_1.DateConvention.DateTime : _b, firstDayOfWeek = _a.firstDayOfWeek, _c = _a.isMonthPickerVisible, isMonthPickerVisible = _c === void 0 ? true : _c, showGoToToday = _a.showGoToToday, _d = _a.showMonthPickerAsOverlay, showMonthPickerAsOverlay = _d === void 0 ? false : _d, _e = _a.showWeekNumbers, showWeekNumbers = _e === void 0 ? false : _e, _f = _a.showSeconds, showSeconds = _f === void 0 ? false : _f, formatDate = _a.formatDate, _g = _a.value, value = _g === void 0 ? this.state.day : _g, _h = _a.strings // Defines the DatePicker control labels
        , dateStrings = _h === void 0 ? new _1.DateTimePickerStrings() : _h // Defines the DatePicker control labels
        ;
        var hours = value != null ? value.getHours() : this.state.hours;
        var minutes = value != null ? value.getMinutes() : this.state.minutes;
        var seconds = value != null ? value.getSeconds() : this.state.seconds;
        // Check if the time element needs to be rendered
        var timeElm = React.createElement("div", { className: "hidden" });
        if (dateConvention === DateTimeConventions_1.DateConvention.DateTime) {
            timeElm = (React.createElement("div", { className: Utilities_1.css(DateTimePicker_module_scss_1.default.row, DateTimePicker_module_scss_1.default.timeRow) },
                React.createElement("div", { className: DateTimePicker_module_scss_1.default.labelCell },
                    React.createElement(Label_1.Label, { className: DateTimePicker_module_scss_1.default.fieldLabel }, strings.DateTimePickerTime)),
                React.createElement("div", { className: DateTimePicker_module_scss_1.default.time },
                    React.createElement("div", { className: DateTimePicker_module_scss_1.default.picker },
                        React.createElement(HoursComponent_1.default, { disabled: disabled, timeConvention: timeConvention, value: hours, onChange: this.dropdownHoursChanged })),
                    React.createElement("div", { className: DateTimePicker_module_scss_1.default.separator },
                        React.createElement(Label_1.Label, null, dateStrings.timeSeparator)),
                    React.createElement("div", { className: DateTimePicker_module_scss_1.default.picker },
                        React.createElement(MinutesComponent_1.default, { disabled: disabled, value: minutes, onChange: this.dropdownMinutesChanged })),
                    showSeconds && (React.createElement("div", { className: DateTimePicker_module_scss_1.default.separator },
                        React.createElement(Label_1.Label, null, dateStrings.timeSeparator))),
                    showSeconds && (React.createElement("div", { className: DateTimePicker_module_scss_1.default.picker },
                        React.createElement(SecondsComponent_1.default, { disabled: disabled, value: seconds, onChange: this.dropdownSecondsChanged }))))));
        }
        // Renders content
        return (React.createElement("div", { className: DateTimePicker_module_scss_1.default.dateTimePicker },
            label && React.createElement(Label_1.Label, null, label),
            React.createElement("div", { className: DateTimePicker_module_scss_1.default.container },
                React.createElement("div", { className: DateTimePicker_module_scss_1.default.row },
                    React.createElement("div", { className: DateTimePicker_module_scss_1.default.labelCell },
                        React.createElement(Label_1.Label, { className: DateTimePicker_module_scss_1.default.fieldLabel }, strings.DateTimePickerDate)),
                    React.createElement("div", { className: DateTimePicker_module_scss_1.default.picker },
                        React.createElement(DatePicker_1.DatePicker, { formatDate: formatDate, disabled: disabled, value: value, strings: dateStrings, isMonthPickerVisible: true, onSelectDate: this.onSelectDate, allowTextInput: false, firstDayOfWeek: firstDayOfWeek, showGoToToday: showGoToToday, showMonthPickerAsOverlay: showMonthPickerAsOverlay, showWeekNumbers: showWeekNumbers }))),
                timeElm),
            React.createElement(ErrorMessage_1.default, { errorMessage: this.state.errorMessage })));
    };
    return DateTimePicker;
}(React.Component));
exports.DateTimePicker = DateTimePicker;
