// office-ui-fabric-react/lib-amd/components/Pivot/Pivot.base.js
define(["require", "exports", "tslib", "react", "../../Utilities", "../../Button", "../../FocusZone", "./PivotItem", "./Pivot.types", "./Pivot.types", "../../Icon", "./Pivot.scss"], function (require, exports, tslib_1, React, Utilities_1, Button_1, FocusZone_1, PivotItem_1, Pivot_types_1, Pivot_types_2, Icon_1, stylesImport) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var styles = stylesImport;
    var PivotItemType = (React.createElement(PivotItem_1.PivotItem, null)).type;
    var PivotBase = /** @class */ (function (_super) {
        tslib_1.__extends(PivotBase, _super);
        function PivotBase(props) {
            var _this = _super.call(this, props) || this;
            _this.focusZone = Utilities_1.createRef();
            _this._renderPivotLink = function (link) {
                var itemKey = link.itemKey, headerButtonProps = link.headerButtonProps;
                var tabId = _this._keyToTabIds[itemKey];
                var onRenderItemLink = link.onRenderItemLink;
                var linkContent;
                if (onRenderItemLink) {
                    linkContent = onRenderItemLink(link, _this._renderLinkContent);
                }
                else {
                    linkContent = _this._renderLinkContent(link);
                }
                return (React.createElement(Button_1.CommandButton, tslib_1.__assign({}, headerButtonProps, { id: tabId, key: itemKey, className: Utilities_1.css('ms-Pivot-link', styles.link, (_a = {},
                        _a['is-selected ' + styles.linkIsSelected] = _this.state.selectedKey === itemKey,
                        _a)), onClick: _this._onLinkClick.bind(_this, itemKey), onKeyPress: _this._onKeyPress.bind(_this, itemKey), ariaLabel: link.ariaLabel, role: 'tab', "aria-selected": _this.state.selectedKey === itemKey, name: link.headerText, keytipProps: link.keytipProps }), linkContent));
                var _a;
            };
            _this._renderLinkContent = function (link) {
                var itemCount = link.itemCount, itemIcon = link.itemIcon, headerText = link.headerText;
                return (React.createElement("span", { className: Utilities_1.css('ms-Pivot-link-content') },
                    itemIcon !== undefined && (React.createElement("span", { className: Utilities_1.css('ms-Pivot-icon', styles.icon) },
                        React.createElement(Icon_1.Icon, { iconName: itemIcon }))),
                    headerText !== undefined && React.createElement("span", { className: Utilities_1.css('ms-Pivot-text', styles.text) },
                        " ",
                        link.headerText),
                    itemCount !== undefined && React.createElement("span", { className: Utilities_1.css('ms-Pivot-count', styles.count) },
                        " (",
                        itemCount,
                        ")")));
            };
            _this._pivotId = Utilities_1.getId('Pivot');
            var links = _this._getPivotLinks(_this.props);
            var selectedKey;
            if (props.initialSelectedKey) {
                selectedKey = props.initialSelectedKey;
            }
            else if (props.initialSelectedIndex) {
                selectedKey = links[props.initialSelectedIndex].itemKey;
            }
            else if (props.selectedKey) {
                selectedKey = props.selectedKey;
            }
            else if (links.length) {
                selectedKey = links[0].itemKey;
            }
            _this.state = {
                links: links,
                selectedKey: selectedKey,
                selectedTabId: _this._keyToTabIds[selectedKey],
            };
            _this._renderPivotLink = _this._renderPivotLink.bind(_this);
            return _this;
        }
        PivotBase.prototype.componentWillReceiveProps = function (nextProps) {
            var _this = this;
            var links = this._getPivotLinks(nextProps);
            this.setState(function (prevState, props) {
                var selectedKey;
                if (_this._isKeyValid(nextProps.selectedKey)) {
                    selectedKey = nextProps.selectedKey;
                }
                else if (_this._isKeyValid(prevState.selectedKey)) {
                    selectedKey = prevState.selectedKey;
                }
                else if (links.length) {
                    selectedKey = links[0].itemKey;
                }
                return {
                    links: links,
                    selectedKey: selectedKey,
                    selectedTabId: _this._keyToTabIds[selectedKey],
                };
            });
        };
        /**
         * Sets focus to the first pivot tab.
         */
        PivotBase.prototype.focus = function () {
            if (this.focusZone.current) {
                this.focusZone.current.focus();
            }
        };
        PivotBase.prototype.render = function () {
            var divProps = Utilities_1.getNativeProps(this.props, Utilities_1.divProperties);
            return (React.createElement("div", tslib_1.__assign({}, divProps),
                this._renderPivotLinks(),
                this._renderPivotItem()));
        };
        /**
         * Renders the set of links to route between pivots
         */
        PivotBase.prototype._renderPivotLinks = function () {
            return (React.createElement(FocusZone_1.FocusZone, { componentRef: this.focusZone, direction: FocusZone_1.FocusZoneDirection.horizontal },
                React.createElement("ul", { className: Utilities_1.css('ms-Pivot', styles.root, (_a = {}, _a['ms-Pivot--large ' + styles.rootIsLarge] = this.props.linkSize === Pivot_types_2.PivotLinkSize.large, _a), (_b = {}, _b['ms-Pivot--tabs ' + styles.rootIsTabs] = this.props.linkFormat === Pivot_types_1.PivotLinkFormat.tabs, _b)), role: 'tablist' }, this.state.links.map(this._renderPivotLink))));
            var _a, _b;
        };
        /**
         * Renders the current Pivot Item
         */
        PivotBase.prototype._renderPivotItem = function () {
            if (this.props.headersOnly) {
                return null;
            }
            var itemKey = this.state.selectedKey;
            var index = this._keyToIndexMapping[itemKey];
            var selectedTabId = this.state.selectedTabId;
            return (React.createElement("div", { role: 'tabpanel', "aria-labelledby": selectedTabId }, React.Children.toArray(this.props.children)[index]));
        };
        /**
         * Gets the set of PivotLinks as arrary of IPivotItemProps
         * The set of Links is determined by child components of type PivotItem
         */
        PivotBase.prototype._getPivotLinks = function (props) {
            var _this = this;
            var links = [];
            this._keyToIndexMapping = {};
            this._keyToTabIds = {};
            React.Children.map(props.children, function (child, index) {
                if (typeof child === 'object' && child.type === PivotItemType) {
                    var pivotItem = child;
                    var itemKey = pivotItem.props.itemKey || index.toString();
                    links.push({
                        headerText: pivotItem.props.headerText || pivotItem.props.linkText,
                        headerButtonProps: pivotItem.props.headerButtonProps,
                        ariaLabel: pivotItem.props.ariaLabel,
                        itemKey: itemKey,
                        itemCount: pivotItem.props.itemCount,
                        itemIcon: pivotItem.props.itemIcon,
                        onRenderItemLink: pivotItem.props.onRenderItemLink,
                        keytipProps: pivotItem.props.keytipProps
                    });
                    _this._keyToIndexMapping[itemKey] = index;
                    _this._keyToTabIds[itemKey] = _this._getTabId(itemKey, index);
                }
            });
            return links;
        };
        /**
         * Generates the Id for the tab button.
         */
        PivotBase.prototype._getTabId = function (itemKey, index) {
            if (this.props.getTabId) {
                return this.props.getTabId(itemKey, index);
            }
            return this._pivotId + ("-Tab" + index);
        };
        /**
         * whether the key exists in the pivot items.
         */
        PivotBase.prototype._isKeyValid = function (itemKey) {
            return itemKey !== undefined && this._keyToIndexMapping[itemKey] !== undefined;
        };
        /**
         * Handles the onClick event on PivotLinks
         */
        PivotBase.prototype._onLinkClick = function (itemKey, ev) {
            ev.preventDefault();
            this._updateSelectedItem(itemKey, ev);
        };
        /**
         * Handle the onKeyPress eventon the PivotLinks
         */
        PivotBase.prototype._onKeyPress = function (itemKey, ev) {
            ev.preventDefault();
            if (ev.which === 13 /* enter */) {
                this._updateSelectedItem(itemKey);
            }
        };
        /**
         * Updates the state with the new selected index
         */
        PivotBase.prototype._updateSelectedItem = function (itemKey, ev) {
            this.setState({
                selectedKey: itemKey,
                selectedTabId: this._keyToTabIds[itemKey]
            });
            if (this.props.onLinkClick && this._keyToIndexMapping[itemKey] >= 0) {
                var index = this._keyToIndexMapping[itemKey];
                // React.Element<any> cannot directly convert to PivotItem.
                var item = React.Children.toArray(this.props.children)[index];
                if (typeof item === 'object' && item.type === PivotItem_1.PivotItem) {
                    this.props.onLinkClick(item, ev);
                }
            }
        };
        return PivotBase;
    }(Utilities_1.BaseComponent));
    exports.PivotBase = PivotBase;
});
