// office-ui-fabric-react/lib/components/Shimmer/examples/Shimmer.Styling.Example.js
// @codepen
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Shimmer } from 'office-ui-fabric-react/lib/Shimmer';
import { mergeStyles } from 'office-ui-fabric-react/lib/Styling';
var wrapperClass = mergeStyles({
    padding: 2,
    selectors: {
        '& > *': {
            margin: '10px 0'
        }
    }
});
var ShimmerStylingExample = /** @class */ (function (_super) {
    tslib_1.__extends(ShimmerStylingExample, _super);
    function ShimmerStylingExample() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._getShimmerStyles = function (props) {
            return {
                shimmerWrapper: [
                    {
                        backgroundColor: '#deecf9',
                        backgroundImage: 'linear-gradient(to right, rgba(255, 255, 255, 0) 0%, #c7e0f4 50%, rgba(255, 255, 255, 0) 100%)'
                    }
                ]
            };
        };
        return _this;
    }
    ShimmerStylingExample.prototype.render = function () {
        return (React.createElement("div", { className: wrapperClass },
            React.createElement(Shimmer, { width: "75%", styles: this._getShimmerStyles }),
            React.createElement(Shimmer, { width: "75%", styles: this._getShimmerStyles }),
            React.createElement(Shimmer, { width: "75%", styles: this._getShimmerStyles }),
            React.createElement(Shimmer, { width: "75%", styles: this._getShimmerStyles }),
            React.createElement(Shimmer, { width: "75%", styles: this._getShimmerStyles })));
    };
    return ShimmerStylingExample;
}(React.Component));
export { ShimmerStylingExample };
