// @uifabric/fluent-theme/lib/fluent/styles/TeachingBubble.styles.js
import { FontSizes } from '../FluentType';
import { FontWeights } from 'office-ui-fabric-react/lib/Styling';
export var TeachingBubbleStyles = function (props) {
    var hasCondensedHeadline = props.hasCondensedHeadline, hasSmallHeadline = props.hasSmallHeadline, theme = props.theme;
    var effects = theme.effects, palette = theme.palette;
    var headlineSize = FontSizes.size14;
    if (!hasCondensedHeadline && !hasSmallHeadline) {
        headlineSize = FontSizes.size20;
    }
    return {
        headline: {
            fontSize: headlineSize,
            fontWeight: FontWeights.semibold
        },
        footer: {
            selectors: {
                '.ms-Button:not(:first-child)': {
                    marginLeft: '16px'
                }
            }
        },
        closeButton: {
            backgroundColor: 'transparent',
            selectors: {
                '&:hover': {
                    color: palette.black
                },
                '&:active': {
                    background: palette.themeDarkAlt,
                    color: palette.black
                }
            }
        },
        subComponentStyles: {
            callout: {
                root: {
                    boxShadow: effects.elevation16
                }
            }
        }
    };
};
