// @microsoft/office-ui-fabric-react-bundle/lib/office-ui-fabric-react-bundle.js
import { initializeIcons } from '@uifabric/icons';
if (DATACENTER) {
    initializeIcons(); 
}
else {
    initializeIcons('/_layouts/15/next/fabric/icons/');
}
export * from 'office-ui-fabric-react/lib/Button'; 
export * from 'office-ui-fabric-react/lib/Callout'; 
export * from 'office-ui-fabric-react/lib/CommandBar'; 
export * from 'office-ui-fabric-react/lib/ContextualMenu'; 
export * from 'office-ui-fabric-react/lib/Dialog'; 
export * from 'office-ui-fabric-react/lib/Fabric'; 
export * from 'office-ui-fabric-react/lib/FocusTrapZone'; 
export * from 'office-ui-fabric-react/lib/FocusZone'; 
export * from 'office-ui-fabric-react/lib/Icon'; 
export * from 'office-ui-fabric-react/lib/Image'; 
export * from 'office-ui-fabric-react/lib/Label'; 
export * from 'office-ui-fabric-react/lib/Link'; 
export * from 'office-ui-fabric-react/lib/List'; 
export * from 'office-ui-fabric-react/lib/MessageBar'; 
export * from 'office-ui-fabric-react/lib/Nav'; 
export * from 'office-ui-fabric-react/lib/Overlay'; 
export * from 'office-ui-fabric-react/lib/Panel'; 
export * from 'office-ui-fabric-react/lib/Persona'; 
export * from 'office-ui-fabric-react/lib/Spinner'; 
export * from 'office-ui-fabric-react/lib/Styling'; 
export * from 'office-ui-fabric-react/lib/TextField'; 
export * from 'office-ui-fabric-react/lib/Utilities'; 
import * as withResponsiveMode from 'office-ui-fabric-react/lib/utilities/decorators/withResponsiveMode';
export var workaround_withResponsiveMode = withResponsiveMode;
import * as withViewport from 'office-ui-fabric-react/lib/utilities/decorators/withViewport';
export var workaround_withViewport = withViewport;
import * as Icon from 'office-ui-fabric-react/lib/Icon';
export var workaround_Icon = Icon;
import * as IconProps from 'office-ui-fabric-react/lib/components/Icon/Icon.types';
export var workaround_IconProps = IconProps;
import * as IconComponent from 'office-ui-fabric-react/lib/components/Icon';
export var workaround_IconComponent = IconComponent;
import * as IconIndex from 'office-ui-fabric-react/lib/components/Icon/index';
export var workaround_IconIndex = IconIndex;
import * as ImageProps from 'office-ui-fabric-react/lib/components/Image/Image.types';
export var workaround_ImageProps = ImageProps;
import * as ImageComponent from 'office-ui-fabric-react/lib/components/Image/Image';
export var workaround_ImageComponent = ImageComponent;
import * as Check from 'office-ui-fabric-react/lib/Check';
export var workaround_Check = Check;
import * as CheckComponent from 'office-ui-fabric-react/lib/components/Check/Check';
export var workaround_CheckComponent = CheckComponent;
import * as CheckIndex from 'office-ui-fabric-react/lib/components/Check/index';
export var workaround_CheckIndex = CheckIndex;
export var workaround_CheckScss = require('office-ui-fabric-react/lib/components/Check/Check.scss.js');
import * as Popup from 'office-ui-fabric-react/lib/Popup';
export var workaround_Popup = Popup;
import * as PopupComponent from 'office-ui-fabric-react/lib/components/Popup/Popup';
export var workaround_PopupComponent = PopupComponent;
import * as PopupIndex from 'office-ui-fabric-react/lib/components/Popup/index';
export var workaround_PopupIndex = PopupIndex;
