// office-ui-fabric-react/lib-amd/components/DetailsList/DetailsRow.styles.js
define(["require", "exports", "tslib", "../../Styling", "../../Utilities"], function (require, exports, tslib_1, Styling_1, Utilities_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var GlobalClassNames = {
        root: 'ms-DetailsRow',
        compact: 'ms-DetailsList--Compact',
        cell: 'ms-DetailsRow-cell',
        cellCheck: 'ms-DetailsRow-cellCheck',
        cellMeasurer: 'ms-DetailsRow-cellMeasurer',
        listCellFirstChild: 'ms-List-cell:first-child',
        isFocusable: "[data-is-focusable='true']",
        isContentUnselectable: 'is-contentUnselectable',
        isSelected: 'is-selected',
        isCheckVisible: 'is-check-visible',
        isRowHeader: 'is-row-header',
        fields: 'ms-DetailsRow-fields'
    };
    exports.DEFAULT_CELL_STYLE_PROPS = {
        cellLeftPadding: 12,
        cellRightPadding: 8,
        cellExtraRightPadding: 24
    };
    // Source of default row heights to share.
    exports.DEFAULT_ROW_HEIGHTS = {
        rowHeight: 42,
        compactRowHeight: 32
    };
    // Constant values
    var values = tslib_1.__assign({}, exports.DEFAULT_ROW_HEIGHTS, { rowVerticalPadding: 11, compactRowVerticalPadding: 6 });
    exports.getStyles = function (props) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
        var theme = props.theme, isSelected = props.isSelected, canSelect = props.canSelect, droppingClassName = props.droppingClassName, anySelected = props.anySelected, isCheckVisible = props.isCheckVisible, checkboxCellClassName = props.checkboxCellClassName, compact = props.compact, className = props.className, _o = props.cellStyleProps, cellStyleProps = _o === void 0 ? exports.DEFAULT_CELL_STYLE_PROPS : _o;
        var _p = theme.palette, neutralPrimary = _p.neutralPrimary, white = _p.white, neutralSecondary = _p.neutralSecondary, neutralLighter = _p.neutralLighter, neutralLight = _p.neutralLight, neutralDark = _p.neutralDark, neutralQuaternaryAlt = _p.neutralQuaternaryAlt;
        var focusBorder = theme.semanticColors.focusBorder;
        var classNames = Styling_1.getGlobalClassNames(GlobalClassNames, theme);
        var colors = {
            // Default
            defaultHeaderText: neutralPrimary,
            defaultMetaText: neutralSecondary,
            defaultBackground: white,
            // Default Hover
            defaultHoverHeaderText: neutralDark,
            defaultHoverMetaText: neutralPrimary,
            defaultHoverBackground: neutralLighter,
            // Selected
            selectedHeaderText: neutralDark,
            selectedMetaText: neutralPrimary,
            selectedBackground: neutralLight,
            // Selected Hover
            selectedHoverHeaderText: neutralDark,
            selectedHoverMetaText: neutralPrimary,
            selectedHoverBackground: neutralQuaternaryAlt,
            // Focus
            focusHeaderText: neutralDark,
            focusMetaText: neutralPrimary,
            focusBackground: neutralLight,
            focusHoverBackground: neutralQuaternaryAlt
        };
        // Selected row styles
        var selectedStyles = [
            Styling_1.getFocusStyle(theme, { inset: -1, borderColor: focusBorder, outlineColor: white }),
            classNames.isSelected,
            {
                color: colors.selectedMetaText,
                background: colors.selectedBackground,
                borderBottom: "1px solid " + white,
                selectors: (_a = {
                        '&:before': {
                            position: 'absolute',
                            display: 'block',
                            top: -1,
                            height: 1,
                            bottom: 0,
                            left: 0,
                            right: 0,
                            content: '',
                            borderTop: "1px solid " + white
                        },
                        // Selected State hover
                        '&:hover': {
                            background: colors.selectedHoverBackground,
                            color: colors.selectedHoverMetaText,
                            selectors: (_b = {},
                                // Selected State hover meta cell
                                _b["." + classNames.cell] = {
                                    selectors: (_c = {},
                                        _c[Styling_1.HighContrastSelector] = {
                                            color: 'HighlightText',
                                            selectors: {
                                                '> a': {
                                                    color: 'HighlightText'
                                                }
                                            }
                                        },
                                        _c)
                                },
                                // Selected State hover Header cell
                                _b["." + classNames.isRowHeader] = {
                                    color: colors.selectedHoverHeaderText,
                                    selectors: (_d = {},
                                        _d[Styling_1.HighContrastSelector] = {
                                            color: 'HighlightText'
                                        },
                                        _d)
                                },
                                // Ensure high-contrast mode overrides default hover background
                                _b[Styling_1.HighContrastSelector] = {
                                    background: 'Highlight'
                                },
                                _b)
                        },
                        // Focus state
                        '&:focus': {
                            background: colors.focusBackground,
                            selectors: (_e = {},
                                // Selected State hover meta cell
                                _e["." + classNames.cell] = {
                                    color: colors.focusMetaText,
                                    selectors: (_f = {},
                                        _f[Styling_1.HighContrastSelector] = {
                                            color: 'HighlightText',
                                            selectors: {
                                                '> a': {
                                                    color: 'HighlightText'
                                                }
                                            }
                                        },
                                        _f)
                                },
                                // Row header cell
                                _e["." + classNames.isRowHeader] = {
                                    color: colors.focusHeaderText,
                                    selectors: (_g = {},
                                        _g[Styling_1.HighContrastSelector] = {
                                            color: 'HighlightText'
                                        },
                                        _g)
                                },
                                // Ensure high-contrast mode overrides default focus background
                                _e[Styling_1.HighContrastSelector] = {
                                    background: 'Highlight'
                                },
                                _e)
                        }
                    },
                    _a[Styling_1.HighContrastSelector] = {
                        background: 'Highlight',
                        color: 'HighlightText',
                        MsHighContrastAdjust: 'none',
                        selectors: {
                            a: {
                                color: 'HighlightText'
                            }
                        }
                    },
                    // Focus and hover state
                    _a['&:focus:hover'] = {
                        background: colors.focusHoverBackground
                    },
                    _a)
            }
        ];
        var cannotSelectStyles = [
            classNames.isContentUnselectable,
            {
                userSelect: 'none',
                cursor: 'default'
            }
        ];
        var rootCompactStyles = {
            minHeight: values.compactRowHeight,
            border: 0
        };
        var cellCompactStyles = {
            minHeight: values.compactRowHeight,
            paddingTop: values.compactRowVerticalPadding,
            paddingBottom: values.compactRowVerticalPadding,
            paddingLeft: cellStyleProps.cellLeftPadding + "px"
        };
        var defaultCellStyles = [
            Styling_1.getFocusStyle(theme, { inset: -1 }),
            classNames.cell,
            {
                display: 'inline-block',
                position: 'relative',
                boxSizing: 'border-box',
                minHeight: values.rowHeight,
                verticalAlign: 'top',
                whiteSpace: 'nowrap',
                overflow: 'hidden',
                textOverflow: 'ellipsis',
                paddingTop: values.rowVerticalPadding,
                paddingBottom: values.rowVerticalPadding,
                paddingLeft: cellStyleProps.cellLeftPadding + "px",
                selectors: (_h = {
                        '& > button': {
                            maxWidth: '100%'
                        }
                    },
                    _h[classNames.isFocusable] = Styling_1.getFocusStyle(theme, { inset: -1, borderColor: neutralSecondary, outlineColor: white }),
                    _h)
            },
            isSelected && {
                selectors: (_j = {},
                    _j[Styling_1.HighContrastSelector] = {
                        background: 'Highlight',
                        color: 'HighlightText',
                        MsHighContrastAdjust: 'none',
                        selectors: {
                            a: {
                                color: 'HighlightText'
                            }
                        }
                    },
                    _j)
            },
            compact && cellCompactStyles
        ];
        return {
            root: [
                classNames.root,
                Styling_1.AnimationClassNames.fadeIn400,
                droppingClassName,
                theme.fonts.small,
                isCheckVisible && classNames.isCheckVisible,
                Styling_1.getFocusStyle(theme, { borderColor: focusBorder, outlineColor: white }),
                {
                    borderBottom: "1px solid " + neutralLighter,
                    background: colors.defaultBackground,
                    color: colors.defaultMetaText,
                    display: 'inline-flex',
                    minWidth: '100%',
                    minHeight: values.rowHeight,
                    whiteSpace: 'nowrap',
                    padding: 0,
                    boxSizing: 'border-box',
                    verticalAlign: 'top',
                    textAlign: 'left',
                    selectors: (_k = {},
                        _k["." + classNames.listCellFirstChild + " &:before"] = {
                            display: 'none'
                        },
                        _k['&:hover'] = {
                            background: colors.defaultHoverBackground,
                            color: colors.defaultHoverMetaText,
                            selectors: (_l = {},
                                _l["." + classNames.isRowHeader] = {
                                    color: colors.defaultHoverHeaderText
                                },
                                _l)
                        },
                        _k['&:hover $check'] = {
                            opacity: 1
                        },
                        _k["." + Utilities_1.IsFocusVisibleClassName + " &:focus $check"] = {
                            opacity: 1
                        },
                        _k)
                },
                isSelected && selectedStyles,
                !canSelect && cannotSelectStyles,
                compact && rootCompactStyles,
                className
            ],
            cellUnpadded: [
                {
                    paddingRight: cellStyleProps.cellRightPadding + "px"
                }
            ],
            cellPadded: [
                {
                    paddingRight: cellStyleProps.cellExtraRightPadding + cellStyleProps.cellRightPadding + "px",
                    selectors: {
                        '&.$checkCell': {
                            paddingRight: 0
                        }
                    }
                }
            ],
            cell: defaultCellStyles,
            cellMeasurer: [
                classNames.cellMeasurer,
                {
                    overflow: 'visible',
                    whiteSpace: 'nowrap'
                }
            ],
            checkCell: [
                defaultCellStyles,
                classNames.cellCheck,
                checkboxCellClassName,
                {
                    padding: 0,
                    // Ensure that the check cell covers the top border of the cell.
                    // This ensures the click target does not leave a spot which would
                    // cause other items to be deselected.
                    paddingTop: 1,
                    marginTop: -1,
                    flexShrink: 0
                }
            ],
            checkCover: [
                {
                    position: 'absolute',
                    top: -1,
                    left: 0,
                    bottom: 0,
                    right: 0,
                    display: 'none'
                },
                anySelected && {
                    display: 'block'
                }
            ],
            fields: [
                classNames.fields,
                {
                    display: 'flex',
                    alignItems: 'stretch'
                }
            ],
            isRowHeader: [
                classNames.isRowHeader,
                {
                    color: colors.defaultHeaderText,
                    fontSize: Styling_1.FontSizes.medium
                },
                isSelected && {
                    color: colors.selectedHeaderText,
                    fontWeight: Styling_1.FontWeights.semibold,
                    selectors: (_m = {},
                        _m[Styling_1.HighContrastSelector] = {
                            color: 'HighlightText'
                        },
                        _m)
                }
            ],
            isMultiline: [
                defaultCellStyles,
                {
                    whiteSpace: 'normal',
                    wordBreak: 'break-word',
                    textOverflow: 'clip'
                }
            ],
            check: []
        };
    };
});
