// @pnp/spfx-controls-react/lib/services/SPService.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var ISPService_1 = require("./ISPService");
var sp_http_1 = require("@microsoft/sp-http");
var SPService = (function () {
    function SPService(_context) {
        this._context = _context;
    }
    /**
     * Get lists or libraries
     *
     * @param options
     */
    SPService.prototype.getLibs = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var filtered, queryUrl, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        queryUrl = this._context.pageContext.web.absoluteUrl + "/_api/web/lists?$select=Title,id,BaseTemplate";
                        if (options.orderBy) {
                            queryUrl += "&$orderby=" + (options.orderBy === ISPService_1.LibsOrderBy.Id ? 'Id' : 'Title');
                        }
                        if (options.baseTemplate) {
                            queryUrl += "&$filter=BaseTemplate eq " + options.baseTemplate;
                            filtered = true;
                        }
                        if (options.includeHidden === false) {
                            queryUrl += filtered ? ' and Hidden eq false' : '&$filter=Hidden eq false';
                            filtered = true;
                        }
                        return [4 /*yield*/, this._context.spHttpClient.get(queryUrl, sp_http_1.SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3: return [2 /*return*/, null];
                }
            });
        });
    };
    /**
     * Get List Items
     */
    SPService.prototype.getListItems = function (filterText, listId, internalColumnName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var returnItems, webAbsoluteUrl, apiUrl, data, results, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/items?$select=Id," + internalColumnName + "&$filter=startswith(" + internalColumnName + ",'" + filterText + "')";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, sp_http_1.SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results && results.value && results.value.length > 0) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, []];
                    case 4:
                        error_1 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_1)];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get list item attachments
     *
     * @param listId
     * @param itemId
     * @param webUrl
     */
    SPService.prototype.getListItemAttachments = function (listId, itemId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/items(" + itemId + ")/AttachmentFiles";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, sp_http_1.SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results && results.value && results.value.length > 0) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                    case 4:
                        error_2 = _a.sent();
                        console.dir(error_2);
                        return [2 /*return*/, Promise.reject(error_2)];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Delete attachment
     *
     * @param fileName
     * @param listId
     * @param itemId
     * @param webUrl
     */
    SPService.prototype.deleteAttachment = function (fileName, listId, itemId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var spOpts, webAbsoluteUrl, apiUrl, data, error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        spOpts = {
                            headers: { "X-HTTP-Method": 'DELETE', }
                        };
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/items(" + itemId + ")/AttachmentFiles/getByFileName('" + encodeURIComponent(fileName) + "')";
                        return [4 /*yield*/, this._context.spHttpClient.post(apiUrl, sp_http_1.SPHttpClient.configurations.v1, spOpts)];
                    case 1:
                        data = _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        error_3 = _a.sent();
                        console.dir(error_3);
                        return [2 /*return*/, Promise.reject(error_3)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Add attachment
     *
     * @param listId
     * @param itemId
     * @param fileName
     * @param file
     * @param webUrl
     */
    SPService.prototype.addAttachment = function (listId, itemId, fileName, file, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var fileExists, spOpts, webAbsoluteUrl, apiUrl, data, error_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, , 6]);
                        // remove special characteres in FileName
                        fileName = fileName.replace(/[^\.\w\s]/gi, '');
                        return [4 /*yield*/, this.checkAttachmentExists(listId, itemId, fileName, webUrl)];
                    case 1:
                        fileExists = _a.sent();
                        if (!fileExists) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.deleteAttachment(fileName, listId, itemId, webUrl)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3:
                        spOpts = {
                            body: file
                        };
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/items(" + itemId + ")/AttachmentFiles/add(FileName='" + encodeURIComponent(fileName) + "')";
                        return [4 /*yield*/, this._context.spHttpClient.post(apiUrl, sp_http_1.SPHttpClient.configurations.v1, spOpts)];
                    case 4:
                        data = _a.sent();
                        return [2 /*return*/];
                    case 5:
                        error_4 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_4)];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get attachement for list item
     *
     * @param listId
     * @param itemId
     * @param fileName
     * @param webUrl
     */
    SPService.prototype.getAttachment = function (listId, itemId, fileName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, file;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/items(" + itemId + ")/AttachmentFiles/GetByFileBame('" + fileName + "'))";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, sp_http_1.SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        file = _a.sent();
                        if (file) {
                            return [2 /*return*/, file];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                }
            });
        });
    };
    /**
     * Check if the attachment exists
     *
     * @param listId
     * @param itemId
     * @param fileName
     * @param webUrl
     */
    SPService.prototype.checkAttachmentExists = function (listId, itemId, fileName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var listName, webAbsoluteUrl, apiUrl, data, results, error_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, , 6]);
                        return [4 /*yield*/, this.getListName(listId, webUrl)];
                    case 1:
                        listName = _a.sent();
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/getfilebyserverrelativeurl('/lists/" + listName + "/Attachments/" + itemId + "/" + fileName + "')";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, sp_http_1.SPHttpClient.configurations.v1)];
                    case 2:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 4];
                        return [4 /*yield*/, data.json()];
                    case 3:
                        results = _a.sent();
                        if (results && results.Exists) {
                            return [2 /*return*/, results.Exists];
                        }
                        _a.label = 4;
                    case 4: return [2 /*return*/, false];
                    case 5:
                        error_5 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_5)];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get the list name
     *
     * @param listId
     * @param webUrl
     */
    SPService.prototype.getListName = function (listId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')?$select=RootFolder/Name&$expand=RootFolder)";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, sp_http_1.SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results) {
                            return [2 /*return*/, results.RootFolder.Name];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return SPService;
}());
exports.default = SPService;
