// @microsoft/sp-http/lib/oauthTokenProvider/AdalAuthContextManager.js
import AdalAuthContext from './AdalAuthContext';
import { Guid, UrlUtilities, _SPKillSwitch } from '@microsoft/sp-core-library';
import { OBOTokenProvider } from './OBOTokenProvider';
var AdalAuthContextManager =  (function () {
    function AdalAuthContextManager() {
        this._loadAdalJsModulePromise = undefined;
        this._authContextDictionary = new Map();
    }
    AdalAuthContextManager.convertTokenParametersToConfig = function (tokenProviderParameters) {
        var extraQueryParameter;
        if (tokenProviderParameters.userEmail &&
            !_SPKillSwitch.isActivated(AdalAuthContextManager.usePromptQueryParamKillSwitchGuid, '12/02/18', 'Use login hint to resolve ambiguous users failure')) {
            extraQueryParameter = "login_hint=" + encodeURIComponent(tokenProviderParameters.userEmail);
        }
        return {
            clientId: tokenProviderParameters.servicePrincipalId,
            redirectUri: tokenProviderParameters.redirectUri,
            instance: UrlUtilities.removeEndSlash(tokenProviderParameters.aadInstanceUrl) + '/',
            loadFrameTimeout: 10000,
            tenant: tokenProviderParameters.aadTenantId,
            navigateToLoginRequestUrl: false,
            extraQueryParameter: extraQueryParameter
        };
    };
    AdalAuthContextManager.prototype.getAuthContext = function (tokenProviderConfiguration, sharePointOBOProviderConfiguration) {
        var _this = this;
        if (!this._loadAdalJsModulePromise) {
            this._loadAdalJsModulePromise = this._loadAdalJs();
        }
        return this._loadAdalJsModulePromise.then(function (adalModule) {
            if (sharePointOBOProviderConfiguration) {
                return new OBOTokenProvider(adalModule.inject(AdalAuthContextManager.convertTokenParametersToConfig(tokenProviderConfiguration)), sharePointOBOProviderConfiguration);
            }
            if (!_this._authContextDictionary.has(tokenProviderConfiguration.servicePrincipalId)) {
                var authContext = new AdalAuthContext(adalModule.inject(AdalAuthContextManager.convertTokenParametersToConfig(tokenProviderConfiguration)), tokenProviderConfiguration.aadUserId);
                _this._authContextDictionary.set(tokenProviderConfiguration.servicePrincipalId, authContext);
            }
            return _this._authContextDictionary.get(tokenProviderConfiguration.servicePrincipalId);
        });
    };
    AdalAuthContextManager.prototype._loadAdalJs = function () {
        return import( 'adal-angular');
    };
    AdalAuthContextManager.usePromptQueryParamKillSwitchGuid = Guid.parse('9a8c4bb8-88c1-4083-ae41-044320eba9d7');
    return AdalAuthContextManager;
}());
export default AdalAuthContextManager;
