// @pnp/spfx-controls-react/lib/controls/peoplepicker/PeoplePickerComponent.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var strings = require("ControlStrings");
var React = require("react");
var telemetry = require("../../common/telemetry");
var PeoplePickerComponent_module_scss_1 = require("./PeoplePickerComponent.module.scss");
var PeopleSearchService_1 = require("../../services/PeopleSearchService");
var Tooltip_1 = require("office-ui-fabric-react/lib/Tooltip");
var PeoplePicker_1 = require("office-ui-fabric-react/lib/components/pickers/PeoplePicker/PeoplePicker");
var Label_1 = require("office-ui-fabric-react/lib/components/Label");
var Icon_1 = require("office-ui-fabric-react/lib/components/Icon");
var sp_lodash_subset_1 = require("@microsoft/sp-lodash-subset");
/**
 * PeoplePicker component
 */
var PeoplePicker = (function (_super) {
    __extends(PeoplePicker, _super);
    function PeoplePicker(props) {
        var _this = _super.call(this, props) || this;
        /**
         * A search field change occured
         */
        _this.onSearchFieldChanged = function (searchText, currentSelected) { return __awaiter(_this, void 0, void 0, function () {
            var results, _a, selectedPersons, mostRecentlyUsedPersons, filteredPersons, recentlyUsed;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(searchText.length > 2)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.peopleSearchService.searchPeople(searchText, this.suggestionsLimit, this.props.principalTypes, this.props.webAbsoluteUrl, this.groupId, this.props.ensureUser)];
                    case 1:
                        results = _b.sent();
                        _a = this.state, selectedPersons = _a.selectedPersons, mostRecentlyUsedPersons = _a.mostRecentlyUsedPersons;
                        filteredPersons = this.removeDuplicates(results, selectedPersons);
                        recentlyUsed = filteredPersons.concat(mostRecentlyUsedPersons);
                        recentlyUsed = sp_lodash_subset_1.uniqBy(recentlyUsed, "text");
                        this.setState({
                            mostRecentlyUsedPersons: recentlyUsed.slice(0, this.suggestionsLimit)
                        });
                        return [2 /*return*/, filteredPersons];
                    case 2: return [2 /*return*/, []];
                }
            });
        }); };
        /**
         * On item selection change event
         */
        _this.onChange = function (items) {
            var triggerUpdate = _this.props.selectedItems;
            _this.setState({
                selectedPersons: items,
                showRequiredError: items.length > 0 ? false : true
            });
            if (triggerUpdate) {
                triggerUpdate(items);
            }
        };
        /**
         * Returns the most recently used person
         *
         * @param currentPersonas
         */
        _this.returnMostRecentlyUsedPerson = function (currentPersonas) {
            var mostRecentlyUsedPersons = _this.state.mostRecentlyUsedPersons;
            return _this.removeDuplicates(mostRecentlyUsedPersons, currentPersonas);
        };
        /**
         * Removes duplicates
         *
         * @param personas
         * @param possibleDupes
         */
        _this.removeDuplicates = function (personas, possibleDupes) {
            return personas.filter(function (persona) { return !_this.listContainsPersona(persona, possibleDupes); });
        };
        /**
         * Checks if list contains the person
         *
         * @param persona
         * @param personas
         */
        _this.listContainsPersona = function (persona, personas) {
            if (!personas || !personas.length || personas.length === 0) {
                return false;
            }
            return personas.filter(function (item) { return item.text === persona.text; }).length > 0;
        };
        _this.peopleSearchService = new PeopleSearchService_1.default(props.context);
        _this.suggestionsLimit = _this.props.suggestionsLimit ? _this.props.suggestionsLimit : 5;
        telemetry.track('ReactPeoplePicker', {
            groupName: !!props.groupName,
            name: !!props.groupName,
            titleText: !!props.titleText
        });
        _this.state = {
            selectedPersons: [],
            mostRecentlyUsedPersons: [],
            showRequiredError: false,
            resolveDelay: _this.props.resolveDelay || 200,
            errorMessage: null
        };
        return _this;
    }
    /**
     * componentWillMount lifecycle hook
     */
    PeoplePicker.prototype.componentWillMount = function () {
        this.getInitialPersons(this.props);
    };
    /**
     * componentWillUpdate lifecycle hook
     */
    PeoplePicker.prototype.componentWillUpdate = function (nextProps, nextState) {
        if (!sp_lodash_subset_1.isEqual(this.props.defaultSelectedUsers, nextProps.defaultSelectedUsers) ||
            this.props.groupName !== nextProps.groupName ||
            this.props.webAbsoluteUrl !== nextProps.webAbsoluteUrl ||
            this.peopleSearchService.getSumOfPrincipalTypes(this.props.principalTypes) !== this.peopleSearchService.getSumOfPrincipalTypes(nextProps.principalTypes)) {
            this.getInitialPersons(nextProps);
        }
    };
    /**
     * Get initial persons
     */
    PeoplePicker.prototype.getInitialPersons = function (props) {
        return __awaiter(this, void 0, void 0, function () {
            var groupName, webAbsoluteUrl, defaultSelectedUsers, ensureUser, principalTypes, _a, selectedPersons, _i, _b, userValue, userResult;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        groupName = props.groupName, webAbsoluteUrl = props.webAbsoluteUrl, defaultSelectedUsers = props.defaultSelectedUsers, ensureUser = props.ensureUser, principalTypes = props.principalTypes;
                        if (!groupName) return [3 /*break*/, 2];
                        _a = this;
                        return [4 /*yield*/, this.peopleSearchService.getGroupId(groupName, webAbsoluteUrl)];
                    case 1:
                        _a.groupId = _c.sent();
                        if (!this.groupId) {
                            this.setState({
                                errorMessage: strings.PeoplePickerGroupNotFound
                            });
                            return [2 /*return*/];
                        }
                        return [3 /*break*/, 3];
                    case 2:
                        this.groupId = null;
                        _c.label = 3;
                    case 3:
                        if (!defaultSelectedUsers) return [3 /*break*/, 8];
                        selectedPersons = [];
                        _i = 0, _b = props.defaultSelectedUsers;
                        _c.label = 4;
                    case 4:
                        if (!(_i < _b.length)) return [3 /*break*/, 7];
                        userValue = _b[_i];
                        return [4 /*yield*/, this.peopleSearchService.searchPersonByEmailOrLogin(userValue, principalTypes, webAbsoluteUrl, this.groupId, ensureUser)];
                    case 5:
                        userResult = _c.sent();
                        if (userResult) {
                            selectedPersons.push(userResult);
                        }
                        _c.label = 6;
                    case 6:
                        _i++;
                        return [3 /*break*/, 4];
                    case 7:
                        this.setState({
                            selectedPersons: selectedPersons
                        });
                        _c.label = 8;
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Default React component render method
     */
    PeoplePicker.prototype.render = function () {
        var suggestionProps = {
            suggestionsHeaderText: strings.peoplePickerSuggestionsHeaderText,
            noResultsFoundText: strings.genericNoResultsFoundText,
            loadingText: strings.peoplePickerLoadingText,
            resultsMaximumNumber: this.props.suggestionsLimit ? this.props.suggestionsLimit : 5,
            searchingText: strings.PeoplePickerSearchText
        };
        var peoplepicker = (React.createElement("div", { id: "people", className: PeoplePickerComponent_module_scss_1.default.defaultClass + " " + (this.props.peoplePickerWPclassName ? this.props.peoplePickerWPclassName : '') },
            this.props.titleText && React.createElement(Label_1.Label, null, this.props.titleText),
            React.createElement(PeoplePicker_1.NormalPeoplePicker, { pickerSuggestionsProps: suggestionProps, onResolveSuggestions: this.onSearchFieldChanged, onEmptyInputFocus: this.returnMostRecentlyUsedPerson, getTextFromItem: function (peoplePersonaMenu) { return peoplePersonaMenu.text; }, className: "ms-PeoplePicker " + (this.props.peoplePickerCntrlclassName ? this.props.peoplePickerCntrlclassName : ''), key: 'normal', removeButtonAriaLabel: 'Remove', inputProps: {
                    'aria-label': 'People Picker'
                }, selectedItems: this.state.selectedPersons, itemLimit: this.props.personSelectionLimit || 1, disabled: this.props.disabled || !!this.state.errorMessage, onChange: this.onChange, resolveDelay: this.state.resolveDelay })));
        return (React.createElement("div", null,
            this.props.showtooltip ? (React.createElement(Tooltip_1.TooltipHost, { content: this.props.tooltipMessage || strings.peoplePickerComponentTooltipMessage, id: 'pntp', calloutProps: { gapSpace: 0 }, directionalHint: this.props.tooltipDirectional || 8 /* leftTopEdge */ }, peoplepicker)) : (React.createElement("div", null, peoplepicker)),
            (this.props.isRequired && this.state.showRequiredError) || (this.state.errorMessage) && (React.createElement("p", { className: "ms-TextField-errorMessage " + PeoplePickerComponent_module_scss_1.default.errorMessage + " " + (this.props.errorMessageClassName ? this.props.errorMessageClassName : '') },
                React.createElement(Icon_1.Icon, { iconName: 'Error', className: PeoplePickerComponent_module_scss_1.default.errorIcon }),
                this.state.errorMessage && React.createElement("span", { "data-automation-id": "error-message" }, this.state.errorMessage),
                (this.props.isRequired && this.state.showRequiredError) && React.createElement("span", { "data-automation-id": "error-message" }, this.props.errorMessage ? this.props.errorMessage : strings.peoplePickerComponentErrorMessage)))));
    };
    return PeoplePicker;
}(React.Component));
exports.PeoplePicker = PeoplePicker;
