// office-ui-fabric-react/lib-amd/components/Rating/examples/Rating.Basic.Example.js
define(["require", "exports", "tslib", "react", "office-ui-fabric-react/lib/Rating", "../../../Styling", "./Rating.Basic.Example.scss"], function (require, exports, tslib_1, React, Rating_1, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var RatingBasicExample = /** @class */ (function (_super) {
        tslib_1.__extends(RatingBasicExample, _super);
        function RatingBasicExample(props) {
            var _this = _super.call(this, props) || this;
            _this._onLargeStarChanged = function (rating) {
                _this.setState({
                    largeStarRating: rating
                });
            };
            _this._onSmallStarChanged = function (rating) {
                _this.setState({
                    smallStarRating: rating
                });
            };
            _this._onTenStarChanged = function (rating) {
                _this.setState({
                    tenStarRating: rating
                });
            };
            _this._onThemedStarChanged = function (rating) {
                _this.setState({
                    themedStarRating: rating
                });
            };
            _this.state = {
                largeStarRating: undefined,
                smallStarRating: 3,
                tenStarRating: undefined,
                themedStarRating: undefined
            };
            _this._customTheme = Styling_1.createTheme(Styling_1.getTheme());
            _this._customTheme.semanticColors.bodySubtext = '#DFDFDF';
            _this._customTheme.semanticColors.bodyTextChecked = '#1E9FE8';
            return _this;
        }
        // tslint:disable:jsx-no-lambda
        RatingBasicExample.prototype.render = function () {
            return (React.createElement("div", { className: 'ms-RatingBasicExample' },
                "Large Stars:",
                React.createElement(Rating_1.Rating, { id: 'largeRatingStar', min: 1, max: 5, size: Rating_1.RatingSize.Large, rating: this.state.largeStarRating, getAriaLabel: this._getRatingComponentAriaLabel, onChanged: this._onLargeStarChanged, onFocus: function () { return console.log('onFocus called'); }, onBlur: function () { return console.log('onBlur called'); }, ariaLabelFormat: '{0} of {1} stars selected' }),
                "Small Stars",
                React.createElement(Rating_1.Rating, { id: 'smallRatingStar', min: 1, max: 5, rating: this.state.smallStarRating, onChanged: this._onSmallStarChanged, getAriaLabel: this._getRatingComponentAriaLabel, onFocus: function () { return console.log('onFocus called'); }, onBlur: function () { return console.log('onBlur called'); }, ariaLabelFormat: '{0} of {1} stars selected' }),
                "10 Small Stars",
                React.createElement(Rating_1.Rating, { id: 'tenRatingStar', min: 1, max: 10, rating: this.state.tenStarRating, onChanged: this._onTenStarChanged, getAriaLabel: this._getRatingComponentAriaLabel, onFocus: function () { return console.log('onFocus called'); }, onBlur: function () { return console.log('onBlur called'); }, ariaLabelFormat: '{0} of {1} stars selected' }),
                "Disabled:",
                React.createElement(Rating_1.Rating, { min: 1, max: 5, rating: this.state.rating, disabled: true, onFocus: function () { return console.log('onFocus called'); }, onBlur: function () { return console.log('onBlur called'); }, ariaLabelFormat: '{0} of {1} stars selected' }),
                "Half star in readOnly mode:",
                React.createElement(Rating_1.Rating, { id: 'readOnlyRatingStar', min: 1, max: 5, rating: 2.5, getAriaLabel: this._getRatingComponentAriaLabel, readOnly: true, ariaLabelFormat: '{0} of {1} stars selected' }),
                "Themed star",
                React.createElement(Rating_1.Rating, { id: 'themedRatingStar', min: 1, max: 5, rating: this.state.themedStarRating, onChanged: this._onThemedStarChanged, getAriaLabel: this._getRatingComponentAriaLabel, onFocus: function () { return console.log('onFocus called'); }, onBlur: function () { return console.log('onBlur called'); }, ariaLabelFormat: '{0} of {1} stars selected', theme: this._customTheme })));
        };
        RatingBasicExample.prototype._getRatingComponentAriaLabel = function (rating, maxRating) {
            return "Rating value is " + rating + " of " + maxRating;
        };
        return RatingBasicExample;
    }(React.Component));
    exports.RatingBasicExample = RatingBasicExample;
});
