// office-ui-fabric-react/lib/components/Stack/examples/Stack.Horizontal.Grow.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Stack } from '../Stack';
import { DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var HorizontalStackGrowExample = /** @class */ (function (_super) {
    tslib_1.__extends(HorizontalStackGrowExample, _super);
    function HorizontalStackGrowExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HorizontalStackGrowExample.prototype.render = function () {
        var stackStyles = {
            root: {
                background: DefaultPalette.themeTertiary
            }
        };
        var stackItemStyles = {
            root: {
                alignItems: 'center',
                background: DefaultPalette.themePrimary,
                color: DefaultPalette.white,
                display: 'flex',
                height: 50,
                justifyContent: 'center'
            }
        };
        var stackTokens = {
            childrenGap: 5,
            padding: 10
        };
        return (React.createElement(Stack, { horizontal: true, styles: stackStyles, tokens: stackTokens },
            React.createElement(Stack.Item, { grow: 3, styles: stackItemStyles }, "Grow is 3"),
            React.createElement(Stack.Item, { grow: 2, styles: stackItemStyles }, "Grow is 2"),
            React.createElement(Stack.Item, { grow: true, styles: stackItemStyles }, "Grow is 1")));
    };
    return HorizontalStackGrowExample;
}(React.Component));
export { HorizontalStackGrowExample };
