// office-ui-fabric-react/lib/components/Stack/examples/Stack.Vertical.Shrink.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Slider } from 'office-ui-fabric-react/lib/Slider';
import { Stack } from '../Stack';
import { mergeStyles, DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var VerticalStackShrinkExample = /** @class */ (function (_super) {
    tslib_1.__extends(VerticalStackShrinkExample, _super);
    function VerticalStackShrinkExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onHeightChange = function (value) {
            _this.setState({ stackHeight: value });
        };
        _this.state = {
            stackHeight: 100
        };
        return _this;
    }
    VerticalStackShrinkExample.prototype.render = function () {
        var stackStyles = {
            root: {
                background: DefaultPalette.themeTertiary,
                height: this.state.stackHeight + "%",
                overflow: 'hidden'
            }
        };
        var stackItemStyles = {
            root: {
                alignItems: 'center',
                background: DefaultPalette.themePrimary,
                color: DefaultPalette.white,
                display: 'flex',
                justifyContent: 'center',
                overflow: 'hidden'
            }
        };
        var containerStyles = {
            height: 200
        };
        var nonShrinkingStackItemStyles = {
            root: {
                alignItems: 'center',
                background: DefaultPalette.themePrimary,
                color: DefaultPalette.white,
                display: 'flex',
                height: 50,
                justifyContent: 'center',
                overflow: 'hidden'
            }
        };
        var outerStackTokens = { childrenGap: 5 };
        var innerStackTokens = {
            childrenGap: 5,
            padding: 10
        };
        return (React.createElement(Stack, { tokens: outerStackTokens },
            React.createElement(Slider, { label: "Change the stack height to see how child items shrink:", min: 1, max: 100, step: 1, defaultValue: 100, showValue: true, onChange: this._onHeightChange }),
            React.createElement("div", { className: mergeStyles(containerStyles) },
                React.createElement(Stack, { styles: stackStyles, tokens: innerStackTokens },
                    React.createElement(Stack.Item, { grow: true, styles: stackItemStyles }, "I shrink"),
                    React.createElement(Stack.Item, { grow: true, styles: stackItemStyles }, "I shrink"),
                    React.createElement(Stack.Item, { grow: true, disableShrink: true, styles: nonShrinkingStackItemStyles }, "I don't shrink"),
                    React.createElement(Stack.Item, { grow: true, styles: stackItemStyles }, "I shrink")))));
    };
    return VerticalStackShrinkExample;
}(React.Component));
export { VerticalStackShrinkExample };
