// office-ui-fabric-react/lib/components/Stack/examples/Stack.Vertical.Wrap.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Slider } from 'office-ui-fabric-react/lib/Slider';
import { Stack } from '../Stack';
import { DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var VerticalStackWrapExample = /** @class */ (function (_super) {
    tslib_1.__extends(VerticalStackWrapExample, _super);
    function VerticalStackWrapExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onHeightChange = function (value) {
            _this.setState({ stackHeight: value });
        };
        _this.state = {
            stackHeight: 420
        };
        return _this;
    }
    VerticalStackWrapExample.prototype.render = function () {
        var stackHeight = this.state.stackHeight;
        var stackStyles = {
            root: {
                background: DefaultPalette.themeTertiary,
                height: stackHeight,
                selectors: {
                    '& span': {
                        alignItems: 'center',
                        background: DefaultPalette.themePrimary,
                        color: DefaultPalette.white,
                        height: 50,
                        display: 'flex',
                        justifyContent: 'center',
                        width: 50
                    }
                }
            }
        };
        var sectionStackTokens = { childrenGap: 10 };
        var wrapStackTokens = { childrenGap: 20 };
        return (React.createElement(Stack, { tokens: sectionStackTokens },
            React.createElement(Slider, { label: "Change the stack height to see how child items wrap onto multiple columns:", min: 1, max: 420, step: 1, defaultValue: 420, showValue: true, onChange: this._onHeightChange }),
            React.createElement(Stack, { wrap: true, styles: stackStyles, tokens: wrapStackTokens },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2"),
                React.createElement("span", null, "3"),
                React.createElement("span", null, "4"),
                React.createElement("span", null, "5"),
                React.createElement("span", null, "6"))));
    };
    return VerticalStackWrapExample;
}(React.Component));
export { VerticalStackWrapExample };
