// office-ui-fabric-react/lib/components/Stack/examples/Stack.Vertical.HorizontalAlign.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Stack } from '../Stack';
import { DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var VerticalStackHorizontalAlignExample = /** @class */ (function (_super) {
    tslib_1.__extends(VerticalStackHorizontalAlignExample, _super);
    function VerticalStackHorizontalAlignExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    VerticalStackHorizontalAlignExample.prototype.render = function () {
        var stackStyles = {
            root: {
                background: DefaultPalette.themeTertiary,
                selectors: {
                    '> *': {
                        alignItems: 'center',
                        background: DefaultPalette.themePrimary,
                        color: DefaultPalette.white,
                        display: 'flex',
                        height: 50,
                        justifyContent: 'center',
                        width: 50
                    }
                }
            }
        };
        var stackTokens = { childrenGap: 5 };
        return (React.createElement(Stack, { tokens: stackTokens },
            React.createElement("span", null, "Left-aligned"),
            React.createElement(Stack, { horizontalAlign: "start", styles: stackStyles },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2")),
            React.createElement("span", null, "Horizontally centered"),
            React.createElement(Stack, { horizontalAlign: "center", styles: stackStyles },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2")),
            React.createElement("span", null, "Right-aligned"),
            React.createElement(Stack, { horizontalAlign: "end", styles: stackStyles },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2"))));
    };
    return VerticalStackHorizontalAlignExample;
}(React.Component));
export { VerticalStackHorizontalAlignExample };
