// office-ui-fabric-react/lib/components/Stack/examples/Stack.Horizontal.Wrap.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Slider } from 'office-ui-fabric-react/lib/Slider';
import { Stack } from '../Stack';
import { DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var HorizontalStackWrapExample = /** @class */ (function (_super) {
    tslib_1.__extends(HorizontalStackWrapExample, _super);
    function HorizontalStackWrapExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onWidthChange = function (value) {
            _this.setState({ stackWidth: value });
        };
        _this.state = {
            stackWidth: 100
        };
        return _this;
    }
    HorizontalStackWrapExample.prototype.render = function () {
        var stackStyles = {
            root: {
                background: DefaultPalette.themeTertiary,
                width: this.state.stackWidth + "%",
                selectors: {
                    '& span': {
                        alignItems: 'center',
                        background: DefaultPalette.themePrimary,
                        color: DefaultPalette.white,
                        display: 'flex',
                        height: 50,
                        justifyContent: 'center',
                        width: 50
                    }
                }
            }
        };
        var sectionStackTokens = { childrenGap: 10 };
        var wrapStackTokens = { childrenGap: 30 };
        return (React.createElement(Stack, { tokens: sectionStackTokens },
            React.createElement(Slider, { label: "Change the stack width to see how child items wrap onto multiple rows:", min: 1, max: 100, step: 1, defaultValue: 100, showValue: true, onChange: this._onWidthChange }),
            React.createElement(Stack, { horizontal: true, wrap: true, styles: stackStyles, tokens: wrapStackTokens },
                React.createElement("span", null, "1"),
                React.createElement("span", null, "2"),
                React.createElement("span", null, "3"),
                React.createElement("span", null, "4"),
                React.createElement("span", null, "5"),
                React.createElement("span", null, "6"),
                React.createElement("span", null, "7"),
                React.createElement("span", null, "8"),
                React.createElement("span", null, "9"),
                React.createElement("span", null, "10"))));
    };
    return HorizontalStackWrapExample;
}(React.Component));
export { HorizontalStackWrapExample };
