// office-ui-fabric-react/lib/components/Stack/examples/Stack.Horizontal.Shrink.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Slider } from 'office-ui-fabric-react/lib/Slider';
import { Stack } from '../Stack';
import { DefaultPalette } from 'office-ui-fabric-react/lib/Styling';
var HorizontalStackShrinkExample = /** @class */ (function (_super) {
    tslib_1.__extends(HorizontalStackShrinkExample, _super);
    function HorizontalStackShrinkExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onWidthChange = function (value) {
            _this.setState({ stackWidth: value });
        };
        _this.state = {
            stackWidth: 100
        };
        return _this;
    }
    HorizontalStackShrinkExample.prototype.render = function () {
        var stackStyles = {
            root: {
                background: DefaultPalette.themeTertiary,
                overflow: 'hidden',
                width: this.state.stackWidth + "%"
            }
        };
        var stackItemStyles = {
            root: {
                alignItems: 'center',
                background: DefaultPalette.themePrimary,
                color: DefaultPalette.white,
                display: 'flex',
                height: 50,
                justifyContent: 'center',
                overflow: 'hidden'
            }
        };
        var nonShrinkingStackItemStyles = {
            root: {
                alignItems: 'center',
                background: DefaultPalette.themePrimary,
                color: DefaultPalette.white,
                display: 'flex',
                height: 50,
                justifyContent: 'center',
                overflow: 'hidden',
                width: 500
            }
        };
        var outerStackTokens = { childrenGap: 5 };
        var innerStackTokens = {
            childrenGap: 5,
            padding: 10
        };
        return (React.createElement(Stack, { tokens: outerStackTokens },
            React.createElement(Slider, { label: "Change the stack width to see how child items shrink:", min: 1, max: 100, step: 1, defaultValue: 100, showValue: true, onChange: this._onWidthChange }),
            React.createElement(Stack, { horizontal: true, styles: stackStyles, tokens: innerStackTokens },
                React.createElement(Stack.Item, { grow: true, styles: stackItemStyles }, "I shrink"),
                React.createElement(Stack.Item, { grow: true, styles: stackItemStyles }, "I shrink"),
                React.createElement(Stack.Item, { grow: true, disableShrink: true, styles: nonShrinkingStackItemStyles }, "I don't shrink"),
                React.createElement(Stack.Item, { grow: true, styles: stackItemStyles }, "I shrink"))));
    };
    return HorizontalStackShrinkExample;
}(React.Component));
export { HorizontalStackShrinkExample };
