// office-ui-fabric-react/lib-es2015/components/Panel/examples/Panel.SmallRight.Example.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { PrimaryButton, DefaultButton } from 'office-ui-fabric-react/lib/Button';
import { ChoiceGroup } from 'office-ui-fabric-react/lib/ChoiceGroup';
import { Panel, PanelType } from 'office-ui-fabric-react/lib/Panel';
var PanelSmallRightExample = /** @class */ (function (_super) {
    tslib_1.__extends(PanelSmallRightExample, _super);
    function PanelSmallRightExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onClosePanel = function () {
            _this.setState({ showPanel: false });
        };
        _this._onRenderFooterContent = function () {
            return (React.createElement("div", null,
                React.createElement(PrimaryButton, { onClick: _this._onClosePanel, style: { 'marginRight': '8px' } }, "Save"),
                React.createElement(DefaultButton, { onClick: _this._onClosePanel }, "Cancel")));
        };
        _this._onShowPanel = function () {
            _this.setState({ showPanel: true });
        };
        _this.state = {
            showPanel: false
        };
        return _this;
    }
    PanelSmallRightExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(DefaultButton, { secondaryText: 'Opens the Sample Panel', onClick: this._onShowPanel, text: 'Open Panel' }),
            React.createElement(Panel, { isOpen: this.state.showPanel, type: PanelType.smallFixedFar, onDismiss: this._onClosePanel, headerText: 'Panel - Small, right-aligned, fixed, with footer', closeButtonAriaLabel: 'Close', onRenderFooterContent: this._onRenderFooterContent },
                React.createElement(ChoiceGroup, { options: [
                        {
                            key: 'A',
                            text: 'Option A'
                        },
                        {
                            key: 'B',
                            text: 'Option B',
                            checked: true
                        },
                        {
                            key: 'C',
                            text: 'Option C',
                            disabled: true
                        },
                        {
                            key: 'D',
                            text: 'Option D',
                            checked: true,
                            disabled: true
                        }
                    ], label: 'Pick one', required: true }))));
    };
    return PanelSmallRightExample;
}(React.Component));
export { PanelSmallRightExample };
