// @microsoft/api-extractor-model/lib/model/ApiEnum.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const ApiDeclaredItem_1 = require("../items/ApiDeclaredItem");
const ApiReleaseTagMixin_1 = require("../mixins/ApiReleaseTagMixin");
const ApiItemContainerMixin_1 = require("../mixins/ApiItemContainerMixin");
const ApiNameMixin_1 = require("../mixins/ApiNameMixin");
/**
 * Represents a TypeScript enum declaration.
 *
 * @remarks
 *
 * This is part of the {@link ApiModel} hierarchy of classes, which are serializable representations of
 * API declarations.
 *
 * `ApiEnum` represents an enum declaration such as `FontSizes` in the example below:
 *
 * ```ts
 * export enum FontSizes {
 *   Small = 100,
 *   Medium = 200,
 *   Large = 300
 * }
 * ```
 *
 * @public
 */
class ApiEnum extends ApiItemContainerMixin_1.ApiItemContainerMixin(ApiNameMixin_1.ApiNameMixin(ApiReleaseTagMixin_1.ApiReleaseTagMixin(ApiDeclaredItem_1.ApiDeclaredItem))) {
    static getCanonicalReference(name) {
        return `(${name}:enum)`;
    }
    constructor(options) {
        super(options);
    }
    /** @override */
    get kind() {
        return "Enum" /* Enum */;
    }
    /** @override */
    get members() {
        return super.members;
    }
    /** @override */
    get canonicalReference() {
        return ApiEnum.getCanonicalReference(this.name);
    }
    /** @override */
    addMember(member) {
        if (member.kind !== "EnumMember" /* EnumMember */) {
            throw new Error('Only ApiEnumMember objects can be added to an ApiEnum');
        }
        super.addMember(member);
    }
}
exports.ApiEnum = ApiEnum;
