// @microsoft/api-extractor-model/lib/model/ApiInterface.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const ApiItemContainerMixin_1 = require("../mixins/ApiItemContainerMixin");
const ApiDeclaredItem_1 = require("../items/ApiDeclaredItem");
const ApiReleaseTagMixin_1 = require("../mixins/ApiReleaseTagMixin");
const HeritageType_1 = require("./HeritageType");
const ApiNameMixin_1 = require("../mixins/ApiNameMixin");
/**
 * Represents a TypeScript class declaration.
 *
 * @remarks
 *
 * This is part of the {@link ApiModel} hierarchy of classes, which are serializable representations of
 * API declarations.
 *
 * `ApiInterface` represents a TypeScript declaration such as this:
 *
 * ```ts
 * export interface X extends Y {
 * }
 * ```
 *
 * @public
 */
class ApiInterface extends ApiItemContainerMixin_1.ApiItemContainerMixin(ApiNameMixin_1.ApiNameMixin(ApiReleaseTagMixin_1.ApiReleaseTagMixin(ApiDeclaredItem_1.ApiDeclaredItem))) {
    constructor(options) {
        super(options);
        this._extendsTypes = [];
        for (const extendsTokenRange of options.extendsTokenRanges) {
            this._extendsTypes.push(new HeritageType_1.HeritageType(this.buildExcerpt(extendsTokenRange)));
        }
    }
    static getCanonicalReference(name) {
        return `(${name}:interface)`;
    }
    /** @override */
    static onDeserializeInto(options, jsonObject) {
        super.onDeserializeInto(options, jsonObject);
        options.extendsTokenRanges = jsonObject.extendsTokenRanges;
    }
    /** @override */
    get kind() {
        return "Interface" /* Interface */;
    }
    /** @override */
    get canonicalReference() {
        return ApiInterface.getCanonicalReference(this.name);
    }
    /**
     * The list of base interfaces that this interface inherits from using the `extends` keyword.
     */
    get extendsTypes() {
        return this._extendsTypes;
    }
    /** @override */
    serializeInto(jsonObject) {
        super.serializeInto(jsonObject);
        jsonObject.extendsTokenRanges = this.extendsTypes.map(x => x.excerpt.tokenRange);
    }
}
exports.ApiInterface = ApiInterface;
