// @microsoft/api-extractor-model/lib/model/ApiPropertySignature.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const ApiPropertyItem_1 = require("../items/ApiPropertyItem");
/**
 * Represents a TypeScript property declaration that belongs to an `ApiInterface`.
 *
 * @remarks
 *
 * This is part of the {@link ApiModel} hierarchy of classes, which are serializable representations of
 * API declarations.
 *
 * `ApiPropertySignature` represents a TypeScript declaration such as the `width` and `height` members in this example:
 *
 * ```ts
 * export interface IWidget {
 *   readonly width: number;
 *   height: number;
 * }
 * ```
 *
 * Compare with {@link ApiProperty}, which represents a property belonging to a class.
 * For example, a class property can be `static` but an interface property cannot.
 *
 * @public
 */
class ApiPropertySignature extends ApiPropertyItem_1.ApiPropertyItem {
    static getCanonicalReference(name) {
        return name;
    }
    constructor(options) {
        super(options);
    }
    /** @override */
    get kind() {
        return "PropertySignature" /* PropertySignature */;
    }
    /** @override */
    get canonicalReference() {
        return ApiPropertySignature.getCanonicalReference(this.name);
    }
}
exports.ApiPropertySignature = ApiPropertySignature;
