// @microsoft/api-extractor-model/lib/model/ApiVariable.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const ApiDeclaredItem_1 = require("../items/ApiDeclaredItem");
const ApiReleaseTagMixin_1 = require("../mixins/ApiReleaseTagMixin");
const ApiNameMixin_1 = require("../mixins/ApiNameMixin");
/**
 * Represents a TypeScript variable declaration.
 *
 * @remarks
 *
 * This is part of the {@link ApiModel} hierarchy of classes, which are serializable representations of
 * API declarations.
 *
 * `ApiVariable` represents an exported `const` or `let` object such as these examples:
 *
 * ```ts
 * // A variable declaration
 * export let verboseLogging: boolean;
 *
 * // A constant variable declaration with an initializer
 * export const canvas: IWidget = createCanvas();
 * ```
 *
 * @public
 */
class ApiVariable extends ApiNameMixin_1.ApiNameMixin(ApiReleaseTagMixin_1.ApiReleaseTagMixin(ApiDeclaredItem_1.ApiDeclaredItem)) {
    /** @override */
    static onDeserializeInto(options, jsonObject) {
        super.onDeserializeInto(options, jsonObject);
        options.variableTypeTokenRange = jsonObject.variableTypeTokenRange;
    }
    static getCanonicalReference(name) {
        return name;
    }
    constructor(options) {
        super(options);
        this.variableTypeExcerpt = this.buildExcerpt(options.variableTypeTokenRange);
    }
    /** @override */
    get kind() {
        return "Variable" /* Variable */;
    }
    /** @override */
    get canonicalReference() {
        return ApiVariable.getCanonicalReference(this.name);
    }
    /** @override */
    serializeInto(jsonObject) {
        super.serializeInto(jsonObject);
        jsonObject.variableTypeTokenRange = this.variableTypeExcerpt.tokenRange;
    }
}
exports.ApiVariable = ApiVariable;
