// office-ui-fabric-react/lib/components/ContextualMenu/examples/ContextualMenu.Section.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var ContextualMenu_1 = require("office-ui-fabric-react/lib/ContextualMenu");
var Button_1 = require("office-ui-fabric-react/lib/Button");
require("./ContextualMenuExample.scss");
var ContextualMenuSectionExample = /** @class */ (function (_super) {
    tslib_1.__extends(ContextualMenuSectionExample, _super);
    function ContextualMenuSectionExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ContextualMenuSectionExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(Button_1.DefaultButton, { id: 'ContextualMenuButton1', text: 'Click for ContextualMenu', menuProps: {
                    items: [
                        {
                            key: 'section',
                            itemType: ContextualMenu_1.ContextualMenuItemType.Section,
                            sectionProps: {
                                topDivider: true,
                                bottomDivider: true,
                                title: 'Actions',
                                items: [
                                    {
                                        key: 'newItem',
                                        name: 'New',
                                    },
                                    {
                                        key: 'deleteItem',
                                        name: 'Delete',
                                    }
                                ]
                            }
                        },
                        {
                            key: 'section',
                            itemType: ContextualMenu_1.ContextualMenuItemType.Section,
                            sectionProps: {
                                title: 'Social',
                                items: [
                                    {
                                        key: 'share',
                                        name: 'Share'
                                    },
                                    {
                                        key: 'print',
                                        name: 'Print'
                                    },
                                    {
                                        key: 'music',
                                        name: 'Music',
                                    },
                                ]
                            }
                        },
                        {
                            key: 'section',
                            itemType: ContextualMenu_1.ContextualMenuItemType.Section,
                            sectionProps: {
                                title: 'Navigation',
                                items: [
                                    {
                                        key: 'Bing',
                                        name: 'Go to Bing',
                                        href: 'http://www.bing.com'
                                    }
                                ]
                            }
                        }
                    ]
                } })));
    };
    return ContextualMenuSectionExample;
}(React.Component));
exports.ContextualMenuSectionExample = ContextualMenuSectionExample;
