// @pnp/spfx-controls-react/lib/controls/chartControl/ChartControl.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ChartControl_module_scss_1 = require("./ChartControl.module.scss");
var Utilities_1 = require("office-ui-fabric-react/lib/Utilities");
var chart_js_1 = require("chart.js");
var PaletteGenerator_1 = require("./PaletteGenerator");
var AccessibleChartTable_1 = require("./AccessibleChartTable");
var telemetry = require("../../common/telemetry");
var ChartControl_types_1 = require("./ChartControl.types");
var ThemeColorHelper_1 = require("../../common/utilities/ThemeColorHelper");
var ChartControl = (function (_super) {
    __extends(ChartControl, _super);
    function ChartControl(props) {
        var _this = _super.call(this, props) || this;
        /**
         * The canvas element that will host the chart
         */
        _this._canvasElem = undefined;
        _this._linkCanvas = function (e) {
            _this._canvasElem = e;
        };
        telemetry.track('ReactChartComponent', {
            type: !!props.type,
            className: !!props.className,
            palette: !!props.palette,
            accessibility: !!props.accessibility.enable
        });
        _this.state = {
            isLoading: false,
            rejected: undefined,
            data: undefined
        };
        return _this;
    }
    /**
     * componentDidMount lifecycle hook
     */
    ChartControl.prototype.componentDidMount = function () {
        if (this.props.datapromise) {
            this._doPromise(this.props.datapromise);
        }
        else {
            this._initChart(this.props, this.props.data);
        }
    };
    /**
     * componentWillReceiveProps lifecycle hook
     *
     * @param nextProps
     */
    ChartControl.prototype.componentWillReceiveProps = function (nextProps) {
        if (nextProps.datapromise !== this.props.datapromise) {
            this.setState({
                isLoading: false
            });
            this._doPromise(nextProps.datapromise);
        }
        else {
            this._destroyChart();
            this._initChart(nextProps, this.props.data);
        }
    };
    /**
     * componentWillUnmount lifecycle hook
     */
    ChartControl.prototype.componentWillUnmount = function () {
        this._destroyChart();
    };
    /**
    * shouldComponentUpdate lifecycle hook
    *
    * @param nextProps
    * @param nextState
    */
    ChartControl.prototype.shouldComponentUpdate = function (nextProps, nextState) {
        var _a = this.props, data = _a.data, options = _a.options, plugins = _a.plugins, className = _a.className, accessibility = _a.accessibility, useTheme = _a.useTheme, palette = _a.palette;
        return data !== nextProps.data ||
            options !== nextProps.options ||
            plugins !== nextProps.plugins ||
            className !== nextProps.className ||
            useTheme !== nextProps.useTheme ||
            palette !== nextProps.palette ||
            accessibility !== nextProps.accessibility;
    };
    /**
     * Default React render method
     */
    ChartControl.prototype.render = function () {
        var _a = this.props, type = _a.type, accessibility = _a.accessibility, useTheme = _a.useTheme, options = _a.options, data = _a.data;
        // If we're still loading, try to show the loading template
        if (this.state.isLoading) {
            if (this.props.loadingtemplate) {
                if (typeof this.props.loadingtemplate === "function") {
                    return this.props.loadingtemplate();
                }
                else {
                    return this.props.loadingtemplate;
                }
            }
        }
        // If promise was rejected, try to show the rejected template
        if (this.state.rejected) {
            if (this.props.rejectedtemplate) {
                if (typeof this.props.rejectedtemplate === "function") {
                    return this.props.rejectedtemplate(this.state.rejected);
                }
                else {
                    return this.props.rejectedtemplate;
                }
            }
        }
        var alternateText = accessibility.alternateText;
        return (React.createElement("div", { className: Utilities_1.css(ChartControl_module_scss_1.default.chartComponent, (useTheme ? ChartControl_module_scss_1.default.themed : null), this.props.className) },
            React.createElement("canvas", { ref: this._linkCanvas, role: 'img', "aria-label": alternateText }),
            accessibility.enable === undefined || accessibility.enable ? (React.createElement(AccessibleChartTable_1.AccessibleChartTable, { chartType: type, data: data || this.state.data, chartOptions: options, className: accessibility.className, caption: accessibility.caption, summary: accessibility.summary, onRenderTable: accessibility.onRenderTable })) : null));
    };
    /**
     * Triggers an update of the chart.
     * This can be safely called after updating the data object.
     * This will update all scales, legends, and then re-render the chart.
     * @param config duration (number): Time for the animation of the redraw in milliseconds
     * lazy (boolean): If true, the animation can be interrupted by other animations
     * easing (string): The animation easing function.
     */
    ChartControl.prototype.update = function (config) {
        this._chart.update(config);
    };
    /**
     * Triggers a redraw of all chart elements.
     * Note, this does not update elements for new data. Use .update() in that case.
     * @param config  duration is the time for the animation of the redraw in milliseconds
      lazy is a boolean. If true, the animation can be interrupted by other animations
     */
    ChartControl.prototype.renderChart = function (config) {
        this._chart.render(config);
    };
    /**
     Use this to stop any current animation loop.
     This will pause the chart during any current animation frame.
     Call .render() to re-animate.
     */
    ChartControl.prototype.stop = function () {
        this._chart.stop();
    };
    /**
     Will clear the chart canvas.
     Used extensively internally between animation frames, but you might find it useful.
     */
    ChartControl.prototype.clear = function () {
        this._chart.clear();
    };
    /**
   Returns a base 64 encoded string of the chart in it's current state.
   @returns {string} A base-64 encoded PNG data URL containing image of the chart in its current state
   */
    ChartControl.prototype.toBase64Image = function () {
        return this._chart.toBase64Image();
    };
    /**
     Return the chartjs instance
     */
    //  tslint:disable-next-line no-any
    ChartControl.prototype.getChart = function () {
        return this._chart;
    };
    /**
    Return the canvass element that contains the chart
    @returns {HTMLCanvasElement} the canvas element containig the chart
     */
    ChartControl.prototype.getCanvas = function () {
        return this._chart.canvas;
    };
    /**
     Looks for the element under the event point,
     then returns all elements from that dataset.
     This is used internally for 'dataset' mode highlighting
     * @param e An array of elements
     */
    ChartControl.prototype.getDatasetAtEvent = function (e) {
        return this.getChart().getDatasetAtEvent(e);
    };
    /**
   * Calling getElementAtEvent(event) on your Chart instance passing an argument of an event,
   * or jQuery event, will return the single element at the event position.
   * If there are multiple items within range, only the first is returned.
   * The value returned from this method is an array with a single parameter.
   * An array is used to keep a consistent API between the get*AtEvent methods.
   * @param e the first element at the event point.
   */
    ChartControl.prototype.getElementAtEvent = function (e) {
        return this.getChart().getElementAtEvent(e);
    };
    /**
    Looks for the element under the event point, then returns all elements
    at the same data index. This is used internally for 'label' mode highlighting.
   Calling getElementsAtEvent(event) on your Chart instance passing an argument of an
   event, or jQuery event, will return the point elements that are at that
   the same position of that event.
    * @param e
    */
    ChartControl.prototype.getElementsAtEvent = function (e) {
        return this.getChart().getElementsAtEvent(e);
    };
    /**
     * Initializes the chart
     * @param props chart control properties
     */
    ChartControl.prototype._initChart = function (props, data) {
        var _this = this;
        var _a = this.props, options = _a.options, type = _a.type, plugins = _a.plugins, useTheme = _a.useTheme;
        // add event handlers -- if they weren't already provided through options
        if (this.props.onClick !== undefined) {
            if (options.onClick === undefined) {
                options.onClick = this.props.onClick;
            }
        }
        // Add onhover
        if (this.props.onHover !== undefined) {
            if (options.onHover === undefined) {
                options.onHover = function (event, activeElements) {
                    _this.props.onHover(_this.getChart(), event, activeElements);
                };
            }
        }
        // Add onResize
        // Note that onResize won't work unless the chart is
        // position: relative and has a height and width defined
        if (this.props.onResize !== undefined) {
            if (options.onResize === undefined) {
                options.onResize = function (newSize) {
                    _this.props.onResize(_this.getChart(), newSize);
                };
            }
        }
        this._applyDatasetPalette(data);
        if (useTheme) {
            this._applyChartThemes();
        }
        this._chart = new chart_js_1.Chart(this._canvasElem, {
            type: type,
            data: data,
            options: options,
            plugins: plugins
        });
    };
    ChartControl.prototype._applyDatasetPalette = function (data) {
        var _this = this;
        try {
            // Get the dataset
            var datasets = data.datasets;
            if (datasets !== undefined) {
                datasets.forEach(function (dataset) {
                    if (dataset.backgroundColor === undefined) {
                        var datasetLength = dataset.data.length;
                        if (datasetLength) {
                            dataset.backgroundColor = PaletteGenerator_1.PaletteGenerator.GetPalette(_this.props.palette, datasetLength);
                        }
                    }
                });
            }
        }
        catch (error) {
        }
    };
    ChartControl.prototype._applyChartThemes = function () {
        try {
            chart_js_1.Chart.defaults.global.defaultFontColor = ThemeColorHelper_1.ThemeColorHelper.GetThemeColor(ChartControl_module_scss_1.default.defaultFontColor);
            chart_js_1.Chart.defaults.global.defaultFontFamily = ChartControl_module_scss_1.default.defaultFontFamily;
            chart_js_1.Chart.defaults.global.defaultFontSize = this._getFontSizeNumber(ChartControl_module_scss_1.default.defaultFontSize);
            chart_js_1.Chart.defaults.global.title.fontColor = ThemeColorHelper_1.ThemeColorHelper.GetThemeColor(ChartControl_module_scss_1.default.titleColor);
            chart_js_1.Chart.defaults.global.title.fontFamily = ChartControl_module_scss_1.default.titleFont;
            chart_js_1.Chart.defaults.global.title.fontSize = this._getFontSizeNumber(ChartControl_module_scss_1.default.titleFontSize);
            chart_js_1.Chart.defaults.global.legend.labels.fontColor = ThemeColorHelper_1.ThemeColorHelper.GetThemeColor(ChartControl_module_scss_1.default.legendColor);
            chart_js_1.Chart.defaults.global.legend.labels.fontFamily = ChartControl_module_scss_1.default.legendFont;
            chart_js_1.Chart.defaults.global.legend.labels.fontSize = this._getFontSizeNumber(ChartControl_module_scss_1.default.legendFontSize);
            chart_js_1.Chart.defaults.global.tooltips.backgroundColor = ThemeColorHelper_1.ThemeColorHelper.GetThemeColor(ChartControl_module_scss_1.default.tooltipBackgroundColor);
            chart_js_1.Chart.defaults.global.tooltips.bodyFontColor = ThemeColorHelper_1.ThemeColorHelper.GetThemeColor(ChartControl_module_scss_1.default.tooltipBodyColor);
            chart_js_1.Chart.defaults.global.tooltips.bodyFontFamily = ChartControl_module_scss_1.default.tooltipFont;
            chart_js_1.Chart.defaults.global.tooltips.bodyFontSize = this._getFontSizeNumber(ChartControl_module_scss_1.default.tooltipFontSize);
            chart_js_1.Chart.defaults.global.tooltips.titleFontColor = ThemeColorHelper_1.ThemeColorHelper.GetThemeColor(ChartControl_module_scss_1.default.tooltipTitleColor);
            chart_js_1.Chart.defaults.global.tooltips.titleFontFamily = ChartControl_module_scss_1.default.tooltipTitleFont;
            chart_js_1.Chart.defaults.global.tooltips.titleFontSize = this._getFontSizeNumber(ChartControl_module_scss_1.default.tooltipTitleFontSize);
            chart_js_1.Chart.defaults.global.tooltips.footerFontColor = ThemeColorHelper_1.ThemeColorHelper.GetThemeColor(ChartControl_module_scss_1.default.tooltipFooterColor);
            chart_js_1.Chart.defaults.global.tooltips.footerFontFamily = ChartControl_module_scss_1.default.tooltipFooterFont;
            chart_js_1.Chart.defaults.global.tooltips.footerFontSize = this._getFontSizeNumber(ChartControl_module_scss_1.default.tooltipFooterFontSize);
            chart_js_1.Chart.defaults.global.tooltips.borderColor = ThemeColorHelper_1.ThemeColorHelper.GetThemeColor(ChartControl_module_scss_1.default.tooltipBorderColor);
            if (chart_js_1.Chart.defaults
                && chart_js_1.Chart.defaults.scale
                && chart_js_1.Chart.defaults.scale.gridLines
                && chart_js_1.Chart.defaults.scale.gridLines.color) {
                chart_js_1.Chart.defaults.scale.gridLines.color = ThemeColorHelper_1.ThemeColorHelper.GetThemeColor(ChartControl_module_scss_1.default.lineColor);
            }
        }
        catch (error) {
        }
    };
    ChartControl.prototype._destroyChart = function () {
        try {
            if (this._chart !== undefined) {
                this._chart.destroy();
            }
        }
        catch (error) {
        }
    };
    // private _getThemeColor(value: string): string {
    //   try {
    //     if (value.indexOf('theme:') > 0) {
    //       // This value has a theme substitution
    //       const themeParts: string[] = value.replace('[', '').replace(']', '').replace('"', '').split(',');
    //       let defaultValue: string = undefined;
    //       let themeValue: string = undefined;
    //       // Break the theme string into it's components
    //       themeParts.forEach(themePart => {
    //         if (themePart.indexOf('theme:') >= 0) {
    //           themeValue = themePart.replace('theme:', '');
    //         } else if (themePart.indexOf('default:') >= 0) {
    //           defaultValue = themePart.replace('default:', '').replace('"', '').trim();
    //         }
    //       });
    //       // If there is a theme value, try to read from environment
    //       if (themeValue !== undefined) {
    //         try {
    //           // This should definitely be easier to do in SPFx!
    //           // tslint:disable-next-line
    //           const themeStateVariable: any = window.__themeState__;
    //           if (themeStateVariable === undefined) {
    //             return defaultValue;
    //           }
    //           const themeState: {} = themeStateVariable.theme;
    //           if (themeState === undefined) {
    //             return defaultValue;
    //           }
    //           for (const varName in themeState) {
    //             if (!themeState.hasOwnProperty(varName)) {
    //               continue;
    //             }
    //             // Cheesy cleanup of variables to remove extra quotes
    //             if (varName === themeValue) {
    //               return themeState[varName].replace('"', '').trim();
    //             }
    //           }
    //         } catch (error) {
    //           // do nothing
    //         }
    //         return defaultValue;
    //       }
    //     }
    //   } catch (error) {
    //   }
    //   return value;
    // }
    // Reads one of the Office Fabric defined font sizes
    // and converts to a number
    ChartControl.prototype._getFontSizeNumber = function (value) {
        try {
            return parseInt(value.replace('px', ''), 10);
        }
        catch (error) {
            return undefined;
        }
    };
    /**
     * Gets the results of a promise and returns it to the chart
     * @param promise
     */
    ChartControl.prototype._doPromise = function (promise) {
        var _this = this;
        this.setState({
            isLoading: true
        }, function () {
            promise.then(function (results) {
                _this.setState({
                    isLoading: false,
                    data: results
                }, function () {
                    _this._initChart(_this.props, results);
                });
            }, function (rejected) {
                _this.setState({
                    isLoading: false,
                    rejected: rejected
                });
            });
        });
    };
    /**
     * Sets default properties
     */
    ChartControl.defaultProps = {
        // We want accessibility on by default
        // -- it's the law in some countries!!!
        accessibility: {
            enable: true
        },
        useTheme: true,
        palette: ChartControl_types_1.ChartPalette.OfficeColorful1,
        // Make charts responsive so that they fit within their
        // parent elements
        options: {
            responsive: true,
            maintainAspectRatio: true
        }
    };
    return ChartControl;
}(React.Component));
exports.ChartControl = ChartControl;
