// office-ui-fabric-react/lib-amd/components/pickers/Suggestions/SuggestionsItem.styles.js
define(["require", "exports", "../../../Styling"], function (require, exports, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var GlobalClassNames = {
        root: 'ms-Suggestions-item',
        itemButton: 'ms-Suggestions-itemButton',
        closeButton: 'ms-Suggestions-closeButton',
        isSuggested: 'is-suggested'
    };
    function getStyles(props) {
        var className = props.className, theme = props.theme, suggested = props.suggested;
        var palette = theme.palette;
        var classNames = Styling_1.getGlobalClassNames(GlobalClassNames, theme);
        return {
            root: [
                classNames.root,
                {
                    display: 'flex',
                    alignItems: 'stretch',
                    boxSizing: 'border-box',
                    width: '100%',
                    position: 'relative',
                    selectors: {
                        '&:hover': {
                            background: palette.neutralLighter
                        },
                        '&:hover .ms-Suggestions-closeButton': {
                            display: 'block'
                        }
                    }
                },
                suggested && [
                    classNames.isSuggested,
                    {
                        background: palette.neutralLight,
                        selectors: (_a = {
                                ':hover': {
                                    background: palette.neutralTertiaryAlt
                                }
                            },
                            _a[Styling_1.HighContrastSelector] = {
                                background: 'Highlight',
                                color: 'HighlightText',
                                MsHighContrastAdjust: 'none'
                            },
                            _a)
                    }
                ],
                className
            ],
            itemButton: [
                classNames.itemButton,
                {
                    width: '100%',
                    padding: 0,
                    border: 'none',
                    height: '100%',
                    selectors: (_b = {},
                        _b[Styling_1.HighContrastSelector] = {
                            color: 'WindowText'
                        },
                        _b[':hover'] = {
                            color: palette.neutralDark
                        },
                        _b)
                }
            ],
            closeButton: [
                classNames.closeButton,
                {
                    display: 'none',
                    color: palette.neutralSecondary,
                    padding: '0 4px',
                    height: 'auto',
                    width: 32,
                    selectors: (_c = {
                            ':hover, :active': {
                                background: palette.neutralTertiaryAlt,
                                color: palette.neutralDark
                            }
                        },
                        _c[Styling_1.HighContrastSelector] = {
                            color: 'WindowText'
                        },
                        _c)
                },
                suggested && {
                    selectors: {
                        ':hover, :active': {
                            background: palette.neutralTertiary,
                            color: palette.neutralPrimary
                        }
                    }
                }
            ]
        };
        var _a, _b, _c;
    }
    exports.getStyles = getStyles;
});
