// office-ui-fabric-react/lib-amd/components/Stack/examples/Stack.Horizontal.HorizontalAlign.Example.js
define(["require", "exports", "tslib", "react", "../Stack", "office-ui-fabric-react/lib/Styling"], function (require, exports, tslib_1, React, Stack_1, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var HorizontalStackHorizontalAlignExample = /** @class */ (function (_super) {
        tslib_1.__extends(HorizontalStackHorizontalAlignExample, _super);
        function HorizontalStackHorizontalAlignExample() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        HorizontalStackHorizontalAlignExample.prototype.render = function () {
            var stackStyles = {
                root: {
                    background: Styling_1.DefaultPalette.themeTertiary,
                    selectors: {
                        '> *': {
                            alignItems: 'center',
                            background: Styling_1.DefaultPalette.themePrimary,
                            color: Styling_1.DefaultPalette.white,
                            display: 'flex',
                            height: 50,
                            justifyContent: 'center',
                            width: 50
                        }
                    }
                }
            };
            var stackTokens = { childrenGap: 5 };
            return (React.createElement(Stack_1.Stack, { tokens: stackTokens },
                React.createElement("span", null, "Left-aligned"),
                React.createElement(Stack_1.Stack, { horizontal: true, horizontalAlign: "start", styles: stackStyles },
                    React.createElement("span", null, "1"),
                    React.createElement("span", null, "2"),
                    React.createElement("span", null, "3")),
                React.createElement("span", null, "Horizontally centered"),
                React.createElement(Stack_1.Stack, { horizontal: true, horizontalAlign: "center", styles: stackStyles },
                    React.createElement("span", null, "1"),
                    React.createElement("span", null, "2"),
                    React.createElement("span", null, "3")),
                React.createElement("span", null, "Right-aligned"),
                React.createElement(Stack_1.Stack, { horizontal: true, horizontalAlign: "end", styles: stackStyles },
                    React.createElement("span", null, "1"),
                    React.createElement("span", null, "2"),
                    React.createElement("span", null, "3")),
                React.createElement("span", null, "Horizontal space around items"),
                React.createElement(Stack_1.Stack, { horizontal: true, horizontalAlign: "space-around", styles: stackStyles },
                    React.createElement("span", null, "1"),
                    React.createElement("span", null, "2"),
                    React.createElement("span", null, "3")),
                React.createElement("span", null, "Horizontal space between items"),
                React.createElement(Stack_1.Stack, { horizontal: true, horizontalAlign: "space-between", styles: stackStyles },
                    React.createElement("span", null, "1"),
                    React.createElement("span", null, "2"),
                    React.createElement("span", null, "3")),
                React.createElement("span", null, "Items horizontally evenly spaced"),
                React.createElement(Stack_1.Stack, { horizontal: true, horizontalAlign: "space-evenly", styles: stackStyles },
                    React.createElement("span", null, "1"),
                    React.createElement("span", null, "2"),
                    React.createElement("span", null, "3"))));
        };
        return HorizontalStackHorizontalAlignExample;
    }(React.Component));
    exports.HorizontalStackHorizontalAlignExample = HorizontalStackHorizontalAlignExample;
});
