// office-ui-fabric-react/lib-amd/components/Shimmer/examples/Shimmer.Styling.Example.js
// @codepen
define(["require", "exports", "tslib", "react", "office-ui-fabric-react/lib/Shimmer", "office-ui-fabric-react/lib/Styling"], function (require, exports, tslib_1, React, Shimmer_1, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var wrapperClass = Styling_1.mergeStyles({
        padding: 2,
        selectors: {
            '& > *': {
                margin: '10px 0'
            }
        }
    });
    var ShimmerStylingExample = /** @class */ (function (_super) {
        tslib_1.__extends(ShimmerStylingExample, _super);
        function ShimmerStylingExample() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this._getShimmerStyles = function (props) {
                return {
                    shimmerWrapper: [
                        {
                            backgroundColor: '#deecf9',
                            backgroundImage: 'linear-gradient(to right, rgba(255, 255, 255, 0) 0%, #c7e0f4 50%, rgba(255, 255, 255, 0) 100%)'
                        }
                    ]
                };
            };
            return _this;
        }
        ShimmerStylingExample.prototype.render = function () {
            return (React.createElement("div", { className: wrapperClass },
                React.createElement(Shimmer_1.Shimmer, { width: "75%", styles: this._getShimmerStyles }),
                React.createElement(Shimmer_1.Shimmer, { width: "75%", styles: this._getShimmerStyles }),
                React.createElement(Shimmer_1.Shimmer, { width: "75%", styles: this._getShimmerStyles }),
                React.createElement(Shimmer_1.Shimmer, { width: "75%", styles: this._getShimmerStyles }),
                React.createElement(Shimmer_1.Shimmer, { width: "75%", styles: this._getShimmerStyles })));
        };
        return ShimmerStylingExample;
    }(React.Component));
    exports.ShimmerStylingExample = ShimmerStylingExample;
});
